<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Summary • CRBHits</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>




<meta property="og:title" content="Summary" />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="articles/V02KaKsVignette.html">KaKs Vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Summary</h1>
    </div>


<hr>
<p>title: ‘CRBHits: From Conditional Reciprocal Best Hits to Codon Alignments and Ka/Ks in R’ tags:</p>
<ul>
<li>R</li>
<li>reciprocal best hit</li>
<li>conditional reciprocal best hit</li>
<li>codon alignment</li>
<li>Ka/Ks</li>
<li>dN/dS</li>
<li>tandem duplicated genes</li>
<li>synteny authors:</li>
<li>name: Kristian K Ullrich orcid: 0000-0003-4308-9626 affiliation: “1” affiliations:</li>
<li>name: Max Planck Institute for Evolutionary Biology, Scientific IT group, August Thienemann Str. 2, 24306 Plön index: 1 date: 29 Oct 2020 bibliography: paper.bib</li>
</ul>
<hr>
<div id="summary" class="section level1">

<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is a coding sequence (CDS) analysis pipeline in <a href="https://cran.r-project.org/">R</a> [@team2013r]. It reimplements the Conditional Reciprocal Best Hit (CRBH) algorithm <a href="https://github.com/cboursnell/crb-blast">crb-blast</a> and covers all necessary steps from sequence similarity searches, codon alignments to Ka/Ks calculations and synteny. The new R package targets ecology, population and evolutionary biologists working in the field of comparative genomics.</p>
<p>The Reciprocal Best Hit (RBH) approach is commonly used in bioinformatics to show that two sequences evolved from a common ancestral gene. In other words, RBH tries to find orthologous protein sequences within and between species. These orthologous sequences can be further analysed to evaluate protein family evolution, infer phylogenetic trees and to annotate protein function [@altenhoff2019inferring]. The initial sequence search step is classically performed with the Basic Local Alignment Search Tool (blast) [@altschul1990basic] and due to evolutionary constraints, in most cases protein coding sequences are compared between two species. Downstream analysis use the resulting RBH to cluster sequence pairs and build so-called orthologous groups like e.g. <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a> [@emms2015orthofinder] and other tools.</p>
<p>The CRBH algorithm was introduced by @aubry2014deep and builds upon the traditional RBH approach to find additional orthologous sequences between two sets of sequences. As described earlier [@aubry2014deep; @scott2017shmlast], CRBH uses the sequence search results to fit an expect value (E-value) cutoff given each RBH to subsequently add sequence pairs to the list of bona-fide orthologs given their alignment length.</p>
<p>Unfortunately, as mentioned by @scott2017shmlast, the original implementation of CRBH (<a href="https://github.com/cboursnell/crb-blast">crb-blast</a>) lag improved blast-like search algorithm to speed up the analysis. As a consequence, @scott2017shmlast ported CRBH to python <a href="https://github.com/camillescott/shmlast">shmlast</a>, while <a href="https://github.com/camillescott/shmlast">shmlast</a> cannot deal with IUPAC nucleotide code so far.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> constitutes a new R package, which build upon previous implementations and ports CRBH into the <a href="https://cran.r-project.org/">R</a> environment, which is popular among biologists. <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> improve CRBH by additional implemented filter steps [@rost1999twilight] and the possibility to apply custom filters prior E-value fitting. Further, the resulting CRBH pairs can be evaluated for the presence of tandem duplicated genes, gene order based syntenic groups and evolutionary rates.</p>
<p><img src="figure1.png" alt="Overview of the two main pipeline function and its subtasks. cds2rbh(): from CDS to CRBHit pairs; rbh2kaks(): from CRBHit pairs to Ka/Ks values.\label{fig:functions}"></p>
</div>
<div id="coding-sequence-analysis-and-synteny" class="section level1">
<h1 class="hasAnchor">
<a href="#coding-sequence-analysis-and-synteny" class="anchor"></a>Coding sequence analysis and synteny</h1>
<p>Calculating synonymous (Ks) and nonsynonymous substitutions (Ka) per orthologous sequence pair is a common task for evolutionary biologists, since its ratio Ka/Ks can be used as an indicator of selective pressure acting on a protein [@kryazhimskiy2008population]. However, this task is computational more demanding and consist of at least two steps, namely codon sequence alignment creation and Ka/Ks calculation. Further, the codon sequence alignment step consist of three subtasks, namely coding nucleotide to protein sequence translation, pairwise protein sequence alignment calculation and converting the protein sequence alignment back into a codon based alignment.</p>
<p>Downstream of CRBH creation, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> features all above mentioned steps and subtasks. <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> has the ability to directly create codon alignments within R with the help of the widely used R package <a href="https://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> [@pages2017biostrings] (more than 200k downloads per year since 2014). These codon alignments can be subsequently used to calculate synonymous and nonsynonymous substitutions per sequence pair and is implemented in a multithreaded fashion either via the R package <a href="https://cran.r-project.org/web/packages/seqinr/index.html">seqinr</a> [@charif2007seqinr] or the use of an R external tool <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> [@wang2010kaks_calculator].</p>
<p>As gene duplication is one driving force in evolution [@ohno1970evolution], the classification of genes as duplicates is one important step to provide us with insights into the molecular events responsible for the current genome architecture of species [@haas2004]. New long-read sequencing technology make more and more chromosome scale assemblies for model and non-model species available. The resulting chromosomal gene order information can be used with sequence similarity scores to classify genes into different types of duplication events, like tandem duplicates or chromosomal segments (syntenic regions) derived from e.g. whole-genome duplication. <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> features this classification step via the integration of the R external tool <a href="http://dagchainer.sourceforge.net/">DAGchainer</a> [@haas2004] and offers the possibility to directly link it with evolutionary rate estimations (see \autoref{fig:synteny}).</p>
</div>
<div id="implementation" class="section level1">
<h1 class="hasAnchor">
<a href="#implementation" class="anchor"></a>Implementation</h1>
<p>Like <a href="https://github.com/camillescott/shmlast">shmlast</a>, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> benefits from the blast-like sequence search software <a href="http://last.cbrc.jp/">LAST</a>[@kielbasa2011adaptive] and plots the fitted model of the CRBH E-value based algorithm. In addition, users can filter the hit pairs prior to CRBH fitting for other criteria like query coverage, protein identity and/or the twilight zone of protein sequence alignments according to</p>
<p>@rost1999twilight. The implemented filter uses equation 2 [see @rost1999twilight]:</p>
<p><br><span class="math display">$$f(x_{\text{hit pair}}) = \begin{cases}
100 \text{ , for } L_{\text{hit pair}} &lt; 11 \\
480 * L^{-0.32 * (1 + e^{\frac{-L}{1000}})} \text{ , for } L_{\text{hit pair}} \leq 450 \\
19.5 \text{ , for } L_{\text{hit pair}} &gt; 450
\end{cases}$$</span><br></p>
<p>where <span class="math inline"><em>x</em><sub>hit pair</sub></span> is the expected protein identity given the alignment length <span class="math inline"><em>L</em><sub>hit pair</sub></span>. If the actual protein identity of a hit pair exceeds the expected protein identity (<span class="math inline"><em>p</em><em>i</em><em>d</em><em>e</em><em>n</em><em>t</em><sub>hit pair</sub> ≥ <em>f</em>(<em>x</em><sub>hit pair</sub>)</span>), it is retained for subsequent CRBH calculation.</p>
<p>In contrast to previous implementations, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> only take coding nucleotide sequences (CDS) as the query and target inputs. This is due to the downstream functionality of <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> to directly calculate codon alignments within R, which rely on CDS. The inputs are translated into protein sequences, aligned globally [@smith1981identification] and converted into codon alignments.</p>
<p>Functions are completely coded in R and only the external prerequisites (<a href="http://last.cbrc.jp/">LAST</a>, <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> and <a href="http://dagchainer.sourceforge.net/">DAGchainer</a>) need to be compiled. However, all of them are forked within <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> and can be easily build with the dedicated R functions <code><a href="reference/make_last.html">make_last()</a></code>, <code><a href="reference/make_KaKs_Calculator2.html">make_KaKs_Calculator2()</a></code> and <code><a href="reference/make_dagchainer.html">make_dagchainer()</a></code>. Further, users can create their own RBH filters before CRBH calculation.</p>
</div>
<div id="functions-and-examples" class="section level1">
<h1 class="hasAnchor">
<a href="#functions-and-examples" class="anchor"></a>Functions and Examples</h1>
<p>The following example shows how to obtain CRBHit pairs between the coding sequences of <em>Schizosaccharomyces pombe</em> (fission yeast) [@wood2012pombase] and <em>Nematostella vectensis</em> (starlet sea anemone) [@apweiler2004protein] by using two URLs as input strings and multiple threads for calculation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a></span><span class="op">)</span>
<span class="co">#set URLs for Schizosaccharomyces pombe (fission yeast)</span>
<span class="co">#and Nematostella vectensis (starlet sea anemone) from NCBI Genomes</span>
<span class="va">cds1.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/945/"</span>,
               <span class="st">"GCF_000002945.1_ASM294v2/"</span>,
               <span class="st">"GCF_000002945.1_ASM294v2_cds_from_genomic.fna.gz"</span><span class="op">)</span>
<span class="va">cds2.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/209/225/"</span>,
               <span class="st">"GCF_000209225.1_ASM20922v1/"</span>,
               <span class="st">"GCF_000209225.1_ASM20922v1_cds_from_genomic.fna.gz"</span><span class="op">)</span>
<span class="co">#calculate CBRBhit pairs</span>
<span class="va">cds1.cds2.crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/cdsfile2rbh.html">cdsfile2rbh</a></span><span class="op">(</span><span class="va">cds1.url</span>, <span class="va">cds2.url</span>, longest.isoform <span class="op">=</span> <span class="cn">TRUE</span>,
                              isoform.source <span class="op">=</span> <span class="st">"NCBI"</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                              threads <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="co">#get help ?cdsfile2rbh</span></code></pre></div>
<p><img src="figure2.png" alt="Accepted condition reciprocal best hits based on RBH fitting.\label{fig:crbh}"></p>
<p>The obtained CRBHit pairs can also be used to calculate synonymous (Ks) and nonsynonymous (Ka) substitutions per hit pair using either the model from @li1993unbiased or from @yang2000estimating.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#download and simultaneously get longest isoform for</span>
<span class="co">#Schizosaccharomyces pombe (fission yeast) and</span>
<span class="co">#Nematostella vectensis (starlet sea anemone)</span>
<span class="va">cds1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cds1.url</span><span class="op">)</span><span class="op">)</span>
<span class="va">cds2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cds2.url</span><span class="op">)</span><span class="op">)</span>
<span class="co">#calculate Ka/Ks values for each CRBHit pair</span>
<span class="va">cds1.cds2.kaks.Li</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span><span class="va">cds1.cds2.crbh</span>, <span class="va">cds1</span>, <span class="va">cds2</span>,
                              model <span class="op">=</span> <span class="st">"Li"</span>, threads <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="va">cds1.cds2.kaks.YN</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span><span class="va">cds1.cds2.crbh</span>, <span class="va">cds1</span>, <span class="va">cds2</span>,
                              model <span class="op">=</span> <span class="st">"YN"</span>, threads <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
                              
<span class="co">#get help ?rbh2kaks</span></code></pre></div>
<p>Given the annotated chromosomal gene positions it is also possible to assign tandem duplicated genes per chromosome and directly compute chains of syntenic genes via the use of the R external tool <a href="http://dagchainer.sourceforge.net/">DAGchainer</a>[@haas2004]. Here, <em>Arabidopsis thaliana</em> is compared to itself (so called selfblast) and syntenic groups vsiualized by their Ks values.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#download and simultaneously get longest isoform for</span>
<span class="co">#Arabidopsis thaliana</span>
<span class="va">cds3.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/"</span>,
               <span class="st">"arabidopsis_thaliana/cds/"</span>,
               <span class="st">"Arabidopsis_thaliana.TAIR10.cds.all.fa.gz"</span><span class="op">)</span>
<span class="va">cds3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cds3.url</span><span class="op">)</span>, <span class="st">"ENSEMBL"</span><span class="op">)</span>
<span class="co">#extract gene position and chromosomal gene order</span>
<span class="va">cds3.genepos</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/cds2genepos.html">cds2genepos</a></span><span class="op">(</span><span class="va">cds3</span>, source <span class="op">=</span> <span class="st">"ENSEMBL"</span><span class="op">)</span>
<span class="co">#calculate CBRBhit pairs</span>
<span class="va">cds3.selfblast.crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds3</span>, <span class="va">cds3</span>, longest.isoform <span class="op">=</span> <span class="cn">TRUE</span>,
                               qcov <span class="op">=</span> <span class="fl">0.5</span>, rost1999 <span class="op">=</span> <span class="cn">TRUE</span>,
                               isoform.source <span class="op">=</span> <span class="st">"ENSEMBL"</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                               threads <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
<span class="co">#compute chains of syntenic genes and plot chr1, chr2, chr3, chr4, chr5</span>
<span class="va">cds3.selfblast.synteny</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/rbh2dagchainer.html">rbh2dagchainer</a></span><span class="op">(</span><span class="va">cds3.selfblast.crbh</span>,
                                         <span class="va">cds3.genepos</span>, <span class="va">cds3.genepos</span>,
                                         plotDotPlot <span class="op">=</span> <span class="cn">TRUE</span>,
                                         select.chr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"5"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#calculate Ka/Ks values for each CRBHit pair</span>
<span class="va">cds3.selfblast.kaks.Li</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span><span class="va">cds3.selfblast.crbh</span>, <span class="va">cds3</span>, <span class="va">cds3</span>,
                                   model <span class="op">=</span> <span class="st">"Li"</span>, threads <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>

<span class="co">#get help ?rbh2dagchainer</span>
<span class="co">#get help ?plot.dagchainer</span>
<span class="co">#get help ?plot.kaks</span></code></pre></div>
<p><img src="figure3.png" alt="Selfblast CRBHit pair results for Arabidopsis thaliana. (A) DAGchainer dotplot per chromosome colored by syntenic group and (B) colored by Ks. (C) Histogram of Ks values colored by syntenic group.\label{fig:synteny}"></p>
<p>Table: Performance comparison for CRBHit pair (<em>Schizosaccharomyces pombe</em> vs. <em>Nematostella vectensis</em>) and Ka/Ks calculations (Intel Xeon CPU E5-2620 v3 @ 2.40GHz; 3411 hit pairs; 2 x Threads).\label{tab:performance}</p>
<table>
<thead><tr class="header">
<th>Number of Threads</th>
<th>1</th>
<th>2</th>
<th>4</th>
<th>8</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Runtime of CRBH(shmlast v1.6) in sec</td>
<td>36 (s)</td>
<td>24 (s)</td>
<td>18 (s)</td>
<td>16 (s)</td>
</tr>
<tr class="even">
<td>Runtime of CRBH(CRBHits) in sec</td>
<td>18 (s)</td>
<td>10 (s)</td>
<td>7 (s)</td>
<td>6 (s)</td>
</tr>
<tr class="odd">
<td>Runtime of kaks.Li in sec</td>
<td>586 (s)</td>
<td>327 (s)</td>
<td>190 (s)</td>
<td>128 (s)</td>
</tr>
<tr class="even">
<td>Runtime of kaks.YN in sec</td>
<td>667 (s)</td>
<td>361 (s)</td>
<td>202 (s)</td>
<td>133 (s)</td>
</tr>
</tbody>
</table>
</div>
<div id="conclusions" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusions" class="anchor"></a>Conclusions</h1>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> implements CRBH in <a href="https://cran.r-project.org/">R</a> (see \autoref{fig:crbh}), can be used to calculate codon alignment based nucleotide diversities (Ka/Ks) and synteny, in a multithreaded fashion (see \autoref{tab:performance}).</p>
</div>
<div id="availability" class="section level1">
<h1 class="hasAnchor">
<a href="#availability" class="anchor"></a>Availability</h1>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is an open source software made available under the MIT license. It can be installed from its gitlab repository using the <a href="https://devtools.r-lib.org">devtools</a> package.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>devtools<span class="op">::</span><span class="kw">install_gitlab</span>(<span class="st">"mpievolbio-it/crbhits"</span>, </span>
<span id="cb4-2"><a href="#cb4-2"></a> <span class="dt">host =</span> <span class="st">"https://gitlab.gwdg.de"</span>)<span class="st">", build_vignettes = TRUE)</span></span></code></pre></div>
<p>The R package website, which contain a detailed HOWTO to install the prerequisites (mentioned above) and package vignettes are availbale at <a href="https://mpievolbio-it.pages.gwdg.de/crbhits">https://mpievolbio-it.pages.gwdg.de/crbhits</a>.</p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


