<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>rbh2dagchainer — rbh2dagchainer • CRBHits</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="rbh2dagchainer — rbh2dagchainer"><meta property="og:description" content="This function runs DAGchainer
(http://dagchainer.sourceforge.net/) given CRBHit pairs and gene positions
for both cds1 and cds2. The default options are set to not compare gene
positions in base pairs but instead using gene order (gene.idx)."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/V02KaKsVignette.html">KaKs Vignette</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">
    <span class="fab fa-gitlab fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>rbh2dagchainer</h1>
    <small class="dont-index">Source: <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/develR/rbh2dagchainer.R" class="external-link"><code>R/rbh2dagchainer.R</code></a></small>
    <div class="hidden name"><code>rbh2dagchainer.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function runs DAGchainer
(http://dagchainer.sourceforge.net/) given CRBHit pairs and gene positions
for both cds1 and cds2. The default options are set to not compare gene
positions in base pairs but instead using gene order (gene.idx).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rbh2dagchainer</span><span class="op">(</span></span>
<span>  <span class="va">rbhpairs</span>,</span>
<span>  selfblast1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  selfblast2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene.position.cds1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene.position.cds2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dagchainerpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">find.package</a></span><span class="op">(</span><span class="st">"CRBHits"</span><span class="op">)</span>, <span class="st">"/extdata/dagchainer/"</span><span class="op">)</span>,</span>
<span>  gap_open_penalty <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gap_extension_penalty <span class="op">=</span> <span class="op">-</span><span class="fl">3</span>,</span>
<span>  gap_length <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  max_match_score <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  max_dist_allowed <span class="op">=</span> <span class="fl">2e+05</span>,</span>
<span>  max_evalue <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  ignore_tandem <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  only_tandem <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min_number_aligned_pairs <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"bp"</span>,</span>
<span>  plotDotPlot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  DotPlotTitle <span class="op">=</span> <span class="st">"DAGchainer results"</span>,</span>
<span>  colorBy <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  kaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ka.max <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  ks.max <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  ka.min <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ks.min <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  select.chr <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>rbhpairs</dt>
<dd><p>(conditional-)reciprocal best hit (CRBHit) pair result
(see <code><a href="cds2rbh.html">cds2rbh</a></code>) [mandatory]</p></dd>


<dt>selfblast1</dt>
<dd><p>(conditional-)reciprocal best hit (CRBHit) pair selfblast
result for cds1 sequences (see <code><a href="cds2rbh.html">cds2rbh</a></code>) [optional]</p></dd>


<dt>selfblast2</dt>
<dd><p>(conditional-)reciprocal best hit (CRBHit) pair selfblast
result for cds2 sequences (see <code><a href="cds2rbh.html">cds2rbh</a></code>) [optional]</p></dd>


<dt>gene.position.cds1</dt>
<dd><p>specify gene position for cds1 sequences
(see <code><a href="cds2genepos.html">cds2genepos</a></code>) [default: NULL]</p></dd>


<dt>gene.position.cds2</dt>
<dd><p>specify gene position for cds2 sequences
(see <code><a href="cds2genepos.html">cds2genepos</a></code>) [default: NULL]</p></dd>


<dt>dagchainerpath</dt>
<dd><p>specify the PATH to the DAGchainer binaries
[default: /extdata/dagchainer/]</p></dd>


<dt>gap_open_penalty</dt>
<dd><p>gap open penalty [default: 0]</p></dd>


<dt>gap_extension_penalty</dt>
<dd><p>gap extension penalty [default: -3]</p></dd>


<dt>gap_length</dt>
<dd><p>length of a gap (avgerage distance expected between two
syntenic genes); if type is set to "idx" use 1 [default: 10000]</p></dd>


<dt>max_match_score</dt>
<dd><p>Maximum match score [default: 50]</p></dd>


<dt>max_dist_allowed</dt>
<dd><p>maximum distance allowed between two matches;
if type is set to "idx" use 20 [default: 200000]</p></dd>


<dt>max_evalue</dt>
<dd><p>Maximum E-value [default: 1e-3]</p></dd>


<dt>ignore_tandem</dt>
<dd><p>ignore tandem duplicates [default = TRUE]</p></dd>


<dt>only_tandem</dt>
<dd><p>only tandem alignments [default = FALSE]</p></dd>


<dt>min_number_aligned_pairs</dt>
<dd><p>Minimum number of Aligned Pairs [default: 5]</p></dd>


<dt>type</dt>
<dd><p>specify if gene order index "idx" or gene base pair position
"bp" should be extracted and used with DAGchainer [default: bp]</p></dd>


<dt>plotDotPlot</dt>
<dd><p>specify if dotplot should be plotted [default: FALSE]</p></dd>


<dt>DotPlotTitle</dt>
<dd><p>specify DotPlot title [default: 'DAGchainer results']</p></dd>


<dt>colorBy</dt>
<dd><p>specify if dagchainer groups should be colored by "Ka", "Ks",
"Ka/Ks" or "none" [default: none]</p></dd>


<dt>kaks</dt>
<dd><p>specify Ka/Ks input obtained via `rbh2kaks()` [default: NULL]</p></dd>


<dt>ka.max</dt>
<dd><p>specify max Ka to be filtered [default: 5]</p></dd>


<dt>ks.max</dt>
<dd><p>specify max Ks to be filtered [default: 5]</p></dd>


<dt>ka.min</dt>
<dd><p>specify min Ka to be filtered [default: 0]</p></dd>


<dt>ks.min</dt>
<dd><p>specify min Ks to be filtered [default: 0]</p></dd>


<dt>select.chr</dt>
<dd><p>filter results for chromosome names [default: NULL]</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>DAGchanier</code> results<br></p>


<p>1: $gene1.chr<br></p>


<p>2: $gene1.seq.id<br></p>


<p>3: $gene1.start<br></p>


<p>4: $gene1.end<br></p>


<p>5: $gene1.mid<br></p>


<p>6: $gene1.idx<br></p>


<p>7: $gene2.chr<br></p>


<p>8: $gene2.seq.id<br></p>


<p>9: $gene2.start<br></p>


<p>10: $gene2.end<br></p>


<p>11: $gene2.mid<br></p>


<p>12: $gene2.idx<br></p>


<p>13: $evalue<br></p>


<p>14: $score<br></p>


    </div>
    <div id="references">
    <h2>References</h2>
    <p>Haas BJ et al. (2004) DAGchainer: a tool for mining segmental
genome duplications and synteny. <em>Bioinformatics.</em>
<b>20(18)</b>, 3643-3646.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="plot_dagchainer.html">plot_dagchainer</a></code>,
<code><a href="cds2genepos.html">cds2genepos</a></code>,
<code><a href="tandemdups.html">tandemdups</a></code>,
<code><a href="rbh2kaks.html">rbh2kaks</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Kristian K Ullrich</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## compile dagchainer</span></span></span>
<span class="r-in"><span><span class="fu">CRBHits</span><span class="fu">::</span><span class="fu"><a href="make_dagchainer.html">make_dagchainer</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## load example sequence data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ath"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## get selfhits CRBHit pairs</span></span></span>
<span class="r-in"><span><span class="va">ath_selfhits_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    cds1<span class="op">=</span><span class="va">ath</span>,</span></span>
<span class="r-in"><span>    cds2<span class="op">=</span><span class="va">ath</span>,</span></span>
<span class="r-in"><span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="rbh2dagchainer-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">## get gene position</span></span></span>
<span class="r-in"><span><span class="va">ath.genepos</span> <span class="op">&lt;-</span> <span class="fu"><a href="cds2genepos.html">cds2genepos</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    cds<span class="op">=</span><span class="va">ath</span>,</span></span>
<span class="r-in"><span>    source<span class="op">=</span><span class="st">"ENSEMBL"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## get DAGchainer results</span></span></span>
<span class="r-in"><span><span class="va">ath_selfblast_crbh.dagchainer</span> <span class="op">&lt;-</span> <span class="fu">rbh2dagchainer</span><span class="op">(</span></span></span>
<span class="r-in"><span>    rbhpairs<span class="op">=</span><span class="va">ath_selfhits_crbh</span>,</span></span>
<span class="r-in"><span>    gene.position.cds1<span class="op">=</span><span class="va">ath.genepos</span>,</span></span>
<span class="r-in"><span>    gene.position.cds2<span class="op">=</span><span class="va">ath.genepos</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/private/var/folders/1_/x30smsdj6mlgnvfrgvx8s4nh0000gp/T/RtmpvGbRX2/temp_libpath2fb5202e5f6f/CRBHits/extdata/dagchainer/run_DAG_chainer.pl -i /var/folders/1_/x30smsdj6mlgnvfrgvx8s4nh0000gp/T//RtmpYdpQM1/file464817fe3cd0 -o 0 -e -3 -g 10000 -M 50 -D 200000 -E 0.001 -A 5 -I -s"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_selfblast_crbh.dagchainer</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   gene1.chr gene1.seq.id gene1.start gene1.end gene1.mid gene1.idx gene2.chr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     AA1:1  AT1G05100.1     1469541   1470881   1470211       437     AA1:1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     AA1:1  AT1G05400.1     1584073   1585421   1584747       468     AA1:1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     AA1:1  AT1G05570.1     1647575   1659611   1653593       484     AA1:1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     AA1:1  AT1G06100.1     1851472   1853019   1852246       546     AA1:1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     AA1:1  AT1G06120.1     1855910   1857718   1856814       548     AA1:1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6     AA1:1  AT1G06290.1     1922208   1926278   1924243       570     AA1:1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   gene2.seq.id gene2.start gene2.end gene2.mid gene2.idx   evalue score</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  AT1G07150.2     2193941   2195798   2194870       654  1.1e-55    50</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  AT1G06810.1     2091738   2092539   2092138       620  8.2e-67    67</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3  AT1G06490.2     1978378   1989409   1983894       588 1.0e-250    87</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  AT1G06360.2     1938862   1940496   1939679       576 2.0e-200    80</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  AT1G06350.1     1935227   1936841   1936034       575 7.2e-177   130</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  AT1G06310.1     1926652   1930397   1928524       571 1.0e-250   162</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    dagchainer_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 AA1:1:AA1:1:1.rev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 AA1:1:AA1:1:1.rev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 AA1:1:AA1:1:1.rev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 AA1:1:AA1:1:1.rev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 AA1:1:AA1:1:1.rev</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 AA1:1:AA1:1:1.rev</span>
<span class="r-in"><span><span class="co">## plot dagchainer</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_dagchainer.html">plot_dagchainer</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    dag<span class="op">=</span><span class="va">ath_selfblast_crbh.dagchainer</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scale for <span style="color: #00BB00;">colour</span> is already present.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span>
<span class="r-plt img"><img src="rbh2dagchainer-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

