<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>rbh2dagchainer — rbh2dagchainer • CRBHits</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="rbh2dagchainer — rbh2dagchainer" />
<meta property="og:description" content="This function runs DAGchainer (http://dagchainer.sourceforge.net/) given CRBHit pairs and gene positions for both cds1 and cds2. The default options are set to not compare gene positions in base pairs but instead using gene order (gene.idx)." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/V02KaKsVignette.html">KaKs Vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>rbh2dagchainer</h1>
    <small class="dont-index">Source: <a href='https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/masterR/rbh2dagchainer.R'><code>R/rbh2dagchainer.R</code></a></small>
    <div class="hidden name"><code>rbh2dagchainer.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function runs DAGchainer (http://dagchainer.sourceforge.net/) given CRBHit pairs and gene positions for both cds1 and cds2. The default options are set to not compare gene positions in base pairs but instead using gene order (gene.idx).</p>
    </div>

    <pre class="usage"><span class='fu'>rbh2dagchainer</span><span class='op'>(</span>
  <span class='va'>rbhpairs</span>,
  gene.position.cds1 <span class='op'>=</span> <span class='cn'>NULL</span>,
  gene.position.cds2 <span class='op'>=</span> <span class='cn'>NULL</span>,
  dagchainerpath <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/find.package.html'>find.package</a></span><span class='op'>(</span><span class='st'>"CRBHits"</span><span class='op'>)</span>, <span class='st'>"/extdata/dagchainer/"</span><span class='op'>)</span>,
  gap_open_penalty <span class='op'>=</span> <span class='fl'>0</span>,
  gap_extension_penalty <span class='op'>=</span> <span class='op'>-</span><span class='fl'>3</span>,
  gap_length <span class='op'>=</span> <span class='fl'>10000</span>,
  max_match_score <span class='op'>=</span> <span class='fl'>50</span>,
  max_dist_allowed <span class='op'>=</span> <span class='fl'>2e+05</span>,
  max_evalue <span class='op'>=</span> <span class='fl'>0.001</span>,
  ignore_tandem <span class='op'>=</span> <span class='cn'>TRUE</span>,
  min_number_aligned_pairs <span class='op'>=</span> <span class='fl'>5</span>,
  type <span class='op'>=</span> <span class='st'>"bp"</span>,
  plotDotPlot <span class='op'>=</span> <span class='cn'>FALSE</span>,
  DotPlotTitle <span class='op'>=</span> <span class='st'>"DAGchainer results"</span>,
  colorBy <span class='op'>=</span> <span class='st'>"none"</span>,
  kaks <span class='op'>=</span> <span class='cn'>NULL</span>,
  ka.max <span class='op'>=</span> <span class='fl'>5</span>,
  ks.max <span class='op'>=</span> <span class='fl'>5</span>,
  ka.min <span class='op'>=</span> <span class='fl'>0</span>,
  ks.min <span class='op'>=</span> <span class='fl'>0</span>,
  select.chr <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>rbhpairs</th>
      <td><p>(conditional-)reciprocal best hit (CRBHit) pair result (see <code><a href='cds2rbh.html'>cds2rbh</a></code>) [mandatory]</p></td>
    </tr>
    <tr>
      <th>gene.position.cds1</th>
      <td><p>specify gene position for cds1 sequences (see <code><a href='cds2genepos.html'>cds2genepos</a></code>) [default: NULL]</p></td>
    </tr>
    <tr>
      <th>gene.position.cds2</th>
      <td><p>specify gene position for cds2 sequences (see <code><a href='cds2genepos.html'>cds2genepos</a></code>) [default: NULL]</p></td>
    </tr>
    <tr>
      <th>dagchainerpath</th>
      <td><p>specify the PATH to the DAGchainer binaries [default: /extdata/dagchainer/]</p></td>
    </tr>
    <tr>
      <th>gap_open_penalty</th>
      <td><p>gap open penalty [default: 0]</p></td>
    </tr>
    <tr>
      <th>gap_extension_penalty</th>
      <td><p>gap extension penalty [default: -3]</p></td>
    </tr>
    <tr>
      <th>gap_length</th>
      <td><p>length of a gap (avgerage distance expected between two syntenic genes); if type is set to "idx" use 1 [default: 10000]</p></td>
    </tr>
    <tr>
      <th>max_match_score</th>
      <td><p>Maximum match score [default: 50]</p></td>
    </tr>
    <tr>
      <th>max_dist_allowed</th>
      <td><p>maximum distance allowed between two matches; if type is set to "idx" use 20 [default: 200000]</p></td>
    </tr>
    <tr>
      <th>max_evalue</th>
      <td><p>Maximum E-value [default: 1e-3]</p></td>
    </tr>
    <tr>
      <th>ignore_tandem</th>
      <td><p>ignore tandem duplicates [default = TRUE]</p></td>
    </tr>
    <tr>
      <th>min_number_aligned_pairs</th>
      <td><p>Minimum number of Aligned Pairs [default: 5]</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>specify if gene order index "idx" or gene base pair position "bp" should be extracted and used with DAGchainer [default: bp]</p></td>
    </tr>
    <tr>
      <th>plotDotPlot</th>
      <td><p>specify if dotplot should be plotted [default: FALSE]</p></td>
    </tr>
    <tr>
      <th>DotPlotTitle</th>
      <td><p>specify DotPlot title [default: 'DAGchainer results']</p></td>
    </tr>
    <tr>
      <th>colorBy</th>
      <td><p>specify if dagchainer groups should be colored by "Ka", "Ks", "Ka/Ks" or "none" [default: none]</p></td>
    </tr>
    <tr>
      <th>kaks</th>
      <td><p>specify Ka/Ks input obtained via `rbh2kaks()` [default: NULL]</p></td>
    </tr>
    <tr>
      <th>ka.max</th>
      <td><p>specify max Ka to be filtered [default: 5]</p></td>
    </tr>
    <tr>
      <th>ks.max</th>
      <td><p>specify max Ks to be filtered [default: 5]</p></td>
    </tr>
    <tr>
      <th>ka.min</th>
      <td><p>specify min Ka to be filtered [default: 0]</p></td>
    </tr>
    <tr>
      <th>ks.min</th>
      <td><p>specify min Ks to be filtered [default: 0]</p></td>
    </tr>
    <tr>
      <th>select.chr</th>
      <td><p>filter results for chromosome names [default: NULL]</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>DAGchanier</code> results<br />
1: $gene1.chr<br />
2: $gene1.seq.id<br />
3: $gene1.start<br />
4: $gene1.end<br />
5: $gene1.mid<br />
6: $gene1.idx<br />
7: $gene2.chr<br />
8: $gene2.seq.id<br />
9: $gene2.start<br />
10: $gene2.end<br />
11: $gene2.mid<br />
12: $gene2.idx<br />
13: $evalue<br />
14: $score<br /></p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Haas BJ et al. (2004) DAGchainer: a tool for mining segmental genome duplications and synteny. <em>Bioinformatics.</em> <b>20(18)</b>, 3643-3646.
## load example sequence data
data("ath", package="CRBHits")
## get selfhits CRBHit pairs
ath_selfhits_crbh &lt;- cds2rbh(ath, ath, plotCurve = TRUE)
## get gene position
ath.genepos &lt;- cds2genepos(ath, "ENSEMBL")
## get DAGchainer results
ath_selfblast_crbh.dagchainer &lt;- rbh2dagchainer(ath_selfhits_crbh,
                                                ath.genepos, ath.genepos)
head(ath_selfblast_crbh.dagchainer)
## plot dagchainer
plot_dagchainer(ath_selfblast_crbh.dagchainer)</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='plot_dagchainer.html'>plot_dagchainer</a></code>,
<code><a href='cds2genepos.html'>cds2genepos</a></code>,
<code><a href='tandemdups.html'>tandemdups</a></code>,
<code><a href='rbh2kaks.html'>rbh2kaks</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Kristian K Ullrich</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


