<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>dNdS Clustering Vignette • CRBHits</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="dNdS Clustering Vignette">
<meta property="og:description" content="CRBHits">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/WGDVignette.html">Paranome-based Whole Genome Duplication Vignette</a>
    </li>
    <li>
      <a href="../articles/dNdSClusteringVignette.html">dNdS Clustering Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>dNdS Clustering Vignette</h1>
                        <h4 class="author">Kristian K Ullrich</h4>
            
            <h4 class="date">2020-05-15</h4>
      
      <small class="dont-index">Source: <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/mastervignettes/dNdSClusteringVignette.Rmd"><code>vignettes/dNdSClusteringVignette.Rmd</code></a></small>
      <div class="hidden name"><code>dNdSClusteringVignette.Rmd</code></div>

    </div>

    
    
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#quickinstall">Quick Installation</a></li>
<li><a href="#crbh">Conditional Reciprocal Best Hits</a></li>
<li><a href="#dNdS">dN/dS Calculations</a></li>
<li><a href="#references">References</a></li>
</ol>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is a reimplementation of the Conditional Reciprocal Best Hit algorithm <a href="https://github.com/cboursnell/crb-blast">crb-blast</a> in R.</p>
</div>
<div id="quick-installation" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-installation" class="anchor"></a>1. <span id="quickinstall">Quick Installation</span>
</h2>
<p>see here for the R package pages and a detailed description of the install process <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/">https://mpievolbio-it.pages.gwdg.de/crbhits/</a></p>
<pre><code><a href="https://devtools.r-lib.org//reference">#see a detailed description of installation prerequisites at
#https://mpievolbio-it.pages.gwdg.de/crbhits/
library(devtools)
#from gitlab
install_gitlab("mpievolbio-it/crbhits", host = "https://gitlab.gwdg.de",
               build_vignettes = TRUE, dependencies = FALSE)
#from github
install_github("kullrich/CRBHits", build_vignettes = TRUE, dependencies = FALSE)</a></code></pre>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> needs <a href="http://last.cbrc.jp/">LAST</a> and <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> to be installed before one can eficiently use it.</p>
<p>Both prerequisites (LAST, KaKs_Calculator2.0) are forked within <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> and can be compiled for Linux/Unix/macOS wiht the functions <code><a href="../reference/make.last.html">make.last()</a></code> and <code><a href="../reference/make.KaKs_Calculator2.html">make.KaKs_Calculator2()</a></code>.</p>
<pre><code><a href="https://rdrr.io/pkg/CRBHits/man">library(CRBHits)
#compile last-1060
make.last()
#compile KaKs_Calculator2.0
make.KaKs_Calculator2()</a></code></pre>
<ol style="list-style-type: decimal">
<li><a href="#crbhpairs">Conditional reciprocal best hit pairs</a></li>
</ol>
<p>To calculate conditional reciprocal best hit pairs between two species, one can directly use an URL to access the coding sequences and calculate the conditional reciprocal best hit pair matrix (crbh).</p>
<p>As an example, here the coding sequences from <em>Arabidopsis thaliana</em> and <em>Arabidopsis lyrata</em> are used as input sequences from the FTP server from <a href="https://plants.ensembl.org/index.html">EnsemblPlants</a>.</p>
<p>The <a href="ftp://ftp.ensemblgenomes.org/pub/plants/release-47/fasta/">release-47</a> is used.</p>
<pre><code><a href="https://rdrr.io/pkg/CRBHits/man">library(CRBHits)
ensemblPlants &lt;- "ftp://ftp.ensemblgenomes.org/pub/plants/release-47/fasta/"
athfile &lt;- "arabidopsis_thaliana/cds/Arabidopsis_thaliana.TAIR10.cds.all.fa.gz"
alyfile &lt;- "arabidopsis_lyrata/cds/Arabidopsis_lyrata.v.1.0.cds.all.fa.gz"</a></code></pre>
<p>The blast-like software <a href="http://last.cbrc.jp/">LAST</a> is used to compare the translated coding sequences against each other and output a blast-like output table including the query and target length.</p>
<p>Taking the length of the obtained pairwise protein alignment one can calculate for each hit pair the query coverage as <span class="math inline">\(\frac{alignment length}{query length}\)</span>.</p>
<p>The data will be filtered with a query coverage of 50% (<code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh(..., qcov = 0.5)</a></code>).</p>
<p>Next to query coverage, the hit pairs will be additionally filtered for the twilight zone of protein sequence alignments according to <a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a> (<code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh(..., rost1999 = TRUE)</a></code>).</p>
<p>The implemented filter uses <code>equation2</code> of <a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a></p>
<p><span class="math display">\[f(x_{\text{hit pair}}) = \cases {100 \text{ , for } L_{\text{hit pair}} &lt; 11 \\ 480 * L^{-0.32 * (1 + e^{\frac{-L}{1000}})} \text{ , for } L_{\text{hit pair}} &lt;= 450 \\ 19.5 \text{ , for } L_{\text{hit pair}} &gt; 450}\]</span> , where <span class="math inline">\(x_{\text{hit pair}}\)</span> is the expected protein identity given the alignemnet length <span class="math inline">\(L_{\text{hit pair}}\)</span>. If the actual <span class="math inline">\(pident_{\text{hit pair}} &gt;= f(x_{\text{hit pair}})\)</span> the hit pair is retained.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">#plot expected pident by alignment length using eq2 from Rost (1999)</span>
<span class="no">get_pident_by_length</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>){
  <span class="no">eq2</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">L</span>){
    <span class="kw">if</span>(<span class="no">L</span> <span class="kw">&lt;=</span> <span class="fl">11</span>){<span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fl">100</span>)}
    <span class="kw">if</span>(<span class="no">L</span> <span class="kw">&lt;=</span> <span class="fl">450</span>){<span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fl">480</span>*(<span class="no">L</span>^(-<span class="fl">0.32</span>*(<span class="fl">1</span>+(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(-<span class="no">L</span>/<span class="fl">1000</span>))))))}
    <span class="kw">if</span>(<span class="no">L</span> <span class="kw">&gt;</span> <span class="fl">450</span>){<span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fl">19.5</span>)}
  }
  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">x</span>, <span class="no">eq2</span>)))
}
<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span>(<span class="no">get_pident_by_length</span>, <span class="fl">11</span>, <span class="fl">500</span>, <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">"alignment length"</span>, <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"pident"</span>,
      <span class="kw">main</span> <span class="kw">=</span> <span class="st">"expected protein identity (eq2; Rost B. 1999)"</span>)</pre></body></html></div>
<p><img src="dNdSClusteringVignette_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<p>The conditional reciprocal best hit pairs can be generated with the function <code>cdsfile2rbh</code> using multiple threads.</p>
<pre><code>ath_aly_crbh &lt;- cdsfile2rbh(paste0(ensemblPlants,athfile), paste0(ensemblPlants,alyfile),
                             qcov = 0.5, rost1999 = TRUE, threads = 8, plotCurve = TRUE)</code></pre>
<ol start="2" style="list-style-type: decimal">
<li><a href="#dNdS">ka/ks - dN/dS calculations</a></li>
</ol>
<p>model = “Li”</p>
<pre><code>cds1 &lt;- Biostrings::readDNAStringSet(paste0(ensemblPlants,athfile))
cds2 &lt;- Biostrings::readDNAStringSet(paste0(ensemblPlants,alyfile))
ath_aly_crbh.kaks.Li &lt;- rbh2kaks(ath_aly_crbh$crbh.pairs, cds1, cds2, model = "Li",
                                 threads = 8)</code></pre>
<p>In some cases the dNdS calculation fails, since e.g. there are no synonymous or nonsynonymous substitutions between cds1 and cds2 or one wants to get rid of high dS values due to substitution saturation. One can easily remove these cases before doing further analysis.</p>
<pre><code><a href="https://rdrr.io/pkg/magrittr/man">library(magrittr)
#remove all values that are negative for either ka or ks
ath_aly_crbh.kaks.Li.red &lt;- ath_aly_crbh.kaks.Li %&gt;%
  subset(ka != 9.999999) %&gt;% subset(ks != 9.999999)</a></code></pre>
<p>model = “YN”</p>
<pre><code>ath_aly_crbh.kaks.YN &lt;- rbh2kaks(ath_aly_crbh$crbh.pairs, cds1, cds2, model = "YN",
                                 threads = 8)</code></pre>
<ol start="3" style="list-style-type: decimal">
<li><a href="#crbhmatrix">Conditional reciprocal best hit matrix</a></li>
</ol>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>Aubry S., Kelly S., Kümpers B. M., Smith-Unna R. D., and Hibberd J. M. (2014). <strong>Deep evolutionary comparison of gene expression identifies parallel recruitment of trans-factors in two independent origins of C4 photosynthesis.</strong> <em>PLoS genetics</em>, <strong>10(6)</strong>. <a href="https://doi.org/10.1371/journal.pgen.1004365">https://doi.org/10.1371/journal.pgen.1004365</a></p>
<p>Charif D., and Lobry J. R. (2007). <strong>SeqinR 1.0-2: a contributed package to the R project for statistical computing devoted to biological sequences retrieval and analysis.</strong> <em>In Structural approaches to sequence evolution</em> (pp. 207-232). Springer, Berlin, Heidelberg. <a href="https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10">https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10</a></p>
<p>Kiełbasa S. M., Wan R., Sato K., Horton P., and Frith M. C. (2011). <strong>Adaptive seeds tame genomic sequence comparison.</strong> <em>Genome research</em>, 21(3), 487-493. <a href="https://doi.org/10.1101/gr.113985.110">https://doi.org/10.1101/gr.113985.110</a></p>
<p>Li W. H. (1993). <strong>Unbiased estimation of the rates of synonymous and nonsynonymous substitution.</strong> <em>Journal of molecular evolution</em>, 36(1), 96-99. <a href="https://doi.org/10.1007/bf02407308">https://doi.org/10.1007/bf02407308</a></p>
<p>Pagès H., Aboyoun P., Gentleman R., and DebRoy S. (2017). <strong>Biostrings: Efficient manipulation of biological strings.</strong> <em>R package version</em>, 2(0).</p>
<p>Rost B. (1999). <strong>Twilight zone of protein sequence alignments.</strong> <em>Protein engineering</em>, 12(2), 85-94. <a href="https://doi.org/10.1093/protein/12.2.85">https://doi.org/10.1093/protein/12.2.85</a></p>
<p>Scott C. (2017). <strong>shmlast: an improved implementation of conditional reciprocal best hits with LAST and Python.</strong> <em>Journal of Open Source Software</em>, 2(9), 142. <a href="https://joss.theoj.org/papers/10.21105/joss.00142">https://joss.theoj.org/papers/10.21105/joss.00142</a></p>
<p>Scrucca L., Fop M., Murphy T. B., and Raftery A. E. (2016) <strong>mclust 5: clustering, classification and density estimation using Gaussian finite mixture models.</strong> <em>The R Journal</em>, 8(1), 289-317. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/</a></p>
<p>Duong T., and Wand M. (2015). <strong>feature: Local Inferential Feature Significance for Multivariate Kernel Density Estimation.</strong> <em>R package version 1.2.13</em>. <a href="https://cran.r-project.org/web/packages/feature/">https://cran.r-project.org/web/packages/feature/</a></p>
<p>Yang Z., and Nielsen R. (2000). <strong>Estimating synonymous and nonsynonymous substitution rates under realistic evolutionary models.</strong> <em>Molecular biology and evolution</em>, 17(1), 32-43. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">https://doi.org/10.1093/oxfordjournals.molbev.a026236</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
