<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CRBHits Basic Vignette â€¢ CRBHits</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CRBHits Basic Vignette">
<meta property="og:description" content="CRBHits">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/V02KaKsVignette.html">KaKs Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="V01CRBHitsBasicVignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CRBHits Basic Vignette</h1>
                        <h4 class="author">Kristian K Ullrich</h4>
            
            <h4 class="date">2021-05-01</h4>
      
      <small class="dont-index">Source: <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/mastervignettes/V01CRBHitsBasicVignette.Rmd"><code>vignettes/V01CRBHitsBasicVignette.Rmd</code></a></small>
      <div class="hidden name"><code>V01CRBHitsBasicVignette.Rmd</code></div>

    </div>

    
    
<div id="crbhit-basic-vignette" class="section level1">
<h1 class="hasAnchor">
<a href="#crbhit-basic-vignette" class="anchor"></a>CRBHit Basic Vignette</h1>
<ul>
<li>includes CRBHit pair calculation</li>
<li>includes CRBHit pair filtering</li>
<li>includes Longest Isoform selection</li>
<li>includes Codon alignments</li>
<li>includes Ka/Ks calculations</li>
</ul>
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#install">Installation</a></li>
<li>
<a href="#crbhalgorithm">Conditional Reciprocal Best Hits - Algorithm</a>
<ol style="list-style-type: decimal">
<li>
<a href="#crbhstep1">1. step: sequence similarity search (last)</a>
<ol style="list-style-type: decimal">
<li><a href="#cdsinput">Input: Coding Sequences (CDS)</a></li>
<li><a href="#isoform">Get Longest Isoform from NCBI or ENSEMBL Input (optional)</a></li>
<li><a href="#sequencesearch">Sequence Similarity Search</a></li>
</ol>
</li>
<li>
<a href="#crbhstep2">2. step: filter hits + RBH extraction</a>
<ol style="list-style-type: decimal">
<li><a href="#filters">Filter blast-like output prior fitting</a></li>
<li><a href="#customfilters">Custom Filter</a></li>
<li><a href="#rbhextraction">Extract RBHs</a></li>
</ol>
</li>
<li>
<a href="#crbhstep3">3. step: fitting RBHs + get CRBHs</a>
<ol style="list-style-type: decimal">
<li><a href="#fittingparam">CRBHit pairs - Fitting Parameter</a></li>
</ol>
</li>
</ol>
</li>
<li>
<a href="#KaKs">Ka/Ks Calculations</a>
<ol style="list-style-type: decimal">
<li><a href="#cds2codon">Codon Alignments - cds2codonaln() Function</a></li>
<li><a href="#cds2kaks">Calculate Ka/Ks values comparing two sequences - cds2kaks() Function</a></li>
<li><a href="#rbh2kaks">Calculate Ka/Ks values for each CRBHit pair - rbh2kaks() Function</a></li>
</ol>
</li>
<li><a href="#references">References</a></li>
</ol>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is a reimplementation of the Conditional Reciprocal Best Hit (CRBH) algorithm <a href="https://github.com/cboursnell/crb-blast">crb-blast</a> in R.</p>
<p>The Reciprocal Best Hit (RBH) approach is commonly used in bioinformatics to show that two sequences evolved from a common ancestral gene. In other words, RBH tries to find orthologous protein sequences within and between species. These orthologous sequences can be further analysed to evaluate protein family evolution, infer phylogenetic trees and to annotate protein function.</p>
<p>The CRBH algorithm builds upon the traditional RBH approach to find additional orthologous sequences between two sets of sequences.</p>
<p>CRBH uses the sequence search results to fit an expect-value (e-value) cutoff given each RBH to subsequently add sequence pairs to the list of <em>bona-fide</em> orthologs given their alignment length. The sequence similarity results can be filtered by common criteria like e-value, query-coverage and more. If chromosome annotations are available for the protein-coding genes (protein order), this information can be directly used to assign tandem duplicates before conducting further processing steps.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> features downstream anaylsis like batch calculating synonymous (Ks) and nonsynonymous substitutions (Ka) per orthologous sequence pair. The codon sequence alignment step consist of three subtasks, namely coding nucleotide to protein sequence translation, pairwise protein sequence alignment calculation and converting the protein sequence alignment back into a codon based alignment.</p>
<p><strong>Note:</strong></p>
<p>In this Vignette the <code>lastpath</code> is defined as <code>vignette.paths[1]</code>, the <code>kakscalcpath</code> as <code>vignette.paths[2]</code> and the <code>dagchainerpath</code> as <code>vignette.paths[3]</code> to be able to build the Vignette.</p>
<p>However, once you have compiled last-1219, KaKs_Calculator2.0 and DAGchainer with the functions <code><a href="../reference/make_last.html">make_last()</a></code>, <code><a href="../reference/make_KaKs_Calculator2.html">make_KaKs_Calculator2()</a></code> and <code><a href="../reference/make_dagchainer.html">make_dagchainer()</a></code> you wonâ€™t need to specify the paths anymore. Please remove them if you would like to repeat the examples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## load vignette specific libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Biostrings</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.cran.dev/curl">curl</a></span><span class="op">)</span><span class="op">)</span>
<span class="co">## compile LAST, KaKs_Calculator2.0 and DAGchainer for the vignette</span>
<span class="va">vignette.paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_vignette.html">make_vignette</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>1. <span id="install">Installation</span>
</h2>
<p>See the R package page for a detailed description of the install process and its dependencies <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/">https://mpievolbio-it.pages.gwdg.de/crbhits/</a>.</p>
<pre><code><a href="https://devtools.r-lib.org/">## see a detailed description of installation prerequisites at
## https://mpievolbio-it.pages.gwdg.de/crbhits/
library(devtools)
## from gitlab
install_gitlab("mpievolbio-it/crbhits", host = "https://gitlab.gwdg.de",
               build_vignettes = TRUE, dependencies = FALSE)
## from github
install_github("kullrich/CRBHits", build_vignettes = TRUE, dependencies = FALSE)</a></code></pre>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> needs <a href="https://gitlab.com/mcfrith/last">LAST</a>, <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> and <a href="http://dagchainer.sourceforge.net/">DAGchainer</a> to be installed before one can efficiently use it.</p>
<p>All prerequisites (LAST, KaKs_Calculator2.0, DAGchainer) are forked within <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> and can be compiled for Linux/Unix/macOS with the functions <code><a href="../reference/make_last.html">make_last()</a></code>, <code><a href="../reference/make_KaKs_Calculator2.html">make_KaKs_Calculator2()</a></code> and <code><a href="../reference/make_dagchainer.html">make_dagchainer()</a></code>.</p>
<p>For the windows platform the user needs to use e.g.Â <a href="https://www.cygwin.com/">https://www.cygwin.com/</a> to be able to compile the prerequisites.</p>
<pre><code><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">library(CRBHits)
## compile last-1219
make_last()
## compile KaKs_Calculator2.0
make_KaKs_Calculator2()
## compile DAGchainer
make_dagchainer()</a></code></pre>
</div>
<div id="conditional-reciprocal-best-hits---algorithm" class="section level2">
<h2 class="hasAnchor">
<a href="#conditional-reciprocal-best-hits---algorithm" class="anchor"></a>2. <span id="crbhalgorithm">Conditional Reciprocal Best Hits - Algorithm</span>
</h2>
<div class="figure">
<img src="cds2rbhoverview.png" width="400" alt=""><p class="caption">Figure: Overview of the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function</p>
</div>
<p>The CRBH algorithm was introduced by <a href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1004365">Aubry S, Kelly S et al.Â (2014)</a> and ported to python <a href="https://pypi.org/project/shmlast/">shmlast</a> (<a href="https://joss.theoj.org/papers/10.21105/joss.00142">Scott C. 2017</a>) which benefits from the blast-like sequence search software <a href="http://last.cbrc.jp/">LAST</a> (<a href="https://genome.cshlp.org/content/21/3/487.long">KieÅ‚basa SM et al.Â 2011</a>).</p>
<p>The CRBH algorithm implementation consists of three steps which are explained in more detail in the following subsections.</p>
<ol style="list-style-type: decimal">
<li>step: sequence similarity search (last)(#crbhstep1)</li>
<li>step: filter hits and extract Reciprocal Best Hits (RBHs)(#crbhstep2)</li>
<li>step: fitting RBHs and get Conditional Reciprocal Best Hits (CRBHs)(#crbhstep3)</li>
</ol>
<p>All of these steps are combined in either the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function (R vector input) or the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function (file input).</p>
<p>To activate/deactivate specific options use the parameters <code>@param</code> given and explained for each function (see <code>?cds2rbh()</code> and <code>?cdsfile2rbh()</code>).</p>
<div class="figure">
<img src="cds2rbh.png" width="800" alt=""><p class="caption">Figure: Individual steps of the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function</p>
</div>
<div id="step-sequence-similarity-search-last" class="section level3">
<h3 class="hasAnchor">
<a href="#step-sequence-similarity-search-last" class="anchor"></a>2.1. <span id="crbhstep1">1. step: sequence similarity search (last)</span>
</h3>
<p>All further described steps are included in the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function. The subsections will just explain in more detail what is happening with the input data.</p>
<div id="input-coding-sequences-cds" class="section level4">
<h4 class="hasAnchor">
<a href="#input-coding-sequences-cds" class="anchor"></a>2.1.1. <span id="cdsinput">Input: Coding Sequences (CDS)</span>
</h4>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> only takes coding nucleotide sequences as the query and target inputs.</p>
<p><strong>Why?</strong> Because <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is implemented to use the same input afterwards to calculate synonymous and nonsynonymous substitutions which must match the protein sequences used to find CRBHit pairs.</p>
<p>One can either use a <code>DNAStringSet</code> vector from the <code>Biostrings</code> R package with the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function or use <strong>raw</strong>,<strong>gzipped</strong> or <strong>URL</strong> coding sequences <strong>fasta</strong> files as inputs with the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> uses either the function <code><a href="../reference/cdsfile2aafile.html">cdsfile2aafile()</a></code> to translate a coding sequence <strong>fasta</strong> file into its corresponding protein sequence <strong>fasta</strong> file or <code><a href="../reference/cds2aa.html">cds2aa()</a></code> to translate a <code>DNAStringSet</code> into a <code>AAStringSet</code>. Thereby it removes all sequences that are not a multiple of three which can not be parsed correctly.</p>
<p><strong>Note:</strong> The user can easily check if all coding sequences would be a multiple of three or rely on the input files generated by other sources.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to check coding sequences if all are a mutiple of three</span>

<span class="co">## load CDS file</span>
<span class="va">cdsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span>
<span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile</span><span class="op">)</span>
<span class="co">## the following statement should return TRUE, if all sequences are a mutiple of three</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu">width</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span> <span class="op">%%</span> <span class="fl">3</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="get-longest-isoform-from-ncbi-or-ensembl-input-optional" class="section level4">
<h4 class="hasAnchor">
<a href="#get-longest-isoform-from-ncbi-or-ensembl-input-optional" class="anchor"></a>2.1.2. <span id="isoform">Get Longest Isoform from NCBI or ENSEMBL Input (optional)</span>
</h4>
<p>It is also possible to use an URL to directly access coding sequences from <a href="https://www.ncbi.nlm.nih.gov/genome">NCBI</a> or <a href="ftp://ftp.ensembl.org/pub/current_fasta">ENSEMBL</a> and reduce this input to the longest annotated isoform.</p>
<pre><code>## example how to access CDS from URL and get longest isoform

## get coding sequences for Homo sapiens from NCBI
HOMSAP.cds.NCBI.url &lt;- paste0(
  "https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/",
  "GCF_000001405.39_GRCh38.p13/",
  "GCF_000001405.39_GRCh38.p13_cds_from_genomic.fna.gz")
HOMSAP.cds.NCBI &lt;- Biostrings::readDNAStringSet(HOMSAP.cds.NCBI.url)
## reduce to the longest isoform
HOMSAP.cds.NCBI.longest &lt;- isoform2longest(HOMSAP.cds.NCBI, "NCBI")
## get coding sequences for Homo sapiens from ENSEMBL
HOMSAP.cds.ENSEMBL.url &lt;- paste0(
  "ftp://ftp.ensembl.org/pub/release-101/fasta/",
  "homo_sapiens/cds/Homo_sapiens.GRCh38.cds.all.fa.gz")
HOMSAP.cds.ENSEMBL.file &lt;- tempfile()
download.file(HOMSAP.cds.ENSEMBL.url, HOMSAP.cds.ENSEMBL.file, quiet = FALSE)
HOMSAP.cds.ENSEMBL &lt;- Biostrings::readDNAStringSet(HOMSAP.cds.ENSEMBL.file)
## reduce to the longest isoform
HOMSAP.cds.ENSEMBL.longest &lt;- isoform2longest(HOMSAP.cds.ENSEMBL, "ENSEMBL")

## get help
#?isoform2longest</code></pre>
<p><strong>Note:</strong> It is also possible to obtain and select the longest/primary isoform from CDS Input sources other than NCBI or Ensembl. This can be achieved by parsing a GTF/GFF3 file. This special case is handled in the <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V02KaKsVignette.html">CRBHits KaKs Vignette</a> under point 1.4.2.</p>
</div>
<div id="sequence-similarity-search" class="section level4">
<h4 class="hasAnchor">
<a href="#sequence-similarity-search" class="anchor"></a>2.1.3. <span id="sequencesearch">Sequence Similarity Search</span>
</h4>
<p>The blast-like software <a href="http://last.cbrc.jp/">LAST</a> is used to compare the translated coding sequences against each other and output a blast-like output table including the query and target length.</p>
<p><strong>Note:</strong> Classical RBH can be performed by disabling the <code>crbh</code> <code>@param</code>of the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> or <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to get CRBHit pairs from two CDS using classical RBH</span>

<span class="co">## define CDS file1</span>
<span class="va">cdsfile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span>
<span class="co">## define CDS file2</span>
<span class="va">cdsfile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"aly.cds.fasta.gz"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span>
<span class="co">## get CDS1</span>
<span class="va">cds1</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile1</span><span class="op">)</span>
<span class="co">## get CDS2</span>
<span class="va">cds2</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile2</span><span class="op">)</span>
<span class="co">## example how to perform classical RBH</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, crbh <span class="op">=</span> <span class="cn">FALSE</span>,
                       lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="co">## show summary</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">)</span>
<span class="co">#&gt;            Length Class      Mode</span>
<span class="co">#&gt; crbh.pairs  3     data.frame list</span>
<span class="co">#&gt; crbh1      16     data.frame list</span>
<span class="co">#&gt; crbh2      16     data.frame list</span>

<span class="co">## get help</span>
<span class="co">#?cds2rbh</span></code></pre></div>
<p>Like <a href="https://pypi.org/project/shmlast/">shmlast</a>, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> plots the fitted model of the CRBH e-value based algorithm. The CRBH algorithm can be performed and visualized as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to get CRBHit pairs using one thread and plot CRBHit algorithm fitting curve</span>

<span class="co">## example how to perform CRBH</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## show summary</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">)</span>
<span class="co">#&gt;               Length Class      Mode    </span>
<span class="co">#&gt; crbh.pairs     3     data.frame list    </span>
<span class="co">#&gt; crbh1         16     data.frame list    </span>
<span class="co">#&gt; crbh2         16     data.frame list    </span>
<span class="co">#&gt; rbh1_rbh2_fit  1     -none-     function</span>

<span class="co">## get help</span>
<span class="co">#?cds2rbh</span></code></pre></div>
<p>Both, <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> and <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function, return a list of three <code>data.frame</code>â€™s which contain the CRBH pairs (<code>$crbh.pairs</code>) retained, the query &gt; target blast-like output matrix (<code>$crbh1</code>) and the target &gt; query blast-like output matrix (<code>$crbh2</code>). For each matrix the <code>$rbh_class</code> indicates if the hit is a Reciprocal Best Hit (<code>rbh</code>) or if it is a conditional secondary hit retained because of RBH fitting (CRBH algorithm) (<code>sec</code>).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example showing cds2rbh() results</span>

<span class="co">## show dimension and first parts of retained hit pairs</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span>
<span class="co">#&gt; [1] 211   3</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span>
<span class="co">#&gt;           aa1                   aa2 rbh_class</span>
<span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh</span>
<span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh</span>
<span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh</span>
<span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh</span>
<span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh</span>
<span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh</span>

<span class="co">## show first retained hit pairs for the query &gt; target matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span><span class="op">)</span>
<span class="co">#&gt;      query_id            subject_id perc_identity alignment_length mismatches</span>
<span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256         31.09              119         73</span>
<span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128        100.00              213          0</span>
<span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125         90.75              281         19</span>
<span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114         92.88              309         19</span>
<span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419         33.33               81         51</span>
<span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326         44.94               89         41</span>
<span class="co">#&gt;   gap_opens q_start q_end s_start s_end   evalue bit_score query_length</span>
<span class="co">#&gt; 1         2     650   768     354   463  4.0e-06      47.0         1910</span>
<span class="co">#&gt; 2         0       1   213       1   213 1.5e-160     493.0          213</span>
<span class="co">#&gt; 3         3       1   275       1   280 5.9e-178     551.0          287</span>
<span class="co">#&gt; 4         1       1   306       1   309 2.9e-218     668.0          308</span>
<span class="co">#&gt; 5         2     450   529      91   169  9.6e-09      53.6          541</span>
<span class="co">#&gt; 6         1     409   497     251   331  3.4e-16      78.3          591</span>
<span class="co">#&gt;   subject_length raw_score rbh_class</span>
<span class="co">#&gt; 1            540        99       rbh</span>
<span class="co">#&gt; 2            213      1110       rbh</span>
<span class="co">#&gt; 3            299      1240       rbh</span>
<span class="co">#&gt; 4            316      1505       rbh</span>
<span class="co">#&gt; 5            189       114       rbh</span>
<span class="co">#&gt; 6            451       170       rbh</span>

<span class="co">## get the number of CRBHit classified as rbh and sec hit pairs</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; rbh sec </span>
<span class="co">#&gt; 181  30</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh2</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; rbh sec </span>
<span class="co">#&gt; 181  30</span></code></pre></div>
<p>If the <code>@param</code> <code>crbh</code> was set to <code>TRUE</code>, the CRBH algorithm was applied and the fitting function will also be returned. The fitting function can than be used for manual plotting as follows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to use the fitting function for manual plotting</span>

<span class="co">## plot fitting function</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="fu">rbh1_rbh2_fit</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,
      from <span class="op">=</span> <span class="fl">1</span>,
      to <span class="op">=</span> <span class="fl">1000</span>,
      xlab <span class="op">=</span> <span class="st">"alnlength"</span>,
      ylab <span class="op">=</span> <span class="st">"-log10(evalue)"</span>,
      main <span class="op">=</span> <span class="st">"CRBH fitting"</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>One can also specify if a hit pair which is only found in one direction should be retained, which will be classified as (<code>single</code>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to retain single direction secondary homologs</span>

<span class="co">## get CRBHit pairs with keepSingleDirection = TRUE</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                        keepSingleDirection <span class="op">=</span> <span class="cn">TRUE</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## get the number of CRBHit classified as rbh and sec hit pairs</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    rbh    sec single </span>
<span class="co">#&gt;    181     30      1</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh2</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    rbh    sec single </span>
<span class="co">#&gt;    181     30      2</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">)</span>
<span class="co">#&gt; NULL</span>

<span class="co">## get help</span>
<span class="co">#?cds2rbh</span></code></pre></div>
</div>
</div>
<div id="step-filter-hits-rbh-extraction" class="section level3">
<h3 class="hasAnchor">
<a href="#step-filter-hits-rbh-extraction" class="anchor"></a>2.2. <span id="crbhstep2">2. step: filter hits + RBH extraction</span>
</h3>
<p>All further described steps are included in the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function. The subsections will just explain in more detail what is happening with the input data.</p>
<p>Until now the hit pairs obtained via the sequence similarity searches were only filtered by default settings (e-value &lt; 0.001). The following subsections will explain the default filters in more detail and show how one can create and apply custom filter on the hit pairs prior RBH extraction and fitting to obtain CRBHit pairs.</p>
<div id="filter-blast-like-output-prior-fitting" class="section level4">
<h4 class="hasAnchor">
<a href="#filter-blast-like-output-prior-fitting" class="anchor"></a>2.2.1. <span id="filters">Filter blast-like output prior fitting</span>
</h4>
<p>The following filters are already defined and can be used out of the box:</p>
<ul>
<li>evalue (<code>@param</code> <code>evalue = 0.001</code>)</li>
<li>query coverage (<code>@param</code> <code>qcov = 0.0</code>)</li>
<li>target coverage (<code>@param</code> <code>tcov = 0.0</code>)</li>
<li>protein identity (<code>@param</code> <code>pident = 0</code>)</li>
<li>alignment length (<code>@param</code> <code>alnlen = 0</code>)</li>
<li>rost1999 (<code>@param</code> <code>rost1999 = FALSE</code>)</li>
</ul>
<p>Taking the length of the obtained pairwise protein alignment one can calculate for each hit pair the query coverage as <span class="math inline">\(\frac{alignment length}{query length}\)</span>.</p>
<p>The data will be filtered with a query coverage of 50% using the <code>@param</code> <code>qcov</code> as follows: <code><a href="../reference/cds2rbh.html">cds2rbh(..., qcov = 0.5)</a></code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to filter prior crbh for query coverage of 50%</span>

<span class="co">## get CRBHit pairs with direct query coverage filter</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>, qcov <span class="op">=</span> <span class="fl">0.5</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span>
<span class="co">#&gt; [1] 145   3</span>

<span class="co">## get help</span>
<span class="co">#?cds2rbh</span></code></pre></div>
<p>In addition, the users can filter for the <strong>twilight zone of protein sequence alignments</strong> according to <a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a>. This <strong>twilight zone</strong> indicates proteins, which might share a common ancestor (homologue genes) but have diverged from each other through time so that similarity is no longer detectable. Here, a certain alignment length and protein identity is at least necessary to indicate <em>bona-fide</em> homologues.</p>
<p>The following equation and plot will show how the relationship between alignment length and protein identity is defined for the <code>rost1999</code> filter.</p>
<p>The implemented <code>rost1999</code> filter uses <code>equation2</code> of <a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a></p>
<p><span class="math display">\[f(x_{\text{hit pair}}) = \cases {100 \text{ , for } L_{\text{hit pair}} &lt; 11 \\ 480 * L^{-0.32 * (1 + e^{\frac{-L}{1000}})} \text{ , for } L_{\text{hit pair}} \leq 450 \\ 19.5 \text{ , for } L_{\text{hit pair}} &gt; 450}\]</span> , where <span class="math inline">\(x_{\text{hit pair}}\)</span> is the expected protein identity given the alignemnet length <span class="math inline">\(L_{\text{hit pair}}\)</span>. If the actual <span class="math inline">\(pident_{\text{hit pair}} &gt;= f(x_{\text{hit pair}})\)</span> the hit pair is retained.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## detailed explanation for the Rost (1999) twilight-zone filter</span>

<span class="co">## define eq2 from Rost (1999)</span>
<span class="va">get_pident_by_length</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span>
  <span class="va">eq2</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">L</span><span class="op">)</span><span class="op">{</span>
    <span class="kw">if</span><span class="op">(</span><span class="va">L</span> <span class="op">&lt;=</span> <span class="fl">11</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">}</span>
    <span class="kw">if</span><span class="op">(</span><span class="va">L</span> <span class="op">&lt;=</span> <span class="fl">450</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fl">480</span><span class="op">*</span><span class="op">(</span><span class="va">L</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">0.32</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">L</span><span class="op">/</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">}</span>
    <span class="kw">if</span><span class="op">(</span><span class="va">L</span> <span class="op">&gt;</span> <span class="fl">450</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fl">19.5</span><span class="op">)</span><span class="op">}</span>
  <span class="op">}</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">eq2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="co">## plot expected pident by alignment length using eq2 from Rost (1999)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span><span class="op">(</span><span class="va">get_pident_by_length</span>, <span class="fl">11</span>, <span class="fl">500</span>, pch <span class="op">=</span> <span class="fl">20</span>, xlab <span class="op">=</span> <span class="st">"alignment length"</span>, ylab <span class="op">=</span> <span class="st">"pident"</span>,
      main <span class="op">=</span> <span class="st">"expected protein identity (eq2; Rost B. 1999)"</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>This <strong>twilight zone</strong> filter can be directly used by the <code>@param</code> <code>rost1999</code> set to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to filter prior crbh for eq2 from Rost (1999)</span>

<span class="co">## get CRBHit pairs with direct twilight-zone filter</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>, rost1999 <span class="op">=</span> <span class="cn">TRUE</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span>
<span class="co">#&gt; [1] 208   3</span>

<span class="co">## get help</span>
<span class="co">#?cds2rbh</span>
<span class="co">## get help</span>
<span class="co">#?filter.alnlen</span>
<span class="co">## get help</span>
<span class="co">#?filter.eval</span>
<span class="co">## get help</span>
<span class="co">#?filter.pident</span>
<span class="co">## get help</span>
<span class="co">#?filter.qcov</span>
<span class="co">## get help</span>
<span class="co">#?filter.rost1999</span>
<span class="co">## get help</span>
<span class="co">#?filter.tcov</span></code></pre></div>
</div>
<div id="custom-filter" class="section level4">
<h4 class="hasAnchor">
<a href="#custom-filter" class="anchor"></a>2.2.2. <span id="filters">Custom Filter</span>
</h4>
<p>The user can also define its own filter function and apply it to the blast-like output prior fitting and calculating CRBHs as follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example for a custom filters</span>

<span class="co">## define custom filter for e.g. bit score (column 12)</span>
<span class="va">myfilter1</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rbh</span>, <span class="va">value</span> <span class="op">=</span> <span class="fl">500.0</span><span class="op">)</span><span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">rbh</span>, <span class="va">bit_score</span> <span class="op">&gt;=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">## define custom filter for e.g. corrected query_coverage</span>
<span class="va">myfilter2</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rbh</span>, <span class="va">value</span> <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">rbh</span>, <span class="op">(</span><span class="op">(</span><span class="va">alignment_length</span><span class="op">-</span><span class="va">mismatches</span><span class="op">-</span><span class="va">gap_opens</span><span class="op">)</span> <span class="op">/</span> <span class="va">query_length</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">## get CRBHit pairs with custom filter list</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                        filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">myfilter1</span>, <span class="va">myfilter2</span><span class="op">)</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span>
<span class="co">#&gt; [1] 58  3</span></code></pre></div>
</div>
<div id="extract-rbhs" class="section level4">
<h4 class="hasAnchor">
<a href="#extract-rbhs" class="anchor"></a>2.2.3. <span id="rbhextraction">Extract RBHs</span>
</h4>
<p>As already explained in the Sequence Similarity Search section (#sequencesearch), both, <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function, return a list of three <code>data.frame</code>â€™s which contain the CRBHit pairs (<code>$crbh.pairs</code>) retained, the query &gt; target blast-like output matrix (<code>$crbh1</code>) and the target &gt; query blast-like output matrix (<code>$crbh2</code>). For each matrix the <code>$rbh_class</code> indicates if the hit is a Reciprocal Best Hit (<code>rbh</code>) or if it is a conditional secondary hit retained because of RBH fitting (CRBH algorithm) (<code>sec</code>).</p>
<p>To extract just the RBHs one can subset based on the <code>rbh_class</code> columns as follows:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example to extract CRBHit pairs classified as rbh</span>

<span class="co">## reduce to rbh_class rbh</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"ath_aly_crbh"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span>, <span class="va">rbh_class</span> <span class="op">==</span> <span class="st">"rbh"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;      query_id            subject_id perc_identity alignment_length mismatches</span>
<span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256         31.09              119         73</span>
<span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128        100.00              213          0</span>
<span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125         90.75              281         19</span>
<span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114         92.88              309         19</span>
<span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419         33.33               81         51</span>
<span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326         44.94               89         41</span>
<span class="co">#&gt;   gap_opens q_start q_end s_start s_end   evalue bit_score query_length</span>
<span class="co">#&gt; 1         2     650   768     354   463  4.0e-06      47.0         1910</span>
<span class="co">#&gt; 2         0       1   213       1   213 1.5e-160     493.0          213</span>
<span class="co">#&gt; 3         3       1   275       1   280 5.9e-178     551.0          287</span>
<span class="co">#&gt; 4         1       1   306       1   309 2.9e-218     668.0          308</span>
<span class="co">#&gt; 5         2     450   529      91   169  9.6e-09      53.6          541</span>
<span class="co">#&gt; 6         1     409   497     251   331  3.4e-16      78.3          591</span>
<span class="co">#&gt;   subject_length raw_score rbh_class</span>
<span class="co">#&gt; 1            540        99       rbh</span>
<span class="co">#&gt; 2            213      1110       rbh</span>
<span class="co">#&gt; 3            299      1240       rbh</span>
<span class="co">#&gt; 4            316      1505       rbh</span>
<span class="co">#&gt; 5            189       114       rbh</span>
<span class="co">#&gt; 6            451       170       rbh</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh2</span>, <span class="va">rbh_class</span> <span class="op">==</span> <span class="st">"rbh"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;                query_id  subject_id perc_identity alignment_length mismatches</span>
<span class="co">#&gt; 1 Al_scaffold_0001_1000 AT1G04940.1         40.48               42         25</span>
<span class="co">#&gt; 2 Al_scaffold_0001_1004 AT1G10280.1         99.27              413          3</span>
<span class="co">#&gt; 3 Al_scaffold_0001_1120 AT1G08040.1         45.67              300        151</span>
<span class="co">#&gt; 4  Al_scaffold_0001_114 AT1G01180.1         92.88              309         19</span>
<span class="co">#&gt; 5 Al_scaffold_0001_1146 AT1G09740.1         32.54              169        105</span>
<span class="co">#&gt; 6 Al_scaffold_0001_1151 AT1G03300.1         33.28              673        377</span>
<span class="co">#&gt;   gap_opens q_start q_end s_start s_end        evalue bit_score query_length</span>
<span class="co">#&gt; 1         0      81   122     166   207  5.400000e-07      46.1          186</span>
<span class="co">#&gt; 2         0       1   413       1   413 9.881313e-324     980.0          413</span>
<span class="co">#&gt; 3         8      83   379      92   382  2.300000e-80     281.0          437</span>
<span class="co">#&gt; 4         1       1   309       1   306 8.100000e-218     668.0          316</span>
<span class="co">#&gt; 5         4      34   197       4   168  1.200000e-16      78.3          274</span>
<span class="co">#&gt; 6        19       7   611       3   671  1.300000e-85     304.0          611</span>
<span class="co">#&gt;   subject_length raw_score rbh_class</span>
<span class="co">#&gt; 1            275        97       rbh</span>
<span class="co">#&gt; 2            413      2212       rbh</span>
<span class="co">#&gt; 3            383       629       rbh</span>
<span class="co">#&gt; 4            308      1505       rbh</span>
<span class="co">#&gt; 5            172       170       rbh</span>
<span class="co">#&gt; 6            671       681       rbh</span></code></pre></div>
</div>
</div>
<div id="step-fitting-rbhs-get-crbhs" class="section level3">
<h3 class="hasAnchor">
<a href="#step-fitting-rbhs-get-crbhs" class="anchor"></a>2.3. <span id="crbhstep3">3. step: fitting RBHs + get CRBHs</span>
</h3>
<p>All further described steps are included in the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function. The subsections will just explain in more detail what is happening with the input data.</p>
<div id="crbhit-pairs---fitting-parameter" class="section level4">
<h4 class="hasAnchor">
<a href="#crbhit-pairs---fitting-parameter" class="anchor"></a>2.3.1. <span id="fittingparam">CRBHit pairs - Fitting Parameter</span>
</h4>
<p>The original implementation of the CRBH algorithm <a href="https://github.com/cboursnell/crb-blast/blob/master/lib/crb-blast/crb-blast.rb">crb-blast.rb line 438-518</a> uses certain parameters for fitting the e-value against the alignment length.</p>
<p>In brief, for each retained reciprocal hit pair (RBH; after filtering) the e-value and the corresponding alignment length of a hit pair and its neighborhood is evaluated. The neighborhood is defined by a weighting parameter of the alignment length (default is set to <code>@param</code> <code>fit.varweight = 0.1</code>) and a minimum neighborhood alignment length (default is set to <code>@param</code> <code>fit.min = 5</code>).</p>
<p>For each unique alignment length <span class="math inline">\(L\)</span> a value <span class="math inline">\(s_{L}\)</span> is defined as the neighborhood alignment length as</p>
<p><span class="math display">\[s_{L} = \cases {\text{fit.min} \text{ , for } L * \text{fit.varweight} &lt; \text{fit.min} \\ L * \text{fit.varweight} \text{ , for } L * \text{fit.varweight} \geq \text{fit.min}}\]</span> E.g. for an alignment length of 100 and the default settings, <span class="math inline">\(s_{100} = 100 * 0.1 = 10\)</span>.</p>
<p>All RBHs, for which <span class="math inline">\(L-s_{L} \leq L_{RBH_i} \leq L + s_{L}\)</span>, in other words, for which their alignment lengths fall between 90 and 110 are than considered for fitting a cutoff at that alignment length using the mean of the <span class="math inline">\(-log10(evalue)\)</span>.</p>
<p><span class="math display">\[\text{cutoff}_{L} = \frac{1}{n} \sum_{i = 1}^{n}{\cases{324 \text{ , for }  evalue_{RBH_i} = 0 \\ -log10(evalue_{\text{RBH}_i}) \text{ , for } evalue_{RBH_i} \ne 0}}\]</span></p>
<p>The user can now alter between <code>mean</code> and <code>median</code> using the <code>@param</code> <code>fit.type = "mean"</code> or <code>fit.type = "median"</code> to get alternative fitting curves.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to get crbh from two coding fasta files using median fitting</span>

<span class="co">## get CRBHit pairs with median fitting</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                        fit.type <span class="op">=</span> <span class="st">"median"</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">## get help</span>
<span class="co">#?cds2rbh</span></code></pre></div>
<p>As described earlier, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> features downstream anaylsis like batch calculating synonymous (Ks) and nonsynonymous substitutions (Ka) per orthologous sequence pair. The following sections will give more details about these downstream functionalities.</p>
</div>
</div>
</div>
<div id="kaks-calculations" class="section level2">
<h2 class="hasAnchor">
<a href="#kaks-calculations" class="anchor"></a>3. <span id="KaKs">Ka/Ks Calculations</span>
</h2>
<div class="figure">
<img src="rbh2kaksoverview.png" width="800" alt=""><p class="caption">Figure: Overview of the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function steps</p>
</div>
<p>The resulting CRBHs (see #crbhalgorithm) can be further processed to e.g.Â filtered for Tandem Duplicates (see <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V02KaKsVignette.html">CRBHits KaKs Vignette</a>) and/or can be used to calculate Ka/Ks values for each CRBH pair (see #rbh2kaks and <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V02KaKsVignette.html">CRBHits KaKs Vignette</a>).</p>
<p>All further described steps are included in the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function. The subsections will just explain in more detail what is happening with the input data.</p>
<p>To activate/deactivate specific options use the parameters <code>@param</code> given and explained for each function (see <code>?rbh2kaks()</code>).</p>
<p><strong>Note:</strong> To be able to perform <em>Tandem Duplicate Assignment</em> with the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function one needs to specify <em>gene positions</em> for both Input CDS (can be extracted automatically from <em>NCBI</em> or <em>ENSEMBL</em> Input (see <code><a href="../reference/cds2genepos.html">cds2genepos()</a></code> function)) and supply <em>selfblast</em> results for both Input CDS.</p>
<div class="figure">
<img src="rbh2kaks.png" width="600" alt=""><p class="caption">Figure: Individual steps of the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function</p>
</div>
<div id="codon-alignments---cds2codonaln-function" class="section level3">
<h3 class="hasAnchor">
<a href="#codon-alignments---cds2codonaln-function" class="anchor"></a>3.1. <span id="cds2codon">Codon Alignments - cds2codonaln() Function</span>
</h3>
<p>To be able to calculate synonymous (Ks) and non-synonymous (Ka) substitutions, one needs to use a codon alignment.</p>
<p>The <code><a href="../reference/cds2codonaln.html">cds2codonaln()</a></code> function takes two single nucleotide sequences as input to obtain such a codon alignment. The function will convert the nucleotide sequences into amino acid sequences, align them with the help of the <code><a href="https://rdrr.io/pkg/Biostrings/man/pairwiseAlignment.html">pairwiseAlignment()</a></code> function from the <code>Biostrings</code> R package and convert this alignment back into a codon alignment.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example to get a codon alignment</span>

<span class="co">## define two CDS</span>
<span class="va">cds1</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html">DNAString</a></span><span class="op">(</span><span class="st">"ATGCAACATTGC"</span><span class="op">)</span>
<span class="va">cds2</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html">DNAString</a></span><span class="op">(</span><span class="st">"ATGCATTGC"</span><span class="op">)</span>
<span class="co">## get codon alignment</span>
<span class="fu"><a href="../reference/cds2codonaln.html">cds2codonaln</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span><span class="op">)</span>
<span class="co">#&gt; DNAStringSet object of length 2:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]    12 ATGCAACATTGC                                      cds1</span>
<span class="co">#&gt; [2]    12 ATG---CATTGC                                      cds2</span>

<span class="co">## get help</span>
<span class="co">#?cds2codonaln</span></code></pre></div>
<p>The user can alter some <code>@param</code> to change the alignment parameters, like <code>type</code>, <code>substitutionMatrix</code>, <code>gapOpening</code> and <code>gapExtension</code> costs as defined by the <code><a href="https://rdrr.io/pkg/Biostrings/man/pairwiseAlignment.html">pairwiseAlignment()</a></code> function from the <code>Biostrings</code> R package.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example to alter the substitionMatrix and use the BLOSUM45 cost matrix</span>
<span class="co">## for the codon alignment</span>

<span class="co">## get codon alignemnt with BLOSUM45 cost matrix</span>
<span class="fu"><a href="../reference/cds2codonaln.html">cds2codonaln</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, substitutionMatrix <span class="op">=</span> <span class="st">"BLOSUM45"</span><span class="op">)</span>
<span class="co">#&gt; DNAStringSet object of length 2:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]    12 ATGCAACATTGC                                      cds1</span>
<span class="co">#&gt; [2]    12 ATG---CATTGC                                      cds2</span></code></pre></div>
<p>The user can also remove gaps in the codon alignment.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example to remove codon gaps</span>

<span class="co">## get codon alignment with gaps removed</span>
<span class="fu"><a href="../reference/cds2codonaln.html">cds2codonaln</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, remove.gaps <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; DNAStringSet object of length 2:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]     9 ATGCATTGC                                         cds1</span>
<span class="co">#&gt; [2]     9 ATGCATTGC                                         cds2</span></code></pre></div>
</div>
<div id="cds2kaks-function" class="section level3">
<h3 class="hasAnchor">
<a href="#cds2kaks-function" class="anchor"></a>3.2. <span id="cds2kaks">cds2kaks() Function</span>
</h3>
<p>The Ka/Ks (dN/dS) values can be obtained either via the codon model of <a href="https://www.ncbi.nlm.nih.gov/pubmed/8433381">Li WH. (1999)</a> implemented in the R package <a href="https://cran.r-project.org/web/packages/seqinr/index.html">seqinr</a> or the model <a href="https://www.ncbi.nlm.nih.gov/pubmed/10666704">Yang Z and Nielson R. (2000)</a> implemented in <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> (see <a href="#quickinstall">Quick Installation</a> how to compile the prerequisites).</p>
<p>The <code><a href="../reference/cds2kaks.html">cds2kaks()</a></code> function takes two single nucleotide sequences as input to obtain a codon alignment and calculate Ka/Ks.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## calculate Ka/Ks on two CDS</span>

<span class="co">## load example sequence data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"ath"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"aly"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span>
<span class="co">## select a sequence pair according to a best hit pair (done for you)</span>
<span class="va">cds1</span> <span class="op">&lt;-</span> <span class="va">ath</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="va">cds2</span> <span class="op">&lt;-</span> <span class="va">aly</span><span class="op">[</span><span class="fl">282</span><span class="op">]</span>
<span class="co">## calculate Ka/Ks values on two CDS using Li model</span>
<span class="fu"><a href="../reference/cds2kaks.html">cds2kaks</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, model <span class="op">=</span> <span class="st">"Li"</span><span class="op">)</span>
<span class="co">#&gt;         ka         ks        vka        vks </span>
<span class="co">#&gt; 1.02338181 9.99999900 0.01140267 9.99999900</span>

<span class="co">## get help</span>
<span class="co">#?cds2kaks</span></code></pre></div>
<p>All parameters which can be used by the <code><a href="../reference/cds2codonaln.html">cds2codonaln()</a></code> function can be overgiven to e.g.Â use an alternative <code>substitutionMatrix</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example to use an alternative substitutionMatrix for the codon alignment</span>
<span class="co">## and obtain Ka/Ks</span>

<span class="co">## calculate Ka/Ks values on two CDS using Li model and BLOSUM45 cost matrix</span>
<span class="fu"><a href="../reference/cds2kaks.html">cds2kaks</a></span><span class="op">(</span><span class="va">cds1</span>, <span class="va">cds2</span>, model <span class="op">=</span> <span class="st">"Li"</span>, substitutionMatrix <span class="op">=</span> <span class="st">"BLOSUM45"</span><span class="op">)</span>
<span class="co">#&gt;          ka          ks         vka         vks </span>
<span class="co">#&gt; 0.878034285 9.999999000 0.009456807 9.999999000</span></code></pre></div>
</div>
<div id="rbh2kaks-function" class="section level3">
<h3 class="hasAnchor">
<a href="#rbh2kaks-function" class="anchor"></a>3.3. <span id="rbh2kaks">rbh2kaks() Function</span>
</h3>
<p>The resulting CRBHit pairs (see <a href="#cds2rbhfunction">cds2rbh()</a>) can be used with the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function to obtain pairwise codon alignments, which are further used to calculate synonymous (Ks) and nonsynonymous (Ka) substitutions using parallelization.</p>
<p><strong>Note:</strong></p>
<p>It is important, that the names of the <code>rbh</code> columns must match the names of the corresponding <code>cds1</code> and <code>cds2</code> <code>DNAStringSet</code> vectors.</p>
<p>However, since one should directly use the same input <code>DNAStringSet</code> vector or input <code>url</code> to calculate the reciprocal best hit pair matrix with the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> or the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function, this should not be a problem.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## example how to get CRBHit pairs from two coding fasta files</span>
<span class="va">cdsfile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span>
<span class="va">cdsfile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"aly.cds.fasta.gz"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span>

<span class="va">ath</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile1</span><span class="op">)</span>
<span class="va">aly</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile2</span><span class="op">)</span>

<span class="co">## the following function calculates CRBHit pairs using one thread and plots the fitted curve</span>
<span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span>cds1 <span class="op">=</span> <span class="va">ath</span>, cds2 <span class="op">=</span> <span class="va">aly</span>,
                        lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>

<span class="co">## calculate Ka/Ks using the CRBHit pairs</span>
<span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span>
<span class="va">ath_aly_crbh.kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span>rbhpairs <span class="op">=</span> <span class="va">ath_aly_crbh</span>,
                              cds1 <span class="op">=</span> <span class="va">ath</span>, cds2 <span class="op">=</span> <span class="va">aly</span>,
                              model <span class="op">=</span> <span class="st">"Li"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh.kaks</span><span class="op">)</span>
<span class="co">#&gt;           aa1                   aa2 rbh_class         ka        ks          vka</span>
<span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh 0.74651201 9.9999990 0.0032172312</span>
<span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh 0.00000000 0.1831839 0.0000000000</span>
<span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh 0.05680851 0.1390746 0.0001612571</span>
<span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh 0.03656808 0.1543978 0.0001014973</span>
<span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh 0.75036088 9.9999990 0.0093948533</span>
<span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh 0.78228833 9.9999990 0.0051807370</span>
<span class="co">#&gt;           vks</span>
<span class="co">#&gt; 1 9.999999000</span>
<span class="co">#&gt; 2 0.003022307</span>
<span class="co">#&gt; 3 0.001531756</span>
<span class="co">#&gt; 4 0.001249826</span>
<span class="co">#&gt; 5 9.999999000</span>
<span class="co">#&gt; 6 9.999999000</span>

<span class="co">## get help</span>
<span class="co">#?rbh2kaks</span></code></pre></div>
<p>The Ka/Ks calculations can be parallelized.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## calculate Ka/Ks using the CRBHit pairs and multiple threads</span>
<span class="va">ath_aly_crbh.kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span>rbhpairs <span class="op">=</span> <span class="va">ath_aly_crbh</span>,
                              cds1 <span class="op">=</span> <span class="va">ath</span>, cds2 <span class="op">=</span> <span class="va">aly</span>,
                              model <span class="op">=</span> <span class="st">"Li"</span>, threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh.kaks</span><span class="op">)</span>
<span class="co">#&gt;           aa1                   aa2 rbh_class         ka        ks          vka</span>
<span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh 0.74651201 9.9999990 0.0032172312</span>
<span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh 0.00000000 0.1831839 0.0000000000</span>
<span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh 0.05680851 0.1390746 0.0001612571</span>
<span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh 0.03656808 0.1543978 0.0001014973</span>
<span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh 0.75036088 9.9999990 0.0093948533</span>
<span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh 0.78228833 9.9999990 0.0051807370</span>
<span class="co">#&gt;           vks</span>
<span class="co">#&gt; 1 9.999999000</span>
<span class="co">#&gt; 2 0.003022307</span>
<span class="co">#&gt; 3 0.001531756</span>
<span class="co">#&gt; 4 0.001249826</span>
<span class="co">#&gt; 5 9.999999000</span>
<span class="co">#&gt; 6 9.999999000</span>

<span class="co">## get help</span>
<span class="co">#?rbh2kaks</span></code></pre></div>
<p>Like for the <code><a href="../reference/cds2kaks.html">cds2kaks()</a></code> function, all parameters which can be used by the <code><a href="../reference/cds2codonaln.html">cds2codonaln()</a></code> function can be overgiven to e.g.Â use an alternative <code>substitutionMatrix</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## calculate Ka/Ks using the CRBHit pairs and multiple threads</span>
<span class="va">ath_aly_crbh.kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span>rbhpairs <span class="op">=</span> <span class="va">ath_aly_crbh</span>,
                              cds1 <span class="op">=</span> <span class="va">ath</span>, cds2 <span class="op">=</span> <span class="va">aly</span>,
                              model <span class="op">=</span> <span class="st">"Li"</span>, threads <span class="op">=</span> <span class="fl">2</span>,
                              substitutionMatrix <span class="op">=</span> <span class="st">"BLOSUM45"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh.kaks</span><span class="op">)</span>
<span class="co">#&gt;           aa1                   aa2 rbh_class         ka        ks          vka</span>
<span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh 0.69109360 9.9999990 0.0029376709</span>
<span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh 0.00000000 0.1831839 0.0000000000</span>
<span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh 0.05638742 0.1366968 0.0001613461</span>
<span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh 0.03656808 0.1543978 0.0001014973</span>
<span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh 0.66169565 9.9999990 0.0073901048</span>
<span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh 0.75382109 9.9999990 0.0047473367</span>
<span class="co">#&gt;           vks</span>
<span class="co">#&gt; 1 9.999999000</span>
<span class="co">#&gt; 2 0.003022307</span>
<span class="co">#&gt; 3 0.001436133</span>
<span class="co">#&gt; 4 0.001249826</span>
<span class="co">#&gt; 5 9.999999000</span>
<span class="co">#&gt; 6 9.999999000</span>

<span class="co">## get help</span>
<span class="co">#?rbh2kaks</span></code></pre></div>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>4. <span id="references">References</span>
</h2>
<p>Aubry S., Kelly S., KÃ¼mpers B. M., Smith-Unna R. D., and Hibberd J. M. (2014). <strong>Deep evolutionary comparison of gene expression identifies parallel recruitment of trans-factors in two independent origins of C4 photosynthesis.</strong> <em>PLoS Genetics</em>, <strong>10(6)</strong>. <a href="https://doi.org/10.1371/journal.pgen.1004365">https://doi.org/10.1371/journal.pgen.1004365</a></p>
<p>Charif D., and Lobry J. R. (2007). <strong>SeqinR 1.0-2: a contributed package to the R project for statistical computing devoted to biological sequences retrieval and analysis.</strong> <em>In Structural approaches to sequence evolution</em> (pp.Â 207-232). Springer, Berlin, Heidelberg. <a href="https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10">https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10</a></p>
<p>Duong T., and Wand M. (2015). <strong>feature: Local Inferential Feature Significance for Multivariate Kernel Density Estimation.</strong> <em>R package version 1.2.13</em>. <a href="https://cran.r-project.org/web/packages/feature/">https://cran.r-project.org/web/packages/feature/</a></p>
<p>Haas B. J., Delcher A. L., Wortman J. R., and Salzberg S. L. (2004). <strong>DAGchainer: a tool for mining segmental genome duplications and synteny.</strong> <em>Bioinformatics</em>, <strong>20(18)</strong>, 3643-3646. <a href="https://doi.org/10.1093/bioinformatics/bth397">https://doi.org/10.1093/bioinformatics/bth397</a></p>
<p>Haug-Baltzell A., Stephens S. A., Davey S., Scheidegger C. E., Lyons E. (2017). <strong>SynMap2 and SynMap3D: web-based wholge-genome synteny browsers.</strong> <em>Bioinformatics</em>, <strong>33(14)</strong>, 2197-2198. <a href="https://academic.oup.com/bioinformatics/article/33/14/2197/3072872">https://academic.oup.com/bioinformatics/article/33/14/2197/3072872</a></p>
<p>KieÅ‚basa S. M., Wan R., Sato K., Horton P., and Frith M. C. (2011). <strong>Adaptive seeds tame genomic sequence comparison.</strong> <em>Genome Research</em>, <strong>21(3)</strong>, 487-493. <a href="https://doi.org/10.1101/gr.113985.110">https://doi.org/10.1101/gr.113985.110</a></p>
<p>Kimura M. (1977). <strong>Preponderance of synonymous changes as evidence for the neutral theory of molecular evolution.</strong> <em>Nature</em>, <strong>267</strong>, 275-276.</p>
<p>Li W. H. (1993). <strong>Unbiased estimation of the rates of synonymous and nonsynonymous substitution.</strong> <em>Journal of Molecular Evolution</em>, <strong>36(1)</strong>, 96-99. <a href="https://doi.org/10.1007/bf02407308">https://doi.org/10.1007/bf02407308</a></p>
<p>Microsoft, and Weston S. (2020). <strong>foreach: Provides Foreach Looping Construct.</strong> <em>R package version</em>, 1.5.1. <a href="https://cran.r-project.org/web/packages/foreach/index.html">foreach</a></p>
<p>Mugal C. F., Wolf J. B. W., Kaj I. (2014). <strong>Why Time Matters: Codon Evolution and the Temproal Dynamics of dN/dS.</strong> <em>Molecular Biology and Evolution</em>, <strong>31(1)</strong>, 212-231.</p>
<p>Ooms J. (2019). <strong>curl: A Modern and Flexible Web Client for R.</strong> <em>R package version</em>, 4.3. <a href="https://cran.r-project.org/web/packages/curl/index.html">curl</a></p>
<p>PagÃ¨s H., Aboyoun P., Gentleman R., and DebRoy S. (2017). <strong>Biostrings: Efficient manipulation of biological strings.</strong> <em>R package version</em>, 2.56.0. <a href="http://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a></p>
<p>Revolution Analytics, and Weston S. (2020). <strong>doMC: Foreach Parallel Adaptor for â€˜parallelâ€™.</strong> <em>R package version</em>, 1.3.7. <a href="https://cran.r-project.org/web/packages/doMC/index.html">doMC</a></p>
<p>Rost B. (1999). <strong>Twilight zone of protein sequence alignments.</strong> <em>Protein Engineering</em>, <strong>12(2)</strong>, 85-94. <a href="https://doi.org/10.1093/protein/12.2.85">https://doi.org/10.1093/protein/12.2.85</a></p>
<p>Scott C. (2017). <strong>shmlast: an improved implementation of conditional reciprocal best hits with LAST and Python.</strong> <em>Journal of Open Source Software</em>, <strong>2(9)</strong>, 142. <a href="https://joss.theoj.org/papers/10.21105/joss.00142">https://joss.theoj.org/papers/10.21105/joss.00142</a></p>
<p>Scrucca L., Fop M., Murphy T. B., and Raftery A. E. (2016) <strong>mclust 5: clustering, classification and density estimation using Gaussian finite mixture models.</strong> <em>The R Journal</em>, 8(1), 289-317. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/</a></p>
<p>Wickham H. (2011). <strong>testthat: Get Started with Testing.</strong> <em>The R Journal</em>, 3(1), 5. <a href="https://cran.r-project.org/web/packages/testthat/index.html">testthat</a></p>
<p>Wickham H. (2019). <strong>stringr: Simple, Consistent Wrappers for Common String Operations.</strong> <em>R package version</em>, 1.4.0. <a href="https://cran.r-project.org/web/packages/stringr/index.html">stringr</a></p>
<p>Wickham H. (2020). <strong>tidyr: Tidy Messy Data.</strong> <em>R package version</em>, 1.1.2. <a href="https://cran.r-project.org/web/packages/tidyr/index.html">tidyr</a></p>
<p>Wickham H., Hester J., and Chang W. (2020). <strong>devtools: Tools to make Developing R Packages Easier.</strong> <em>R package version</em>, (2.3.2). <a href="https://cran.r-project.org/web/packages/devtools/index.html">devtools</a></p>
<p>Wickham H., FranÃ§ois R., Henry L., and MÃ¼ller K. (2020). <strong>dplyr: A Grammar of Data Manipulation.</strong> <em>R package version</em>, 1.0.2. <a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a></p>
<p>Yang Z., and Nielsen R. (2000). <strong>Estimating synonymous and nonsynonymous substitution rates under realistic evolutionary models.</strong> <em>Molecular Biology and Evolution</em>, <strong>17(1)</strong>, 32-43. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">https://doi.org/10.1093/oxfordjournals.molbev.a026236</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
