<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CRBHits Basic Vignette • CRBHits</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CRBHits Basic Vignette">
<meta property="og:description" content="CRBHits">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/V02KaKsVignette.html">KaKs Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">
    <span class="fab fa-gitlab fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CRBHits Basic Vignette</h1>
                        <h4 data-toc-skip class="author">Kristian K
Ullrich</h4>
            
            <h4 data-toc-skip class="date">2023-12-22</h4>
      
      <small class="dont-index">Source: <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/develvignettes/V01CRBHitsBasicVignette.Rmd" class="external-link"><code>vignettes/V01CRBHitsBasicVignette.Rmd</code></a></small>
      <div class="hidden name"><code>V01CRBHitsBasicVignette.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="crbhit-basic-vignette">CRBHit Basic Vignette<a class="anchor" aria-label="anchor" href="#crbhit-basic-vignette"></a>
</h2>
<ul>
<li>includes CRBHit pair calculation</li>
<li>includes CRBHit pair filtering</li>
<li>includes Longest Isoform selection</li>
<li>includes Codon alignments</li>
<li>includes Ka/Ks calculations</li>
</ul>
<div class="section level3">
<h3 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h3>
<ol style="list-style-type: decimal">
<li><a href="#install">Installation</a></li>
<li>
<a href="#crbhalgorithm">Conditional Reciprocal Best Hits -
Algorithm</a>
<ol style="list-style-type: decimal">
<li>
<a href="#crbhstep1">1. step: sequence similarity search (last)</a>
<ol style="list-style-type: decimal">
<li><a href="#cdsinput">Input: Coding Sequences (CDS)</a></li>
<li><a href="#isoform">Get Longest Isoform from NCBI or ENSEMBL Input
(optional)</a></li>
<li><a href="#sequencesearch">Sequence Similarity Search</a></li>
</ol>
</li>
<li>
<a href="#crbhstep2">2. step: filter hits + RBH extraction</a>
<ol style="list-style-type: decimal">
<li><a href="#filters">Filter blast-like output prior fitting</a></li>
<li><a href="#customfilters">Custom Filter</a></li>
<li><a href="#rbhextraction">Extract RBHs</a></li>
</ol>
</li>
<li>
<a href="#crbhstep3">3. step: fitting RBHs + get CRBHs</a>
<ol style="list-style-type: decimal">
<li><a href="#fittingparam">CRBHit pairs - Fitting Parameter</a></li>
</ol>
</li>
</ol>
</li>
<li>
<a href="#KaKs">Ka/Ks Calculations</a>
<ol style="list-style-type: decimal">
<li><a href="#cds2codon">Codon Alignments - cds2codonaln()
Function</a></li>
<li><a href="#cds2kaks">Calculate Ka/Ks values comparing two sequences -
cds2kaks() Function</a></li>
<li><a href="#rbh2kaks">Calculate Ka/Ks values for each CRBHit pair -
rbh2kaks() Function</a></li>
</ol>
</li>
<li><a href="#references">References</a></li>
</ol>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a> is
a reimplementation of the Conditional Reciprocal Best Hit (CRBH)
algorithm <a href="https://github.com/cboursnell/crb-blast" class="external-link">crb-blast</a> in R.</p>
<p>The Reciprocal Best Hit (RBH) approach is commonly used in
bioinformatics to show that two sequences evolved from a common
ancestral gene. In other words, RBH tries to find orthologous protein
sequences within and between species. These orthologous sequences can be
further analysed to evaluate protein family evolution, infer
phylogenetic trees and to annotate protein function.</p>
<p>The CRBH algorithm builds upon the traditional RBH approach to find
additional orthologous sequences between two sets of sequences.</p>
<p>CRBH uses the sequence search results to fit an expect-value
(e-value) cutoff given each RBH to subsequently add sequence pairs to
the list of <em>bona-fide</em> orthologs given their alignment length.
The sequence similarity results can be filtered by common criteria like
e-value, query-coverage and more. If chromosome annotations are
available for the protein-coding genes (protein order), this information
can be directly used to assign tandem duplicates before conducting
further processing steps.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a>
features downstream anaylsis like batch calculating synonymous (Ks) and
nonsynonymous substitutions (Ka) per orthologous sequence pair. The
codon sequence alignment step consist of three subtasks, namely coding
nucleotide to protein sequence translation, pairwise protein sequence
alignment calculation and converting the protein sequence alignment back
into a codon based alignment.</p>
<p><strong>Note:</strong></p>
<p>In this Vignette the <code>lastpath</code> is defined as
<code>vignette.paths[1]</code>, the <code>kakscalcpath</code> as
<code>vignette.paths[2]</code> and the <code>dagchainerpath</code> as
<code>vignette.paths[3]</code> to be able to build the Vignette.</p>
<p>However, once you have compiled last-1521, KaKs_Calculator2.0 and
DAGchainer with the functions <code><a href="../reference/make_last.html">make_last()</a></code>,
<code><a href="../reference/make_KaKs_Calculator2.html">make_KaKs_Calculator2()</a></code> and <code><a href="../reference/make_dagchainer.html">make_dagchainer()</a></code>
you won’t need to specify the paths anymore. Please remove them if you
would like to repeat the examples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load vignette specific libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.r-universe.dev/curl" class="external-link">curl</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## compile LAST, KaKs_Calculator2.0 and DAGchainer for the vignette</span></span>
<span><span class="va">vignette.paths</span> <span class="op">&lt;-</span> <span class="fu">CRBHits</span><span class="fu">::</span><span class="fu"><a href="../reference/make_vignette.html">make_vignette</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="installation">1. <span id="install">Installation</span><a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>See the R package page for a detailed description of the install
process and its dependencies <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/" class="external-link">https://mpievolbio-it.pages.gwdg.de/crbhits/</a>.</p>
<pre><code><span><span class="co">## see a detailed description of installation prerequisites at</span></span>
<span><span class="co">## https://mpievolbio-it.pages.gwdg.de/crbhits/</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="co">## from gitlab</span></span>
<span><span class="fu">install_gitlab</span><span class="op">(</span><span class="st">"mpievolbio-it/crbhits"</span>, host <span class="op">=</span> <span class="st">"https://gitlab.gwdg.de"</span>,</span>
<span>               build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, dependencies <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">## from github</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"kullrich/CRBHits"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, dependencies <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a>
needs <a href="https://gitlab.com/mcfrith/last" class="external-link">LAST</a>, <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download" class="external-link">KaKs_Calculator2.0</a>
and <a href="http://dagchainer.sourceforge.net/" class="external-link">DAGchainer</a> to be
installed before one can efficiently use it.</p>
<p>All prerequisites (LAST, KaKs_Calculator2.0, DAGchainer) are forked
within <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a> and can
be compiled for Linux/Unix/macOS with the functions
<code><a href="../reference/make_last.html">make_last()</a></code>, <code><a href="../reference/make_KaKs_Calculator2.html">make_KaKs_Calculator2()</a></code> and
<code><a href="../reference/make_dagchainer.html">make_dagchainer()</a></code>.</p>
<p>For the windows platform the user needs to use e.g. <a href="https://www.cygwin.com/" class="external-link">https://www.cygwin.com/</a> to be able to
compile the prerequisites.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a></span><span class="op">)</span></span>
<span><span class="co">## compile last-1521</span></span>
<span><span class="fu"><a href="../reference/make_last.html">make_last</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## compile KaKs_Calculator2.0</span></span>
<span><span class="fu"><a href="../reference/make_KaKs_Calculator2.html">make_KaKs_Calculator2</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">## compile DAGchainer</span></span>
<span><span class="fu"><a href="../reference/make_dagchainer.html">make_dagchainer</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="conditional-reciprocal-best-hits---algorithm">2. <span id="crbhalgorithm">Conditional Reciprocal Best Hits -
Algorithm</span><a class="anchor" aria-label="anchor" href="#conditional-reciprocal-best-hits---algorithm"></a>
</h3>
<div class="float">
<img src="cds2rbhoverview.png" width="400" alt="Figure: Overview of the cds2rbh() function"><div class="figcaption">Figure: Overview of the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code>
function</div>
</div>
<p>The CRBH algorithm was introduced by <a href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1004365" class="external-link">Aubry
S, Kelly S et al. (2014)</a> and ported to python <a href="https://pypi.org/project/shmlast/" class="external-link">shmlast</a> (<a href="https://joss.theoj.org/papers/10.21105/joss.00142" class="external-link">Scott C.
2017</a>) which benefits from the blast-like sequence search software <a href="http://last.cbrc.jp/" class="external-link">LAST</a> (<a href="https://genome.cshlp.org/content/21/3/487.long" class="external-link">Kiełbasa SM et
al. 2011</a>).</p>
<p>The CRBH algorithm implementation consists of three steps which are
explained in more detail in the following subsections.</p>
<ol style="list-style-type: decimal">
<li>step: sequence similarity search (last)(#crbhstep1)</li>
<li>step: filter hits and extract Reciprocal Best Hits
(RBHs)(#crbhstep2)</li>
<li>step: fitting RBHs and get Conditional Reciprocal Best Hits
(CRBHs)(#crbhstep3)</li>
</ol>
<p>All of these steps are combined in either the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code>
function (R vector input) or the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function
(file input).</p>
<p>To activate/deactivate specific options use the parameters
<code>@param</code> given and explained for each function (see
<code>?cds2rbh()</code> and <code>?cdsfile2rbh()</code>).</p>
<div class="float">
<img src="cds2rbh.png" width="800" alt="Figure: Individual steps of the cds2rbh() function"><div class="figcaption">Figure: Individual steps of the
<code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function</div>
</div>
<div class="section level4">
<h4 id="step-sequence-similarity-search-last">2.1. <span id="crbhstep1">1. step: sequence similarity search
(last)</span><a class="anchor" aria-label="anchor" href="#step-sequence-similarity-search-last"></a>
</h4>
<p>All further described steps are included in the
<code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function. The
subsections will just explain in more detail what is happening with the
input data.</p>
<div class="section level5">
<h5 id="input-coding-sequences-cds">2.1.1. <span id="cdsinput">Input: Coding Sequences (CDS)</span><a class="anchor" aria-label="anchor" href="#input-coding-sequences-cds"></a>
</h5>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a>
only takes coding nucleotide sequences as the query and target
inputs.</p>
<p><strong>Why?</strong> Because <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a> is
implemented to use the same input afterwards to calculate synonymous and
nonsynonymous substitutions which must match the protein sequences used
to find CRBHit pairs.</p>
<p>One can either use a <code>DNAStringSet</code> vector from the
<code>Biostrings</code> R package with the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code>
function or use <strong>raw</strong>,<strong>gzipped</strong> or
<strong>URL</strong> coding sequences <strong>fasta</strong> files as
inputs with the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a>
uses either the function <code><a href="../reference/cdsfile2aafile.html">cdsfile2aafile()</a></code> to translate a
coding sequence <strong>fasta</strong> file into its corresponding
protein sequence <strong>fasta</strong> file or <code>cds2aa()</code> to
translate a <code>DNAStringSet</code> into a <code>AAStringSet</code>.
Thereby it removes all sequences that are not a multiple of three which
can not be parsed correctly.</p>
<p><strong>Note:</strong> The user can easily check if all coding
sequences would be a multiple of three or rely on the input files
generated by other sources.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to check coding sequences if all are a mutiple of three</span></span>
<span></span>
<span><span class="co">## load CDS file</span></span>
<span><span class="va">cdsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="va">cds</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile</span><span class="op">)</span></span>
<span><span class="co">## the following statement should return TRUE,</span></span>
<span><span class="co">## if all sequences are a mutiple of three</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu">Biostrings</span><span class="fu">::</span><span class="fu">width</span><span class="op">(</span><span class="va">cds</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">3</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="get-longest-isoform-from-ncbi-or-ensembl-input-optional">2.1.2. <span id="isoform">Get Longest Isoform from NCBI or ENSEMBL
Input (optional)</span><a class="anchor" aria-label="anchor" href="#get-longest-isoform-from-ncbi-or-ensembl-input-optional"></a>
</h5>
<p>It is also possible to use an URL to directly access coding sequences
from <a href="https://www.ncbi.nlm.nih.gov/genome" class="external-link">NCBI</a> or <a href="ftp://ftp.ensembl.org/pub/current_fasta">ENSEMBL</a> and reduce
this input to the longest annotated isoform.</p>
<pre><code><span><span class="co">## example how to access CDS from URL and get longest isoform</span></span>
<span></span>
<span><span class="co">## get coding sequences for Homo sapiens from NCBI</span></span>
<span><span class="va">HOMSAP.cds.NCBI.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/"</span>,</span>
<span>  <span class="st">"GCF_000001405.39_GRCh38.p13/"</span>,</span>
<span>  <span class="st">"GCF_000001405.39_GRCh38.p13_cds_from_genomic.fna.gz"</span><span class="op">)</span></span>
<span><span class="va">HOMSAP.cds.NCBI</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">HOMSAP.cds.NCBI.url</span><span class="op">)</span></span>
<span><span class="co">## reduce to the longest isoform</span></span>
<span><span class="va">HOMSAP.cds.NCBI.longest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="va">HOMSAP.cds.NCBI</span>, <span class="st">"NCBI"</span><span class="op">)</span></span>
<span><span class="co">## get coding sequences for Homo sapiens from ENSEMBL</span></span>
<span><span class="va">HOMSAP.cds.ENSEMBL.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="st">"ftp://ftp.ensembl.org/pub/release-101/fasta/"</span>,</span>
<span>  <span class="st">"homo_sapiens/cds/Homo_sapiens.GRCh38.cds.all.fa.gz"</span><span class="op">)</span></span>
<span><span class="va">HOMSAP.cds.ENSEMBL.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">HOMSAP.cds.ENSEMBL.url</span>, <span class="va">HOMSAP.cds.ENSEMBL.file</span>, quiet<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">HOMSAP.cds.ENSEMBL</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">HOMSAP.cds.ENSEMBL.file</span><span class="op">)</span></span>
<span><span class="co">## reduce to the longest isoform</span></span>
<span><span class="va">HOMSAP.cds.ENSEMBL.longest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="va">HOMSAP.cds.ENSEMBL</span>, <span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?isoform2longest</span></span></code></pre>
<p><strong>Note:</strong> It is also possible to obtain and select the
longest/primary isoform from CDS Input sources other than NCBI or
Ensembl. This can be achieved by parsing a GTF/GFF3 file. This special
case is handled in the <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V02KaKsVignette.html" class="external-link">CRBHits
KaKs Vignette</a> under point 1.2.2, 1.2.3 and 1.4.2.</p>
</div>
<div class="section level5">
<h5 id="sequence-similarity-search">2.1.3. <span id="sequencesearch">Sequence Similarity
Search</span><a class="anchor" aria-label="anchor" href="#sequence-similarity-search"></a>
</h5>
<p>The blast-like software <a href="http://last.cbrc.jp/" class="external-link">LAST</a> is
used to compare the translated coding sequences against each other and
output a blast-like output table including the query and target
length.</p>
<p><strong>Note:</strong> Classical RBH can be performed by disabling
the <code>crbh</code> <code>@param</code>of the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code>
or <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to get CRBHit pairs from two CDS using classical RBH</span></span>
<span></span>
<span><span class="co">## define CDS file1</span></span>
<span><span class="va">cdsfile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="co">## define CDS file2</span></span>
<span><span class="va">cdsfile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"aly.cds.fasta.gz"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="co">## get CDS1</span></span>
<span><span class="va">cds1</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile1</span><span class="op">)</span></span>
<span><span class="co">## get CDS2</span></span>
<span><span class="va">cds2</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile2</span><span class="op">)</span></span>
<span><span class="co">## example how to perform classical RBH</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    crbh<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## show summary</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">)</span></span>
<span><span class="co">#&gt;            Length Class      Mode</span></span>
<span><span class="co">#&gt; crbh.pairs  3     data.frame list</span></span>
<span><span class="co">#&gt; crbh1      16     data.frame list</span></span>
<span><span class="co">#&gt; crbh2      16     data.frame list</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?cds2rbh</span></span></code></pre></div>
<p>Like <a href="https://pypi.org/project/shmlast/" class="external-link">shmlast</a>, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a> plots
the fitted model of the CRBH e-value based algorithm. The CRBH algorithm
can be performed and visualized as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to get CRBHit pairs using one thread</span></span>
<span><span class="co">## and plot CRBHit algorithm fitting curve</span></span>
<span></span>
<span><span class="co">## example how to perform CRBH</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## show summary</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">)</span></span>
<span><span class="co">#&gt;               Length Class      Mode    </span></span>
<span><span class="co">#&gt; crbh.pairs     3     data.frame list    </span></span>
<span><span class="co">#&gt; crbh1         16     data.frame list    </span></span>
<span><span class="co">#&gt; crbh2         16     data.frame list    </span></span>
<span><span class="co">#&gt; rbh1_rbh2_fit  1     -none-     function</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?cds2rbh</span></span></code></pre></div>
<p>Both, <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> and <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function,
return a list of three <code>data.frame</code>’s which contain the CRBH
pairs (<code>$crbh.pairs</code>) retained, the query &gt; target
blast-like output matrix (<code>$crbh1</code>) and the target &gt; query
blast-like output matrix (<code>$crbh2</code>). For each matrix the
<code>$rbh_class</code> indicates if the hit is a Reciprocal Best Hit
(<code>rbh</code>) or if it is a conditional secondary hit retained
because of RBH fitting (CRBH algorithm) (<code>sec</code>).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example showing cds2rbh() results</span></span>
<span></span>
<span><span class="co">## show dimension and first parts of retained hit pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 211   3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt;           aa1                   aa2 rbh_class</span></span>
<span><span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh</span></span>
<span><span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh</span></span>
<span><span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh</span></span>
<span><span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh</span></span>
<span><span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh</span></span>
<span><span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh</span></span>
<span></span>
<span><span class="co">## show first retained hit pairs for the query &gt; target matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span><span class="op">)</span></span>
<span><span class="co">#&gt;      query_id            subject_id perc_identity alignment_length mismatches</span></span>
<span><span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256         31.09              119         73</span></span>
<span><span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128        100.00              213          0</span></span>
<span><span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125         90.75              281         19</span></span>
<span><span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114         92.88              309         19</span></span>
<span><span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419         33.33               81         51</span></span>
<span><span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326         44.94               89         41</span></span>
<span><span class="co">#&gt;   gap_opens q_start q_end s_start s_end   evalue bit_score query_length</span></span>
<span><span class="co">#&gt; 1         2     650   768     354   463  4.0e-06      47.0         1910</span></span>
<span><span class="co">#&gt; 2         0       1   213       1   213 1.5e-160     493.0          213</span></span>
<span><span class="co">#&gt; 3         3       1   275       1   280 5.9e-178     551.0          287</span></span>
<span><span class="co">#&gt; 4         1       1   306       1   309 2.9e-218     668.0          308</span></span>
<span><span class="co">#&gt; 5         2     450   529      91   169  9.6e-09      53.6          541</span></span>
<span><span class="co">#&gt; 6         1     409   497     251   331  3.4e-16      78.3          591</span></span>
<span><span class="co">#&gt;   subject_length raw_score rbh_class</span></span>
<span><span class="co">#&gt; 1            540        99       rbh</span></span>
<span><span class="co">#&gt; 2            213      1110       rbh</span></span>
<span><span class="co">#&gt; 3            299      1240       rbh</span></span>
<span><span class="co">#&gt; 4            316      1505       rbh</span></span>
<span><span class="co">#&gt; 5            189       114       rbh</span></span>
<span><span class="co">#&gt; 6            451       170       rbh</span></span>
<span></span>
<span><span class="co">## get the number of CRBHit classified as rbh and sec hit pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; rbh sec </span></span>
<span><span class="co">#&gt; 181  30</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh2</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; rbh sec </span></span>
<span><span class="co">#&gt; 181  30</span></span></code></pre></div>
<p>If the <code>@param</code> <code>crbh</code> was set to
<code>TRUE</code>, the CRBH algorithm was applied and the fitting
function will also be returned. The fitting function can than be used
for manual plotting as follows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to use the fitting function for manual plotting</span></span>
<span></span>
<span><span class="co">## plot fitting function</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/curve.html" class="external-link">curve</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="fu">rbh1_rbh2_fit</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,</span>
<span>    from<span class="op">=</span><span class="fl">1</span>,</span>
<span>    to<span class="op">=</span><span class="fl">1000</span>,</span>
<span>    xlab<span class="op">=</span><span class="st">"alnlength"</span>,</span>
<span>    ylab<span class="op">=</span><span class="st">"-log10(evalue)"</span>,</span>
<span>    main<span class="op">=</span><span class="st">"CRBH fitting"</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>One can also specify if a hit pair which is only found in one
direction should be retained, which will be classified as
(<code>single</code>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to retain single direction secondary homologs</span></span>
<span></span>
<span><span class="co">## get CRBHit pairs with keepSingleDirection = TRUE</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    keepSingleDirection<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get the number of CRBHit classified as rbh and sec hit pairs</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    rbh    sec single </span></span>
<span><span class="co">#&gt;    181     30      1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh2</span><span class="op">$</span><span class="va">rbh_class</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    rbh    sec single </span></span>
<span><span class="co">#&gt;    181     30      2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?cds2rbh</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="step-filter-hits-rbh-extraction">2.2. <span id="crbhstep2">2. step: filter hits + RBH
extraction</span><a class="anchor" aria-label="anchor" href="#step-filter-hits-rbh-extraction"></a>
</h4>
<p>All further described steps are included in the
<code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function. The
subsections will just explain in more detail what is happening with the
input data.</p>
<p>Until now the hit pairs obtained via the sequence similarity searches
were only filtered by default settings (e-value &lt; 0.001). The
following subsections will explain the default filters in more detail
and show how one can create and apply custom filter on the hit pairs
prior RBH extraction and fitting to obtain CRBHit pairs.</p>
<div class="section level5">
<h5 id="filter-blast-like-output-prior-fitting">2.2.1. <span id="filters">Filter blast-like output prior
fitting</span><a class="anchor" aria-label="anchor" href="#filter-blast-like-output-prior-fitting"></a>
</h5>
<p>The following filters are already defined and can be used out of the
box:</p>
<ul>
<li>evalue (<code>@param</code> <code>evalue = 0.001</code>)</li>
<li>query coverage (<code>@param</code> <code>qcov = 0.0</code>)</li>
<li>target coverage (<code>@param</code> <code>tcov = 0.0</code>)</li>
<li>protein identity (<code>@param</code> <code>pident = 0</code>)</li>
<li>alignment length (<code>@param</code> <code>alnlen = 0</code>)</li>
<li>rost1999 (<code>@param</code> <code>rost1999 = FALSE</code>)</li>
</ul>
<p>Taking the length of the obtained pairwise protein alignment one can
calculate for each hit pair the query coverage as <span class="math inline">\(\frac{alignment length}{query
length}\)</span>.</p>
<p>The data will be filtered with a query coverage of 50% using the
<code>@param</code> <code>qcov</code> as follows:
<code>cds2rbh(..., qcov = 0.5)</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to filter prior crbh for query coverage of 50%</span></span>
<span></span>
<span><span class="co">## get CRBHit pairs with direct query coverage filter</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    qcov<span class="op">=</span><span class="fl">0.5</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 145   3</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?cds2rbh</span></span></code></pre></div>
<p>In addition, the users can filter for the <strong>twilight zone of
protein sequence alignments</strong> according to <a href="https://academic.oup.com/peds/article/12/2/85/1550637" class="external-link">Rost B.
(1999)</a>. This <strong>twilight zone</strong> indicates proteins,
which might share a common ancestor (homologue genes) but have diverged
from each other through time so that similarity is no longer detectable.
Here, a certain alignment length and protein identity is at least
necessary to indicate <em>bona-fide</em> homologues.</p>
<p>The following equation and plot will show how the relationship
between alignment length and protein identity is defined for the
<code>rost1999</code> filter.</p>
<p>The implemented <code>rost1999</code> filter uses
<code>equation2</code> of <a href="https://academic.oup.com/peds/article/12/2/85/1550637" class="external-link">Rost B.
(1999)</a></p>
<p><span class="math display">\[f(x_{\text{hit pair}}) = \cases {
100 \text{ , for } L_{\text{hit pair}} &lt; 11 \\
480 * L^{-0.32 * (1 + e^{\frac{-L}{1000}})
} \text{ , for } L_{\text{hit pair}} \leq 450 \\
19.5 \text{ , for } L_{\text{hit pair}} &gt; 450}\]</span> , where <span class="math inline">\(x_{\text{hit pair}}\)</span> is the expected
protein identity given the alignemnet length <span class="math inline">\(L_{\text{hit pair}}\)</span>. If the actual <span class="math inline">\(pident_{\text{hit pair}} &gt;= f(x_{\text{hit
pair}})\)</span> the hit pair is retained.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## detailed explanation for the Rost (1999) twilight-zone filter</span></span>
<span></span>
<span><span class="co">## define eq2 from Rost (1999)</span></span>
<span><span class="va">get_pident_by_length</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">eq2</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">L</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="va">L</span><span class="op">&lt;=</span><span class="fl">11</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">}</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="va">L</span><span class="op">&lt;=</span><span class="fl">450</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fl">480</span><span class="op">*</span><span class="op">(</span><span class="va">L</span><span class="op">^</span><span class="op">(</span><span class="op">-</span><span class="fl">0.32</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="va">L</span><span class="op">/</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">}</span></span>
<span>        <span class="kw">if</span><span class="op">(</span><span class="va">L</span><span class="op">&gt;</span><span class="fl">450</span><span class="op">)</span><span class="op">{</span><span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fl">19.5</span><span class="op">)</span><span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">eq2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## plot expected pident by alignment length using eq2 from Rost (1999)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/curve.html" class="external-link">curve</a></span><span class="op">(</span><span class="va">get_pident_by_length</span>,</span>
<span>    <span class="fl">11</span>,</span>
<span>    <span class="fl">500</span>,</span>
<span>    pch<span class="op">=</span><span class="fl">20</span>,</span>
<span>    xlab<span class="op">=</span><span class="st">"alignment length"</span>,</span>
<span>    ylab<span class="op">=</span><span class="st">"pident"</span>,</span>
<span>    main<span class="op">=</span><span class="st">"expected protein identity (eq2; Rost B. 1999)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>This <strong>twilight zone</strong> filter can be directly used by
the <code>@param</code> <code>rost1999</code> set to
<code>TRUE</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to filter prior crbh for eq2 from Rost (1999)</span></span>
<span></span>
<span><span class="co">## get CRBHit pairs with direct twilight-zone filter</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    rost1999<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 208   3</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?cds2rbh</span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?filter.alnlen</span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?filter.eval</span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?filter.pident</span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?filter.qcov</span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?filter.rost1999</span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?filter.tcov</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="custom-filter">2.2.2. <span id="filters">Custom Filter</span><a class="anchor" aria-label="anchor" href="#custom-filter"></a>
</h5>
<p>The user can also define its own filter function and apply it to the
blast-like output prior fitting and calculating CRBHs as follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example for a custom filters</span></span>
<span></span>
<span><span class="co">## define custom filter for e.g. bit score (column 12)</span></span>
<span><span class="va">myfilter1</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rbh</span>, <span class="va">value</span><span class="op">=</span><span class="fl">500.0</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rbh</span>, <span class="va">bit_score</span><span class="op">&gt;=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">## define custom filter for e.g. corrected query_coverage</span></span>
<span><span class="va">myfilter2</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rbh</span>, <span class="va">value</span><span class="op">=</span><span class="fl">0.5</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rbh</span>,</span>
<span>        <span class="op">(</span><span class="op">(</span><span class="va">alignment_length</span><span class="op">-</span><span class="va">mismatches</span><span class="op">-</span><span class="va">gap_opens</span><span class="op">)</span> <span class="op">/</span> <span class="va">query_length</span><span class="op">)</span><span class="op">&gt;=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">## get CRBHit pairs with custom filter list</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    filter<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">myfilter1</span>, <span class="va">myfilter2</span><span class="op">)</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 58  3</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="extract-rbhs">2.2.3. <span id="rbhextraction">Extract RBHs</span><a class="anchor" aria-label="anchor" href="#extract-rbhs"></a>
</h5>
<p>As already explained in the Sequence Similarity Search section
(#sequencesearch), both, <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and
<code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function, return a list of three
<code>data.frame</code>’s which contain the CRBHit pairs
(<code>$crbh.pairs</code>) retained, the query &gt; target blast-like
output matrix (<code>$crbh1</code>) and the target &gt; query blast-like
output matrix (<code>$crbh2</code>). For each matrix the
<code>$rbh_class</code> indicates if the hit is a Reciprocal Best Hit
(<code>rbh</code>) or if it is a conditional secondary hit retained
because of RBH fitting (CRBH algorithm) (<code>sec</code>).</p>
<p>To extract just the RBHs one can subset based on the
<code>rbh_class</code> columns as follows:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example to extract CRBHit pairs classified as rbh</span></span>
<span></span>
<span><span class="co">## reduce to rbh_class rbh</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ath_aly_crbh"</span>, package <span class="op">=</span> <span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh1</span>, <span class="va">rbh_class</span><span class="op">==</span><span class="st">"rbh"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      query_id            subject_id perc_identity alignment_length mismatches</span></span>
<span><span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256         31.09              119         73</span></span>
<span><span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128        100.00              213          0</span></span>
<span><span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125         90.75              281         19</span></span>
<span><span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114         92.88              309         19</span></span>
<span><span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419         33.33               81         51</span></span>
<span><span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326         44.94               89         41</span></span>
<span><span class="co">#&gt;   gap_opens q_start q_end s_start s_end   evalue bit_score query_length</span></span>
<span><span class="co">#&gt; 1         2     650   768     354   463  4.0e-06      47.0         1910</span></span>
<span><span class="co">#&gt; 2         0       1   213       1   213 1.5e-160     493.0          213</span></span>
<span><span class="co">#&gt; 3         3       1   275       1   280 5.9e-178     551.0          287</span></span>
<span><span class="co">#&gt; 4         1       1   306       1   309 2.9e-218     668.0          308</span></span>
<span><span class="co">#&gt; 5         2     450   529      91   169  9.6e-09      53.6          541</span></span>
<span><span class="co">#&gt; 6         1     409   497     251   331  3.4e-16      78.3          591</span></span>
<span><span class="co">#&gt;   subject_length raw_score rbh_class</span></span>
<span><span class="co">#&gt; 1            540        99       rbh</span></span>
<span><span class="co">#&gt; 2            213      1110       rbh</span></span>
<span><span class="co">#&gt; 3            299      1240       rbh</span></span>
<span><span class="co">#&gt; 4            316      1505       rbh</span></span>
<span><span class="co">#&gt; 5            189       114       rbh</span></span>
<span><span class="co">#&gt; 6            451       170       rbh</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh2</span>, <span class="va">rbh_class</span><span class="op">==</span><span class="st">"rbh"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;                query_id  subject_id perc_identity alignment_length mismatches</span></span>
<span><span class="co">#&gt; 1 Al_scaffold_0001_1000 AT1G04940.1         40.48               42         25</span></span>
<span><span class="co">#&gt; 2 Al_scaffold_0001_1004 AT1G10280.1         99.27              413          3</span></span>
<span><span class="co">#&gt; 3 Al_scaffold_0001_1120 AT1G08040.1         45.67              300        151</span></span>
<span><span class="co">#&gt; 4  Al_scaffold_0001_114 AT1G01180.1         92.88              309         19</span></span>
<span><span class="co">#&gt; 5 Al_scaffold_0001_1146 AT1G09740.1         32.54              169        105</span></span>
<span><span class="co">#&gt; 6 Al_scaffold_0001_1151 AT1G03300.1         33.28              673        377</span></span>
<span><span class="co">#&gt;   gap_opens q_start q_end s_start s_end        evalue bit_score query_length</span></span>
<span><span class="co">#&gt; 1         0      81   122     166   207  5.400000e-07      46.1          186</span></span>
<span><span class="co">#&gt; 2         0       1   413       1   413 9.881313e-324     980.0          413</span></span>
<span><span class="co">#&gt; 3         8      83   379      92   382  2.300000e-80     281.0          437</span></span>
<span><span class="co">#&gt; 4         1       1   309       1   306 8.100000e-218     668.0          316</span></span>
<span><span class="co">#&gt; 5         4      34   197       4   168  1.200000e-16      78.3          274</span></span>
<span><span class="co">#&gt; 6        19       7   611       3   671  1.300000e-85     304.0          611</span></span>
<span><span class="co">#&gt;   subject_length raw_score rbh_class</span></span>
<span><span class="co">#&gt; 1            275        97       rbh</span></span>
<span><span class="co">#&gt; 2            413      2212       rbh</span></span>
<span><span class="co">#&gt; 3            383       629       rbh</span></span>
<span><span class="co">#&gt; 4            308      1505       rbh</span></span>
<span><span class="co">#&gt; 5            172       170       rbh</span></span>
<span><span class="co">#&gt; 6            671       681       rbh</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="step-fitting-rbhs-get-crbhs">2.3. <span id="crbhstep3">3. step: fitting RBHs + get
CRBHs</span><a class="anchor" aria-label="anchor" href="#step-fitting-rbhs-get-crbhs"></a>
</h4>
<p>All further described steps are included in the
<code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> and <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function. The
subsections will just explain in more detail what is happening with the
input data.</p>
<div class="section level5">
<h5 id="crbhit-pairs---fitting-parameter">2.3.1. <span id="fittingparam">CRBHit pairs - Fitting
Parameter</span><a class="anchor" aria-label="anchor" href="#crbhit-pairs---fitting-parameter"></a>
</h5>
<p>The original implementation of the CRBH algorithm <a href="https://github.com/cboursnell/crb-blast/blob/master/lib/crb-blast/crb-blast.rb" class="external-link">crb-blast.rb
line 438-518</a> uses certain parameters for fitting the e-value against
the alignment length.</p>
<p>In brief, for each retained reciprocal hit pair (RBH; after
filtering) the e-value and the corresponding alignment length of a hit
pair and its neighborhood is evaluated. The neighborhood is defined by a
weighting parameter of the alignment length (default is set to
<code>@param</code> <code>fit.varweight = 0.1</code>) and a minimum
neighborhood alignment length (default is set to <code>@param</code>
<code>fit.min = 5</code>).</p>
<p>For each unique alignment length <span class="math inline">\(L\)</span> a value <span class="math inline">\(s_{L}\)</span> is defined as the neighborhood
alignment length as</p>
<p><span class="math display">\[s_{L} = \cases {
\text{fit.min} \text{ , for } L * \text{fit.varweight} &lt;
\text{fit.min} \\
L * \text{fit.varweight} \text{ , for } L * \text{fit.varweight} \geq
\text{fit.min}}\]</span> E.g. for an alignment length of 100 and the
default settings, <span class="math inline">\(s_{100} = 100 * 0.1 =
10\)</span>.</p>
<p>All RBHs, for which <span class="math inline">\(L-s_{L} \leq
L_{RBH_i} \leq L + s_{L}\)</span>, in other words, for which their
alignment lengths fall between 90 and 110 are than considered for
fitting a cutoff at that alignment length using the mean of the <span class="math inline">\(-log10(evalue)\)</span>.</p>
<p><span class="math display">\[\text{cutoff}_{L} = \frac{1}{n} \sum_{i
= 1}^{n}{
\cases{324 \text{ , for }  evalue_{RBH_i} = 0 \\
-log10(evalue_{\text{RBH}_i}) \text{ , for } evalue_{RBH_i} \ne
0}}\]</span></p>
<p>The user can now alter between <code>mean</code> and
<code>median</code> using the <code>@param</code>
<code>fit.type = "mean"</code> or <code>fit.type = "median"</code> to
get alternative fitting curves.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to get crbh from two coding fasta files using median fitting</span></span>
<span></span>
<span><span class="co">## get CRBHit pairs with median fitting</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    plotCurve<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>    fit.type<span class="op">=</span><span class="st">"median"</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?cds2rbh</span></span></code></pre></div>
<p>As described earlier, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits" class="external-link">CRBHits</a> features
downstream anaylsis like batch calculating synonymous (Ks) and
nonsynonymous substitutions (Ka) per orthologous sequence pair. The
following sections will give more details about these downstream
functionalities.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="kaks-calculations">3. <span id="KaKs">Ka/Ks Calculations</span><a class="anchor" aria-label="anchor" href="#kaks-calculations"></a>
</h3>
<div class="float">
<img src="rbh2kaksoverview.png" width="800" alt="Figure: Overview of the rbh2kaks() function steps"><div class="figcaption">Figure: Overview of the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code>
function steps</div>
</div>
<p>The resulting CRBHs (see #crbhalgorithm) can be further processed to
e.g. filtered for Tandem Duplicate (see <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V02KaKsVignette.html" class="external-link">CRBHits
KaKs Vignette</a>) and/or can be used to calculate Ka/Ks values for each
CRBH pair (see #rbh2kaks and <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V02KaKsVignette.html" class="external-link">CRBHits
KaKs Vignette</a>).</p>
<p>All further described steps are included in the
<code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function. The subsections will just explain in
more detail what is happening with the input data.</p>
<p>To activate/deactivate specific options use the parameters
<code>@param</code> given and explained for each function (see
<code>?rbh2kaks()</code>).</p>
<p><strong>Note:</strong> To be able to perform <em>Tandem Duplicate
Assignment</em> with the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function one needs to
specify <em>gene positions</em> for both Input CDS (can be extracted
automatically from <em>NCBI</em> or <em>ENSEMBL</em> Input (see
<code><a href="../reference/cds2genepos.html">cds2genepos()</a></code> function)) and supply <em>selfblast</em>
results for both Input CDS.</p>
<div class="float">
<img src="rbh2kaks.png" width="600" alt="Figure: Individual steps of the rbh2kaks() function"><div class="figcaption">Figure: Individual steps of the
<code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function</div>
</div>
<div class="section level4">
<h4 id="codon-alignments---cds2codonaln-function">3.1. <span id="cds2codon">Codon Alignments - cds2codonaln()
Function</span><a class="anchor" aria-label="anchor" href="#codon-alignments---cds2codonaln-function"></a>
</h4>
<p>To be able to calculate synonymous (Ks) and non-synonymous (Ka)
substitutions, one needs to use a codon alignment.</p>
<p>The <code><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/cds2codonaln.html" class="external-link">MSA2dist::cds2codonaln()</a></code> function takes two single
nucleotide sequences as input to obtain such a codon alignment. The
function will convert the nucleotide sequences into amino acid
sequences, align them with the help of the
<code><a href="https://rdrr.io/pkg/Biostrings/man/pairwiseAlignment.html" class="external-link">pairwiseAlignment()</a></code> function from the
<code>Biostrings</code> R package and convert this alignment back into a
codon alignment.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example to get a codon alignment</span></span>
<span></span>
<span><span class="co">## define two CDS</span></span>
<span><span class="va">cds1</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="st">"ATGCAACATTGC"</span><span class="op">)</span></span>
<span><span class="va">cds2</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html" class="external-link">DNAString</a></span><span class="op">(</span><span class="st">"ATGCATTGC"</span><span class="op">)</span></span>
<span><span class="co">## get codon alignment</span></span>
<span><span class="fu">MSA2dist</span><span class="fu">::</span><span class="fu"><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/cds2codonaln.html" class="external-link">cds2codonaln</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 2:</span></span>
<span><span class="co">#&gt;     width seq                                               names               </span></span>
<span><span class="co">#&gt; [1]    12 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                      cds1</span></span>
<span><span class="co">#&gt; [2]    12 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span>---<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                      cds2</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?MSA2dist::cds2codonaln</span></span></code></pre></div>
<p>The user can alter some <code>@param</code> to change the alignment
parameters, like <code>type</code>, <code>substitutionMatrix</code>,
<code>gapOpening</code> and <code>gapExtension</code> costs as defined
by the <code><a href="https://rdrr.io/pkg/Biostrings/man/pairwiseAlignment.html" class="external-link">pairwiseAlignment()</a></code> function from the
<code>Biostrings</code> R package.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example to alter the substitionMatrix and use the BLOSUM45 cost matrix</span></span>
<span><span class="co">## for the codon alignment</span></span>
<span></span>
<span><span class="co">## get codon alignemnt with BLOSUM45 cost matrix</span></span>
<span><span class="fu">MSA2dist</span><span class="fu">::</span><span class="fu"><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/cds2codonaln.html" class="external-link">cds2codonaln</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    substitutionMatrix<span class="op">=</span><span class="st">"BLOSUM45"</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 2:</span></span>
<span><span class="co">#&gt;     width seq                                               names               </span></span>
<span><span class="co">#&gt; [1]    12 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">AA</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                      cds1</span></span>
<span><span class="co">#&gt; [2]    12 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span>---<span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                      cds2</span></span></code></pre></div>
<p>The user can also remove gaps in the codon alignment.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example to remove codon gaps</span></span>
<span></span>
<span><span class="co">## get codon alignment with gaps removed</span></span>
<span><span class="fu">MSA2dist</span><span class="fu">::</span><span class="fu"><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/cds2codonaln.html" class="external-link">cds2codonaln</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">cds1</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">cds2</span>,</span>
<span>    remove.gaps<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; DNAStringSet object of length 2:</span></span>
<span><span class="co">#&gt;     width seq                                               names               </span></span>
<span><span class="co">#&gt; [1]     9 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                         cds1</span></span>
<span><span class="co">#&gt; [2]     9 <span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">T</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span><span style="color: #000000; background-color: #FFAFAF;">A</span><span style="color: #000000; background-color: #FFD7AF;">TT</span><span style="color: #000000; background-color: #AFFFFF;">G</span><span style="color: #000000; background-color: #AFFFAF;">C</span>                                         cds2</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="msa2distdnastring2kaks-function">3.2. <span id="cds2kaks">MSA2dist::dnastring2kaks()
Function</span><a class="anchor" aria-label="anchor" href="#msa2distdnastring2kaks-function"></a>
</h4>
<p>The Ka/Ks (dN/dS) values can be obtained either via the codon model
of <a href="https://www.ncbi.nlm.nih.gov/pubmed/8433381" class="external-link">Li WH.
(1999)</a> implemented in the R package <a href="https://cran.r-project.org/web/packages/seqinr/index.html" class="external-link">seqinr</a>
or the model <a href="https://www.ncbi.nlm.nih.gov/pubmed/10666704" class="external-link">Yang
Z and Nielson R. (2000)</a> implemented in <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download" class="external-link">KaKs_Calculator2.0</a>
(see <a href="#quickinstall">Quick Installation</a> how to compile the
prerequisites).</p>
<p>The <code><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/dnastring2kaks.html" class="external-link">MSA2dist::dnastring2kaks()</a></code> function takes single
nucleotide sequences as input to obtain a codon alignment and calculate
Ka/Ks.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate Ka/Ks on two CDS</span></span>
<span></span>
<span><span class="co">## load example sequence data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ath"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"aly"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="co">## select a sequence pair according to a best hit pair (done for you)</span></span>
<span><span class="va">cds1</span> <span class="op">&lt;-</span> <span class="va">ath</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">cds2</span> <span class="op">&lt;-</span> <span class="va">aly</span><span class="op">[</span><span class="fl">282</span><span class="op">]</span></span>
<span><span class="co">## calculate Ka/Ks values on two CDS using Li model</span></span>
<span><span class="fu">MSA2dist</span><span class="fu">::</span><span class="fu"><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/dnastring2kaks.html" class="external-link">dnastring2kaks</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cds1</span>,</span>
<span>    <span class="va">cds2</span><span class="op">)</span>,</span>
<span>    model<span class="op">=</span><span class="st">"Li"</span>,</span>
<span>    isMSA<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                               OUT</span></span>
<span><span class="co">#&gt; Comp1                                                                                                                                                                                                                                                                           1</span></span>
<span><span class="co">#&gt; Comp2                                                                                                                                                                                                                                                                           2</span></span>
<span><span class="co">#&gt; seq1                                         AT1G01010.1_cds_chromosome:TAIR10:1:3631:5899:1_gene:AT1G01010_gene_biotype:protein_coding_transcript_biotype:protein_coding_gene_symbol:NAC001_description:NAC_domain-containing_protein_1_[Source:UniProtKB/Swiss-Prot;Acc:Q0WV96]</span></span>
<span><span class="co">#&gt; seq2  Al_scaffold_0001_283_cds_chromosome:v.1.0:1:1082169:1084227:1_gene:Al_scaffold_0001_283_gene_biotype:protein_coding_transcript_biotype:protein_coding_description:DTW_domain-containing_protein_[Source:Projected_from_Arabidopsis_thaliana_(AT1G03687)_TAIR;Acc:AT1G03687]</span></span>
<span><span class="co">#&gt; ka                                                                                                                                                                                                                                                               1.02338181273445</span></span>
<span><span class="co">#&gt; ks                                                                                                                                                                                                                                                                       9.999999</span></span>
<span><span class="co">#&gt; vka                                                                                                                                                                                                                                                            0.0114026707748616</span></span>
<span><span class="co">#&gt; vks                                                                                                                                                                                                                                                                      9.999999</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?MSA2dist::dnastring2kaks</span></span></code></pre></div>
<p>All parameters which can be used by the
<code><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/cds2codonaln.html" class="external-link">MSA2dist::cds2codonaln()</a></code> function can be overgiven to
e.g. use an alternative <code>substitutionMatrix</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example to use an alternative substitutionMatrix for the codon alignment</span></span>
<span><span class="co">## and obtain Ka/Ks</span></span>
<span></span>
<span><span class="co">## calculate Ka/Ks values on two CDS using Li model and BLOSUM45 cost matrix</span></span>
<span><span class="fu">MSA2dist</span><span class="fu">::</span><span class="fu"><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/dnastring2kaks.html" class="external-link">dnastring2kaks</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">cds1</span>,</span>
<span>    <span class="va">cds2</span><span class="op">)</span>,</span>
<span>    model<span class="op">=</span><span class="st">"Li"</span>,</span>
<span>    isMSA<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>    substitutionMatrix<span class="op">=</span><span class="st">"BLOSUM45"</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                               OUT</span></span>
<span><span class="co">#&gt; Comp1                                                                                                                                                                                                                                                                           1</span></span>
<span><span class="co">#&gt; Comp2                                                                                                                                                                                                                                                                           2</span></span>
<span><span class="co">#&gt; seq1                                         AT1G01010.1_cds_chromosome:TAIR10:1:3631:5899:1_gene:AT1G01010_gene_biotype:protein_coding_transcript_biotype:protein_coding_gene_symbol:NAC001_description:NAC_domain-containing_protein_1_[Source:UniProtKB/Swiss-Prot;Acc:Q0WV96]</span></span>
<span><span class="co">#&gt; seq2  Al_scaffold_0001_283_cds_chromosome:v.1.0:1:1082169:1084227:1_gene:Al_scaffold_0001_283_gene_biotype:protein_coding_transcript_biotype:protein_coding_description:DTW_domain-containing_protein_[Source:Projected_from_Arabidopsis_thaliana_(AT1G03687)_TAIR;Acc:AT1G03687]</span></span>
<span><span class="co">#&gt; ka                                                                                                                                                                                                                                                              0.878034285447893</span></span>
<span><span class="co">#&gt; ks                                                                                                                                                                                                                                                                       9.999999</span></span>
<span><span class="co">#&gt; vka                                                                                                                                                                                                                                                           0.00945680717088392</span></span>
<span><span class="co">#&gt; vks                                                                                                                                                                                                                                                                      9.999999</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="rbh2kaks-function">3.3. <span id="rbh2kaks">rbh2kaks() Function</span><a class="anchor" aria-label="anchor" href="#rbh2kaks-function"></a>
</h4>
<p>The resulting CRBHit pairs (see <a href="#cds2rbhfunction">cds2rbh()</a>) can be used with the
<code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function to obtain pairwise codon alignments,
which are further used to calculate synonymous (Ks) and nonsynonymous
(Ka) substitutions using parallelization.</p>
<p><strong>Note:</strong></p>
<p>It is important, that the names of the <code>rbh</code> columns must
match the names of the corresponding <code>cds1</code> and
<code>cds2</code> <code>DNAStringSet</code> vectors.</p>
<p>However, since one should directly use the same input
<code>DNAStringSet</code> vector or input <code>url</code> to calculate
the reciprocal best hit pair matrix with the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> or
the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function, this should not be a
problem.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## example how to get CRBHit pairs from two coding fasta files</span></span>
<span><span class="va">cdsfile1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span><span class="va">cdsfile2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"fasta"</span>, <span class="st">"aly.cds.fasta.gz"</span>, package<span class="op">=</span><span class="st">"CRBHits"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ath</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile1</span><span class="op">)</span></span>
<span><span class="va">aly</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">readDNAStringSet</a></span><span class="op">(</span><span class="va">cdsfile2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## the following function calculates CRBHit pairs</span></span>
<span><span class="co">## using one thread and plots the fitted curve</span></span>
<span><span class="va">ath_aly_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span></span>
<span>    cds1<span class="op">=</span><span class="va">ath</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">aly</span>,</span>
<span>    lastpath<span class="op">=</span><span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## calculate Ka/Ks using the CRBHit pairs</span></span>
<span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh</span><span class="op">$</span><span class="va">crbh.pairs</span><span class="op">)</span></span>
<span><span class="va">ath_aly_crbh.kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span></span>
<span>    rbhpairs<span class="op">=</span><span class="va">ath_aly_crbh</span>,</span>
<span>    cds1<span class="op">=</span><span class="va">ath</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">aly</span>,</span>
<span>    model<span class="op">=</span><span class="st">"Li"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh.kaks</span><span class="op">)</span></span>
<span><span class="co">#&gt;           aa1                   aa2 rbh_class         ka        ks          vka</span></span>
<span><span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh 0.74651201 9.9999990 0.0032172312</span></span>
<span><span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh 0.00000000 0.1831839 0.0000000000</span></span>
<span><span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh 0.05680851 0.1390746 0.0001612571</span></span>
<span><span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh 0.03656808 0.1543978 0.0001014973</span></span>
<span><span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh 0.75036088 9.9999990 0.0093948533</span></span>
<span><span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh 0.78228833 9.9999990 0.0051807370</span></span>
<span><span class="co">#&gt;           vks</span></span>
<span><span class="co">#&gt; 1 9.999999000</span></span>
<span><span class="co">#&gt; 2 0.003022307</span></span>
<span><span class="co">#&gt; 3 0.001531756</span></span>
<span><span class="co">#&gt; 4 0.001249826</span></span>
<span><span class="co">#&gt; 5 9.999999000</span></span>
<span><span class="co">#&gt; 6 9.999999000</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?rbh2kaks</span></span></code></pre></div>
<p>The Ka/Ks calculations can be parallelized.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate Ka/Ks using the CRBHit pairs and multiple threads</span></span>
<span><span class="va">ath_aly_crbh.kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span></span>
<span>    rbhpairs<span class="op">=</span><span class="va">ath_aly_crbh</span>,</span>
<span>    cds1<span class="op">=</span><span class="va">ath</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">aly</span>,</span>
<span>    model<span class="op">=</span><span class="st">"Li"</span>,</span>
<span>    threads<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh.kaks</span><span class="op">)</span></span>
<span><span class="co">#&gt;           aa1                   aa2 rbh_class         ka        ks          vka</span></span>
<span><span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh 0.74651201 9.9999990 0.0032172312</span></span>
<span><span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh 0.00000000 0.1831839 0.0000000000</span></span>
<span><span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh 0.05680851 0.1390746 0.0001612571</span></span>
<span><span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh 0.03656808 0.1543978 0.0001014973</span></span>
<span><span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh 0.75036088 9.9999990 0.0093948533</span></span>
<span><span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh 0.78228833 9.9999990 0.0051807370</span></span>
<span><span class="co">#&gt;           vks</span></span>
<span><span class="co">#&gt; 1 9.999999000</span></span>
<span><span class="co">#&gt; 2 0.003022307</span></span>
<span><span class="co">#&gt; 3 0.001531756</span></span>
<span><span class="co">#&gt; 4 0.001249826</span></span>
<span><span class="co">#&gt; 5 9.999999000</span></span>
<span><span class="co">#&gt; 6 9.999999000</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?rbh2kaks</span></span></code></pre></div>
<p>Like for the <code><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/dnastring2kaks.html" class="external-link">MSA2dist::dnastring2kaks()</a></code> function, all
parameters which can be used by the
<code><a href="https://mpievolbio-it.pages.gwdg.de/MSA2dist/reference/cds2codonaln.html" class="external-link">MSA2dist::cds2codonaln()</a></code> function can be overgiven to
e.g. use an alternative <code>substitutionMatrix</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## calculate Ka/Ks using the CRBHit pairs and multiple threads</span></span>
<span><span class="va">ath_aly_crbh.kaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span><span class="op">(</span></span>
<span>    rbhpairs<span class="op">=</span><span class="va">ath_aly_crbh</span>,</span>
<span>    cds1<span class="op">=</span><span class="va">ath</span>,</span>
<span>    cds2<span class="op">=</span><span class="va">aly</span>,</span>
<span>    model<span class="op">=</span><span class="st">"Li"</span>,</span>
<span>    threads<span class="op">=</span><span class="fl">2</span>,</span>
<span>    substitutionMatrix<span class="op">=</span><span class="st">"BLOSUM45"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ath_aly_crbh.kaks</span><span class="op">)</span></span>
<span><span class="co">#&gt;           aa1                   aa2 rbh_class         ka        ks          vka</span></span>
<span><span class="co">#&gt; 1 AT1G01040.1 Al_scaffold_0001_3256       rbh 0.69109360 9.9999990 0.0029376709</span></span>
<span><span class="co">#&gt; 2 AT1G01050.1  Al_scaffold_0001_128       rbh 0.00000000 0.1831839 0.0000000000</span></span>
<span><span class="co">#&gt; 3 AT1G01080.3  Al_scaffold_0001_125       rbh 0.05638742 0.1366968 0.0001613461</span></span>
<span><span class="co">#&gt; 4 AT1G01180.1  Al_scaffold_0001_114       rbh 0.03656808 0.1543978 0.0001014973</span></span>
<span><span class="co">#&gt; 5 AT1G01190.2 Al_scaffold_0001_4419       rbh 0.66169565 9.9999990 0.0073901048</span></span>
<span><span class="co">#&gt; 6 AT1G01260.3 Al_scaffold_0001_1326       rbh 0.75382109 9.9999990 0.0047473367</span></span>
<span><span class="co">#&gt;           vks</span></span>
<span><span class="co">#&gt; 1 9.999999000</span></span>
<span><span class="co">#&gt; 2 0.003022307</span></span>
<span><span class="co">#&gt; 3 0.001436133</span></span>
<span><span class="co">#&gt; 4 0.001249826</span></span>
<span><span class="co">#&gt; 5 9.999999000</span></span>
<span><span class="co">#&gt; 6 9.999999000</span></span>
<span></span>
<span><span class="co">## get help</span></span>
<span><span class="co">#?rbh2kaks</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="references">4. <span id="references">References</span><a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<p>Aubry S., Kelly S., Kümpers B. M., Smith-Unna R. D., and Hibberd J.
M. (2014). <strong>Deep evolutionary comparison of gene expression
identifies parallel recruitment of trans-factors in two independent
origins of C4 photosynthesis.</strong> <em>PLoS Genetics</em>,
<strong>10(6)</strong>. <a href="https://doi.org/10.1371/journal.pgen.1004365" class="external-link">https://doi.org/10.1371/journal.pgen.1004365</a></p>
<p>Charif D., and Lobry J. R. (2007). <strong>SeqinR 1.0-2: a
contributed package to the R project for statistical computing devoted
to biological sequences retrieval and analysis.</strong> <em>In
Structural approaches to sequence evolution</em> (pp. 207-232).
Springer, Berlin, Heidelberg. <a href="https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10" class="external-link">https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10</a></p>
<p>Duong T., and Wand M. (2015). <strong>feature: Local Inferential
Feature Significance for Multivariate Kernel Density
Estimation.</strong> <em>R package version 1.2.13</em>. <a href="https://cran.r-project.org/web/packages/feature/" class="external-link">https://cran.r-project.org/web/packages/feature/</a></p>
<p>Haas B. J., Delcher A. L., Wortman J. R., and Salzberg S. L. (2004).
<strong>DAGchainer: a tool for mining segmental genome duplications and
synteny.</strong> <em>Bioinformatics</em>, <strong>20(18)</strong>,
3643-3646. <a href="https://doi.org/10.1093/bioinformatics/bth397" class="external-link">https://doi.org/10.1093/bioinformatics/bth397</a></p>
<p>Haug-Baltzell A., Stephens S. A., Davey S., Scheidegger C. E., Lyons
E. (2017). <strong>SynMap2 and SynMap3D: web-based wholge-genome synteny
browsers.</strong> <em>Bioinformatics</em>, <strong>33(14)</strong>,
2197-2198. <a href="https://academic.oup.com/bioinformatics/article/33/14/2197/3072872" class="external-link">https://academic.oup.com/bioinformatics/article/33/14/2197/3072872</a></p>
<p>Kiełbasa S. M., Wan R., Sato K., Horton P., and Frith M. C. (2011).
<strong>Adaptive seeds tame genomic sequence comparison.</strong>
<em>Genome Research</em>, <strong>21(3)</strong>, 487-493. <a href="https://doi.org/10.1101/gr.113985.110" class="external-link">https://doi.org/10.1101/gr.113985.110</a></p>
<p>Kimura M. (1977). <strong>Preponderance of synonymous changes as
evidence for the neutral theory of molecular evolution.</strong>
<em>Nature</em>, <strong>267</strong>, 275-276.</p>
<p>Li W. H. (1993). <strong>Unbiased estimation of the rates of
synonymous and nonsynonymous substitution.</strong> <em>Journal of
Molecular Evolution</em>, <strong>36(1)</strong>, 96-99. <a href="https://doi.org/10.1007/bf02407308" class="external-link">https://doi.org/10.1007/bf02407308</a></p>
<p>Microsoft, and Weston S. (2020). <strong>foreach: Provides Foreach
Looping Construct.</strong> <em>R package version</em>, 1.5.1. <a href="https://cran.r-project.org/web/packages/foreach/index.html" class="external-link">foreach</a></p>
<p>Mugal C. F., Wolf J. B. W., Kaj I. (2014). <strong>Why Time Matters:
Codon Evolution and the Temproal Dynamics of dN/dS.</strong>
<em>Molecular Biology and Evolution</em>, <strong>31(1)</strong>,
212-231.</p>
<p>Ooms J. (2019). <strong>curl: A Modern and Flexible Web Client for
R.</strong> <em>R package version</em>, 4.3. <a href="https://cran.r-project.org/web/packages/curl/index.html" class="external-link">curl</a></p>
<p>Pagès H., Aboyoun P., Gentleman R., and DebRoy S. (2017).
<strong>Biostrings: Efficient manipulation of biological
strings.</strong> <em>R package version</em>, 2.56.0. <a href="http://bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link">Biostrings</a></p>
<p>Revolution Analytics, and Weston S. (2020). <strong>doMC: Foreach
Parallel Adaptor for ‘parallel’.</strong> <em>R package version</em>,
1.3.7. <a href="https://cran.r-project.org/web/packages/doMC/index.html" class="external-link">doMC</a></p>
<p>Rost B. (1999). <strong>Twilight zone of protein sequence
alignments.</strong> <em>Protein Engineering</em>,
<strong>12(2)</strong>, 85-94. <a href="https://doi.org/10.1093/protein/12.2.85" class="external-link">https://doi.org/10.1093/protein/12.2.85</a></p>
<p>Scott C. (2017). <strong>shmlast: an improved implementation of
conditional reciprocal best hits with LAST and Python.</strong>
<em>Journal of Open Source Software</em>, <strong>2(9)</strong>, 142. <a href="https://joss.theoj.org/papers/10.21105/joss.00142" class="external-link">https://joss.theoj.org/papers/10.21105/joss.00142</a></p>
<p>Scrucca L., Fop M., Murphy T. B., and Raftery A. E. (2016)
<strong>mclust 5: clustering, classification and density estimation
using Gaussian finite mixture models.</strong> <em>The R Journal</em>,
8(1), 289-317. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/" class="external-link">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/</a></p>
<p>Wickham H. (2011). <strong>testthat: Get Started with
Testing.</strong> <em>The R Journal</em>, 3(1), 5. <a href="https://cran.r-project.org/web/packages/testthat/index.html" class="external-link">testthat</a></p>
<p>Wickham H. (2019). <strong>stringr: Simple, Consistent Wrappers for
Common String Operations.</strong> <em>R package version</em>, 1.4.0. <a href="https://cran.r-project.org/web/packages/stringr/index.html" class="external-link">stringr</a></p>
<p>Wickham H. (2020). <strong>tidyr: Tidy Messy Data.</strong> <em>R
package version</em>, 1.1.2. <a href="https://cran.r-project.org/web/packages/tidyr/index.html" class="external-link">tidyr</a></p>
<p>Wickham H., Hester J., and Chang W. (2020). <strong>devtools: Tools
to make Developing R Packages Easier.</strong> <em>R package
version</em>, (2.3.2). <a href="https://cran.r-project.org/web/packages/devtools/index.html" class="external-link">devtools</a></p>
<p>Wickham H., François R., Henry L., and Müller K. (2020).
<strong>dplyr: A Grammar of Data Manipulation.</strong> <em>R package
version</em>, 1.0.2. <a href="https://cran.r-project.org/web/packages/dplyr/index.html" class="external-link">dplyr</a></p>
<p>Yang Z., and Nielsen R. (2000). <strong>Estimating synonymous and
nonsynonymous substitution rates under realistic evolutionary
models.</strong> <em>Molecular Biology and Evolution</em>,
<strong>17(1)</strong>, 32-43. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a026236" class="external-link">https://doi.org/10.1093/oxfordjournals.molbev.a026236</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
