<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CRBHits Basic Vignette • CRBHits</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CRBHits Basic Vignette">
<meta property="og:description" content="CRBHits">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/V02dNdSClusteringVignette.html">dNdS Clustering Vignette</a>
    </li>
    <li>
      <a href="../articles/V03WGDVignette.html">Paranome-based Whole Genome Duplication Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="V01CRBHitsBasicVignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CRBHits Basic Vignette</h1>
                        <h4 class="author">Kristian K Ullrich</h4>
            
            <h4 class="date">2020-08-25</h4>
      
      <small class="dont-index">Source: <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/mastervignettes/V01CRBHitsBasicVignette.Rmd"><code>vignettes/V01CRBHitsBasicVignette.Rmd</code></a></small>
      <div class="hidden name"><code>V01CRBHitsBasicVignette.Rmd</code></div>

    </div>

    
    
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#install">Installation</a></li>
<li>
<a href="#crbh">Conditional Reciprocal Best Hits</a>
<ol style="list-style-type: decimal">
<li><a href="#crbhalgorithm">Conditional Reciprocal Best Hits - Algorithm</a></li>
<li><a href="#input">Coding Sequences Input</a></li>
<li><a href="#cds2rbhfunction">cds2rbh() Function</a></li>
<li><a href="#filters">Filter blast-like output prior fitting</a></li>
<li><a href="#customfilters">Custom Filter</a></li>
<li><a href="#fittingparam">Fitting Parameter</a></li>
</ol>
</li>
<li>
<a href="#dNdS">dN/dS Calculations</a>
<ol style="list-style-type: decimal">
<li><a href="#cds2codon">Codon Alignments - cds2codonaln() Function</a></li>
<li><a href="#cds2kaks">cds2kaks() Function</a></li>
<li><a href="#rbh2kaks">rbh2kaks() Function</a></li>
</ol>
</li>
<li><a href="#references">References</a></li>
</ol>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is a reimplementation of the Conditional Reciprocal Best Hit algorithm <a href="https://github.com/cboursnell/crb-blast">crb-blast</a> in R.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>1. <span id="install">Installation</span>
</h2>
<p>See the R package page for a detailed description of the install process and its dependencies <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/">https://mpievolbio-it.pages.gwdg.de/crbhits/</a>.</p>
<pre><code><a href="https://devtools.r-lib.org//reference">#see a detailed description of installation prerequisites at
#https://mpievolbio-it.pages.gwdg.de/crbhits/
library(devtools)
#from gitlab
install_gitlab("mpievolbio-it/crbhits", host = "https://gitlab.gwdg.de",
               build_vignettes = TRUE, dependencies = FALSE)
#from github
install_github("kullrich/CRBHits", build_vignettes = TRUE, dependencies = FALSE)</a></code></pre>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> needs <a href="http://last.cbrc.jp/">LAST</a> and <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> to be installed before one can efficiently use it.</p>
<p>Both prerequisites (LAST, KaKs_Calculator2.0) are forked within <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> and can be compiled for Linux/Unix/macOS with the functions <code><a href="../reference/make.last.html">make.last()</a></code> and <code><a href="../reference/make.KaKs_Calculator2.html">make.KaKs_Calculator2()</a></code>.</p>
<p>For the windows platform the user needs to use e.g. <a href="https://www.cygwin.com/">https://www.cygwin.com/</a> to be able to compile both prerequisites.</p>
<pre><code><a href="https://rdrr.io/pkg/CRBHits/man">library(CRBHits)
#compile last-1080
make.last()
#compile KaKs_Calculator2.0
make.KaKs_Calculator2()</a></code></pre>
<p><strong>Note:</strong></p>
<p>In this Vignette the <code>lastpath</code> is defined as <code>vignette.paths[1]</code> and the <code>kakscalcpath</code> as <code>vignette.paths[2]</code> to be able to build the Vignette.</p>
<p>However, once you have compiled last-1080 and KaKs_Calculator2.0 with the functions <code><a href="../reference/make.last.html">make.last()</a></code> and <code><a href="../reference/make.KaKs_Calculator2.html">make.KaKs_Calculator2()</a></code> you won’t need to specify the paths anymore. Please remove them if you would like to repeat the examples.</p>
</div>
<div id="conditional-reciprocal-best-hits" class="section level2">
<h2 class="hasAnchor">
<a href="#conditional-reciprocal-best-hits" class="anchor"></a>2. <span id="crbh">Conditional Reciprocal Best Hits</span>
</h2>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is a reimplementation of the Conditional Reciprocal Best Hit algorithm <a href="https://github.com/cboursnell/crb-blast">crb-blast</a> in R.</p>
<div id="conditional-reciprocal-best-hits---algorithm" class="section level3">
<h3 class="hasAnchor">
<a href="#conditional-reciprocal-best-hits---algorithm" class="anchor"></a>2.1 <span id="crbhalgorithm">Conditional Reciprocal Best Hits - Algorithm</span>
</h3>
<p>The algorithm was introduced by <a href="http://www.plosgenetics.org/article/info%3Adoi%2F10.1371%2Fjournal.pgen.1004365">Aubry S, Kelly S et al. (2014)</a> and ported to python <a href="https://pypi.org/project/shmlast/">shmlast</a> (<a href="https://joss.theoj.org/papers/10.21105/joss.00142">Scott C. 2017</a>) which benefits from the blast-like sequence search software <a href="http://last.cbrc.jp/">LAST</a> (<a href="https://genome.cshlp.org/content/21/3/487.long">Kiełbasa SM et al. 2011</a>).</p>
<p>The CRBH algorithm builds upon the classical reciprocal best hit (RBH) approach to find orthologous sequences between two sets of sequences by defining an expect-value cutoff per alignment length. Further, this cutoff is used to retain secondary hits as additional bona-fide homologs.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> only takes coding nucleotide sequences as the query and target inputs.</p>
<p><strong>Why?</strong> Because <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is implemented to use the same input afterwards to calculate synonymous and nonsynonymous substitutions which must match the protein sequences used during the conditional reciprocal best hit step.</p>
</div>
<div id="coding-sequences-input" class="section level3">
<h3 class="hasAnchor">
<a href="#coding-sequences-input" class="anchor"></a>2.2 <span id="input">Coding Sequences Input</span>
</h3>
<p>One can either use <strong>raw</strong> or <strong>gzipped</strong> coding sequences <strong>fasta</strong> files as inputs with the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function or use a <code>DNAStringSet</code> vector from the <code>Biostrings</code> R package with the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function.</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> uses either the function <code><a href="../reference/cdsfile2aafile.html">cdsfile2aafile()</a></code> to translate a coding sequence fasta file into its corresponding protein sequence fasta file or <code><a href="../reference/cds2aa.html">cds2aa()</a></code> to translate a <code>DNAStringSet</code> into a <code>AAStringSet</code>. Thereby it removes all sequences that are not a multiple of three which can not be parsed corecctly.</p>
<p><strong>Note:</strong> The user can easily check if all coding sequences would be a multiple of three or rely on the input files generated by other sources.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">CRBHits</span>)
<span class="co">#compile LAST and KaKs_Calculator2.0 for the vignette</span>
<span class="no">vignette.paths</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/make.vignette.html">make.vignette</a></span>()

<span class="co">#example how to check coding sequences if all are a mutiple of three</span>
<span class="no">cdsfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CRBHits"</span>)
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span>(<span class="no">cdsfile</span>)

<span class="co">#the following statement should return TRUE, if all sequences are a mutiple of three</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span>(<span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu">width</span>(<span class="no">cds</span>) <span class="kw">%%</span> <span class="fl">3</span> <span class="kw">==</span> <span class="fl">0</span>)
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>The blast-like software <a href="http://last.cbrc.jp/">LAST</a> is used to compare the translated coding sequences against each other and output a blast-like output table including the query and target length.</p>
</div>
<div id="cds2rbh-function" class="section level3">
<h3 class="hasAnchor">
<a href="#cds2rbh-function" class="anchor"></a>2.3 <span id="cds2rbhfunction">cds2rbh() Function</span>
</h3>
<p>Like <a href="https://pypi.org/project/shmlast/">shmlast</a>, <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> plots the fitted model of the conditional reciprocal best hit evalue based algorithm.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">#example how to get crbh from two coding fasta files</span>
<span class="no">cdsfile1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CRBHits"</span>)
<span class="no">cdsfile2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"fasta"</span>, <span class="st">"aly.cds.fasta.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CRBHits"</span>)

<span class="no">cds1</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span>(<span class="no">cdsfile1</span>)
<span class="no">cds2</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span>(<span class="no">cdsfile2</span>)

<span class="co">#the following function calculates crbh matrix using one thread and plots the fitted curve</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">plotCurve</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">ath_aly_crbh</span>)
<span class="co">#&gt;               Length Class      Mode    </span>
<span class="co">#&gt; crbh.pairs     2     data.frame list    </span>
<span class="co">#&gt; crbh1         16     data.frame list    </span>
<span class="co">#&gt; crbh2         16     data.frame list    </span>
<span class="co">#&gt; rbh1_rbh2_fit  1     -none-     function</span></pre></body></html></div>
<p>Classical RBH can be performed by disabeling the <code>crbh</code> <code>@param</code>of the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> or <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co">#example how to perform classical rbh</span>
<span class="no">ath_aly_rbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">crbh</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
                       <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p>Both, <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> and <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function, return a list of three <code>data.frame</code>’s which contain the conditional reciprocal best hit pairs (<code>$crbh.pairs</code>) retained, the query &gt; target blast-like output matrix (<code>$crbh1</code>) and the target &gt; query blast-like output matrix (<code>$crbh2</code>). For each matrix the <code>$rbh_class</code> indicates if the hit is a reciprocal best hit (<code>rbh</code>) or if it is a secondary hit retained because of crbh (<code>sec</code>).</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co">#show dimension and first retained hit pairs</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>)
<span class="co">#&gt; [1] 798   2</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>)
<span class="co">#&gt;           aa1          aa2</span>
<span class="co">#&gt; 1 AT1G01010.1 AL1G11530.t1</span>
<span class="co">#&gt; 2 AT1G01020.5 AL1G11510.t1</span>
<span class="co">#&gt; 3 AT1G01030.1 AL1G11500.t1</span>
<span class="co">#&gt; 4 AT1G01040.1 AL1G11480.t1</span>
<span class="co">#&gt; 5 AT1G01050.1 AL1G11470.t1</span>
<span class="co">#&gt; 6 AT1G01060.6 AL1G11460.t1</span>

<span class="co">#show first retained hit pairs for the query &gt; target matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh1</span>)
<span class="co">#&gt;      query_id   subject_id perc_identity alignment_length mismatches gap_opens</span>
<span class="co">#&gt; 1 AT1G01010.1 AL1G11530.t1         71.61              472         86         8</span>
<span class="co">#&gt; 2 AT1G01020.5 AL1G11510.t1         88.94              199         22         0</span>
<span class="co">#&gt; 3 AT1G01030.1 AL1G11500.t1         96.31              298          8         1</span>
<span class="co">#&gt; 4 AT1G01040.1 AL1G11480.t1         96.39             1909         58         6</span>
<span class="co">#&gt; 5 AT1G01050.1 AL1G11470.t1        100.00              213          0         0</span>
<span class="co">#&gt; 6 AT1G01060.6 AL1G11460.t1         86.51              593         63         7</span>
<span class="co">#&gt;   q_start q_end s_start s_end   evalue bit_score query_length subject_length</span>
<span class="co">#&gt; 1       1   430       1   466 7.7e-224       696          430            466</span>
<span class="co">#&gt; 2       1   199      48   246 2.6e-124       389          199            246</span>
<span class="co">#&gt; 3       1   298       1   295 9.1e-221       680          359            323</span>
<span class="co">#&gt; 4       6  1910       4  1905  0.0e+00      4180         1910           1905</span>
<span class="co">#&gt; 5       1   213       1   213 2.0e-160       493          213            213</span>
<span class="co">#&gt; 6      60   644      62   645  0.0e+00      1130          644            645</span>
<span class="co">#&gt;   raw_score rbh_class</span>
<span class="co">#&gt; 1      1570       rbh</span>
<span class="co">#&gt; 2       874       rbh</span>
<span class="co">#&gt; 3      1534       rbh</span>
<span class="co">#&gt; 4      9471       rbh</span>
<span class="co">#&gt; 5      1110       rbh</span>
<span class="co">#&gt; 6      2557       rbh</span>

<span class="co">#get the number of rbh and sec hit pairs</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh1</span>$<span class="no">rbh_class</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; rbh sec </span>
<span class="co">#&gt; 745  53</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh2</span>$<span class="no">rbh_class</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; rbh sec </span>
<span class="co">#&gt; 745  53</span></pre></body></html></div>
<p>If the <code>@param</code> <code>crbh</code> was set to <code>TRUE</code>, the conditional reciprocal best hit algorithm was applied and the fitting function will also be returned. The fitting function can than be used for manual plotting.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co">#example how to use the fitting function for manual plotting</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span>(<span class="no">ath_aly_crbh</span>$<span class="fu">rbh1_rbh2_fit</span>(<span class="no">x</span>),
      <span class="kw">from</span> <span class="kw">=</span> <span class="fl">1</span>,
      <span class="kw">to</span> <span class="kw">=</span> <span class="fl">1000</span>,
      <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">"alnlength"</span>,
      <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"-log10(evalue)"</span>,
      <span class="kw">main</span> <span class="kw">=</span> <span class="st">"CRBH fitting"</span>)</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>One can also specify if a hit pair which is only found in one direction should be retained, which will be classified as (<code>single</code>).</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co">#example how to retain single direction secondary homologs</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">plotCurve</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">keepSingleDirection</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh1</span>$<span class="no">rbh_class</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;    rbh    sec single </span>
<span class="co">#&gt;    745     53      2</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh2</span>$<span class="no">rbh_class</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;    rbh    sec single </span>
<span class="co">#&gt;    745     53      5</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">ath_aly_crbh</span>)
<span class="co">#&gt; NULL</span></pre></body></html></div>
</div>
<div id="filter-blast-like-output-prior-fitting" class="section level3">
<h3 class="hasAnchor">
<a href="#filter-blast-like-output-prior-fitting" class="anchor"></a>2.4 <span id="filters">Filter blast-like output prior fitting</span>
</h3>
<p>In addition users can filter the hit pairs prior fitting for other cirteria like evalue, protein identity and/or the twilight zone of protein sequence alignments according to <a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a>.</p>
<p>The follwoing filters are already defined can be used out of the box:</p>
<ul>
<li>evalue (<code>@param</code> <code>evalue = 0.001</code>)</li>
<li>query coverage (<code>@param</code> <code>qcov = 0.5</code>)</li>
<li>target coverage (<code>@param</code> <code>tcov = 0.5</code>)</li>
<li>protein identity (<code>@param</code> <code>pident = 50</code>)</li>
<li>alignment length (<code>@param</code> <code>alnlen = 50</code>)</li>
<li>rost1999 (<code>@param</code> <code>rost1999 = TRUE</code>)</li>
</ul>
<p>Taking the length of the obtained pairwise protein alignment one can calculate for each hit pair the query coverage as <span class="math inline">\(\frac{alignment length}{query length}\)</span>.</p>
<p>The data will be filtered with a query coverage of 50% (<code><a href="../reference/cds2rbh.html">cds2rbh(..., qcov = 0.5)</a></code>).</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">#example how to filter prior crbh for query coverage</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">plotCurve</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">qcov</span> <span class="kw">=</span> <span class="fl">0.5</span>,
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>)
<span class="co">#&gt; [1] 776   2</span></pre></body></html></div>
<p>The implemented <code>rost1999</code> filter uses <code>equation2</code> of <a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a></p>
<p><span class="math display">\[f(x_{\text{hit pair}}) = \cases {100 \text{ , for } L_{\text{hit pair}} &lt; 11 \\ 480 * L^{-0.32 * (1 + e^{\frac{-L}{1000}})} \text{ , for } L_{\text{hit pair}} \leq 450 \\ 19.5 \text{ , for } L_{\text{hit pair}} &gt; 450}\]</span> , where <span class="math inline">\(x_{\text{hit pair}}\)</span> is the expected protein identity given the alignemnet length <span class="math inline">\(L_{\text{hit pair}}\)</span>. If the actual <span class="math inline">\(pident_{\text{hit pair}} &gt;= f(x_{\text{hit pair}})\)</span> the hit pair is retained.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co">#define eq2 from Rost (1999)</span>
<span class="no">get_pident_by_length</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">x</span>){
  <span class="no">eq2</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">L</span>){
    <span class="kw">if</span>(<span class="no">L</span> <span class="kw">&lt;=</span> <span class="fl">11</span>){<span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fl">100</span>)}
    <span class="kw">if</span>(<span class="no">L</span> <span class="kw">&lt;=</span> <span class="fl">450</span>){<span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fl">480</span>*(<span class="no">L</span>^(-<span class="fl">0.32</span>*(<span class="fl">1</span>+(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(-<span class="no">L</span>/<span class="fl">1000</span>))))))}
    <span class="kw">if</span>(<span class="no">L</span> <span class="kw">&gt;</span> <span class="fl">450</span>){<span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fl">19.5</span>)}
  }
  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">x</span>, <span class="no">eq2</span>)))
}

<span class="co">#plot expected pident by alignment length using eq2 from Rost (1999)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span>(<span class="no">get_pident_by_length</span>, <span class="fl">11</span>, <span class="fl">500</span>, <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">xlab</span> <span class="kw">=</span> <span class="st">"alignment length"</span>, <span class="kw">ylab</span> <span class="kw">=</span> <span class="st">"pident"</span>,
      <span class="kw">main</span> <span class="kw">=</span> <span class="st">"expected protein identity (eq2; Rost B. 1999)"</span>)</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>This can be directly used by the <code>@param</code> <code>rost1999</code> set to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co">#example how to filter prior crbh for eq2 from Rost (1999)</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">plotCurve</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">rost1999</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>)
<span class="co">#&gt; [1] 797   2</span></pre></body></html></div>
</div>
<div id="custom-filter" class="section level3">
<h3 class="hasAnchor">
<a href="#custom-filter" class="anchor"></a>2.5 <span id="filters">Custom Filter</span>
</h3>
<p>The user can also define its own function to filter the blast-like output prior crbh.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co">#example for a custom filter for e.g. bit score (column 12)</span>
<span class="no">myfilter</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">rbh</span>, <span class="no">value</span> <span class="kw">=</span> <span class="fl">500.0</span>){
  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">rbh</span>[<span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">rbh</span>[, <span class="fl">12</span>]) <span class="kw">&gt;=</span> <span class="no">value</span> , , <span class="kw">drop</span> <span class="kw">=</span> <span class="fl">FALSE</span>])
}

<span class="co">#example how to filter prior crbh with a custom filter</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">plotCurve</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">filter</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">myfilter</span>),
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>)
<span class="co">#&gt; [1] 570   2</span></pre></body></html></div>
</div>
<div id="fitting-parameter" class="section level3">
<h3 class="hasAnchor">
<a href="#fitting-parameter" class="anchor"></a>2.6. <span id="fittingparam">Fitting Parameter</span>
</h3>
<p>The original implementation of the Conditional Reciprocal Best Hit algorithm <a href="https://github.com/cboursnell/crb-blast/blob/master/lib/crb-blast/crb-blast.rb">crb-blast.rb line 438-518</a> uses certain parameters for fitting the evalue against the alignment length.</p>
<p>In brief, for each retained reciprocal hit pair (RBH; after filtering) the evalue and the corresponding alignment length of a hit pair and its neighborhood is evaluated. The neighborhood is defined by a weighting parameter of the alignment length (default is set to <code>@param</code> <code>fit.varweight = 0.1</code>) and a minimum neighborhood alignment length (default is set to <code>@param</code> <code>fit.min = 5</code>).</p>
<p>For each unique alignment length <span class="math inline">\(L\)</span> a value <span class="math inline">\(s_{L}\)</span> is defined as the neighborhood alignment length as</p>
<p><span class="math display">\[s_{L} = \cases {\text{fit.min} \text{ , for } L * \text{fit.varweight} &lt; \text{fit.min} \\ L * \text{fit.varweight} \text{ , for } L * \text{fit.varweight} \geq \text{fit.min}}\]</span> E.g. for an alignment length of 100 and the default settings, <span class="math inline">\(s_{100} = 100 * 0.1 = 10\)</span>.</p>
<p>All RBHs, for which <span class="math inline">\(L-s_{L} \leq L_{RBH_i} \leq L + s_{L}\)</span>, in other words, for which their alignment lengths fall between 90 and 110 are than considered for fitting a cutoff at that alignment length using the mean of the <span class="math inline">\(-log10(evalue)\)</span>.</p>
<p><span class="math display">\[\text{cutoff}_{L} = \frac{1}{n} \sum_{i = 1}^{n}{\cases{324 \text{ , for }  evalue_{RBH_i} = 0 \\ -log10(evalue_{\text{RBH}_i}) \text{ , for } evalue_{RBH_i} \ne 0}}\]</span></p>
<p>The user can now alter between <code>mean</code> and <code>median</code> using the <code>@param</code> <code>fit.type = "mean"</code> or <code>fit.type = "median"</code> to get alternative fitting curves.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co">#example how to get crbh from two coding fasta files using median fitting</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">plotCurve</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                        <span class="kw">fit.type</span> <span class="kw">=</span> <span class="st">"median"</span>,
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])</pre></body></html></div>
<p><img src="V01CRBHitsBasicVignette_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
</div>
<div id="dnds-calculations" class="section level2">
<h2 class="hasAnchor">
<a href="#dnds-calculations" class="anchor"></a>3. <span id="dNdS">dN/dS Calculations</span>
</h2>
<div id="codon-alignments---cds2codonaln-function" class="section level3">
<h3 class="hasAnchor">
<a href="#codon-alignments---cds2codonaln-function" class="anchor"></a>3.1. <span id="cds2codon">Codon Alignments - cds2codonaln() Function</span>
</h3>
<p>To be able to calculate synonymous and non-synonymous substitutions, one needs to use a codon alignment.</p>
<p>The <code><a href="../reference/cds2codonaln.html">cds2codonaln()</a></code> function takes two single nucleotide sequences as input to obtain such a codon alignment. The function will convert the nucleotide sequences into amino acid sequences, align them with the help of the <code>pairwiseAlignment()</code> function from the <code>Biostrings</code> R package and convert this alignment back into a codon alignment.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="co">#example to get a codon alignment</span>
<span class="no">cds1</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html">DNAString</a></span>(<span class="st">"ATGCAACATTGC"</span>)
<span class="no">cds2</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/DNAString-class.html">DNAString</a></span>(<span class="st">"ATGCATTGC"</span>)
<span class="fu"><a href="../reference/cds2codonaln.html">cds2codonaln</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>)
<span class="co">#&gt; DNAStringSet object of length 2:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]    12 ATGCAACATTGC                                      cds1</span>
<span class="co">#&gt; [2]    12 ATG---CATTGC                                      cds2</span></pre></body></html></div>
<p>The user can alter some <code>@param</code> to change the alignment parameters, like <code>type</code>, <code>substitutionMatrix</code>, <code>gapOpening</code> and <code>gapExtension</code> costs as defined by the <code>pairwiseAlignment()</code> function from the <code>Biostrings</code> R package.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">#example to alter the substitionMatrix and use the BLOSUM45 cost matrix</span>
<span class="fu"><a href="../reference/cds2codonaln.html">cds2codonaln</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">substitutionMatrix</span> <span class="kw">=</span> <span class="st">"BLOSUM45"</span>)
<span class="co">#&gt; DNAStringSet object of length 2:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]    12 ATGCAACATTGC                                      cds1</span>
<span class="co">#&gt; [2]    12 ATG---CATTGC                                      cds2</span></pre></body></html></div>
<p>The user can also remove gaps in the codon alignment.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="co">#example to remove codon gaps</span>
<span class="fu"><a href="../reference/cds2codonaln.html">cds2codonaln</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">remove.gaps</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#&gt; DNAStringSet object of length 2:</span>
<span class="co">#&gt;     width seq                                               names               </span>
<span class="co">#&gt; [1]     9 ATGCATTGC                                         cds1</span>
<span class="co">#&gt; [2]     9 ATGCATTGC                                         cds2</span></pre></body></html></div>
</div>
<div id="cds2kaks-function" class="section level3">
<h3 class="hasAnchor">
<a href="#cds2kaks-function" class="anchor"></a>3.2. <span id="cds2kaks">cds2kaks() Function</span>
</h3>
<p>The ka/ks (dN/dS) values can be obtained either via the codon model of <a href="https://www.ncbi.nlm.nih.gov/pubmed/8433381">Li WH. (1999)</a> implemented in the R package <a href="https://cran.r-project.org/web/packages/seqinr/index.html">seqinr</a> or the model <a href="https://www.ncbi.nlm.nih.gov/pubmed/10666704">Yang Z and Nielson R. (2000)</a> implemented in <a href="https://sourceforge.net/projects/kakscalculator2/files/KaKs_Calculator2.0.tar.gz/download">KaKs_Calculator2.0</a> (see <a href="#quickinstall">Quick Installation</a> how to compile the prerequisites).</p>
<p>The <code><a href="../reference/cds2kaks.html">cds2kaks()</a></code> function takes two single nucleotide sequences as input to obtain a codon alignment and calculate dN/dS.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="co">##load example sequence data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"ath"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"CRBHits"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"aly"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"CRBHits"</span>)
<span class="co">##select a sequence pair according to a best hit pair (done for you)</span>
<span class="no">cds1</span> <span class="kw">&lt;-</span> <span class="no">ath</span>[<span class="fl">1</span>]
<span class="no">cds2</span> <span class="kw">&lt;-</span> <span class="no">aly</span>[<span class="fl">282</span>]
<span class="fu"><a href="../reference/cds2kaks.html">cds2kaks</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">model</span> <span class="kw">=</span> <span class="st">"Li"</span>)
<span class="co">#&gt;           ka           ks          vka          vks </span>
<span class="co">#&gt; 0.1061722655 0.2628911134 0.0002637032 0.0023637311</span></pre></body></html></div>
<p>All parameters which can be used by the <code><a href="../reference/cds2codonaln.html">cds2codonaln()</a></code> function can be overgiven to e.g. use an alternative <code>substitutionMatrix</code>.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co">#example to use an alternative substitutionMatrix</span>
<span class="fu"><a href="../reference/cds2kaks.html">cds2kaks</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">model</span> <span class="kw">=</span> <span class="st">"Li"</span>, <span class="kw">substitutionMatrix</span> <span class="kw">=</span> <span class="st">"BLOSUM45"</span>)
<span class="co">#&gt;           ka           ks          vka          vks </span>
<span class="co">#&gt; 0.1070171100 0.2719592468 0.0002625809 0.0024722288</span></pre></body></html></div>
</div>
<div id="rbh2kaks-function" class="section level3">
<h3 class="hasAnchor">
<a href="#rbh2kaks-function" class="anchor"></a>3.3. <span id="rbh2kaks">rbh2kaks() Function</span>
</h3>
<p>The resultings conditional reciprocal best hit pair matrix (see <a href="#cds2rbhfunction">cds2rbh()</a>) can be used with the <code><a href="../reference/rbh2kaks.html">rbh2kaks()</a></code> function to obtain pairwise codon alignments, which are further used to calculate synonymous and nonsynonymous substitutions using parallelization.</p>
<p><strong>Note:</strong></p>
<p>It is important, that the names of the <code>rbh</code> columns must match the names of the corresponding <code>cds1</code> and <code>cds2</code> <code>DNAStringSet</code> vectors.</p>
<p>However, since one should directly use the same input <code>DNAStringSet</code> vector or input <code>url</code> to calculate the reciprocal best hit pair matrix with the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> or the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function, this should not be a problem.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="co">#example how to get crbh from two coding fasta files</span>
<span class="no">cdsfile1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"fasta"</span>, <span class="st">"ath.cds.fasta.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CRBHits"</span>)
<span class="no">cdsfile2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"fasta"</span>, <span class="st">"aly.cds.fasta.gz"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"CRBHits"</span>)

<span class="no">cds1</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span>(<span class="no">cdsfile1</span>)
<span class="no">cds2</span> <span class="kw">&lt;-</span> <span class="kw pkg">Biostrings</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span>(<span class="no">cdsfile2</span>)

<span class="co">#the following function calculates crbh matrix using one thread and plots the fitted curve</span>
<span class="no">ath_aly_crbh</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span>(<span class="no">cds1</span>, <span class="no">cds2</span>,
                        <span class="kw">lastpath</span> <span class="kw">=</span> <span class="no">vignette.paths</span>[<span class="fl">1</span>])

<span class="co">#calculate dN/dS usign the crbh matrix</span>
<span class="no">ath_aly_crbh.kaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>[<span class="fl">1</span>:<span class="fl">20</span>, ],
                              <span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">model</span> <span class="kw">=</span> <span class="st">"Li"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ath_aly_crbh.kaks</span>)
<span class="co">#&gt;           aa1          aa2         ka        ks          vka          vks</span>
<span class="co">#&gt; 1 AT1G01010.1 AL1G11530.t1 0.10617227 0.2628911 2.637032e-04 0.0023637311</span>
<span class="co">#&gt; 2 AT1G01020.5 AL1G11510.t1 0.05027159 0.1076133 2.572181e-04 0.0017464252</span>
<span class="co">#&gt; 3 AT1G01030.1 AL1G11500.t1 0.05606632 0.2329670 1.808118e-04 0.0026270370</span>
<span class="co">#&gt; 4 AT1G01040.1 AL1G11480.t1 0.01388844 0.1121138 6.907144e-06 0.0001460956</span>
<span class="co">#&gt; 5 AT1G01050.1 AL1G11470.t1 0.00000000 0.1831839 0.000000e+00 0.0030223075</span>
<span class="co">#&gt; 6 AT1G01060.6 AL1G11460.t1 0.09054187 0.1854908 1.323098e-04 0.0007490874</span></pre></body></html></div>
<p>The dN/dS calculations can be parallelized.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co">#calculate dN/dS usign the crbh matrix and multiple threads</span>
<span class="no">ath_aly_crbh.kaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>[<span class="fl">1</span>:<span class="fl">20</span>, ],
                              <span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">model</span> <span class="kw">=</span> <span class="st">"Li"</span>, <span class="kw">threads</span> <span class="kw">=</span> <span class="fl">2</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ath_aly_crbh.kaks</span>)
<span class="co">#&gt;           aa1          aa2         ka        ks          vka          vks</span>
<span class="co">#&gt; 1 AT1G01010.1 AL1G11530.t1 0.10617227 0.2628911 2.637032e-04 0.0023637311</span>
<span class="co">#&gt; 2 AT1G01020.5 AL1G11510.t1 0.05027159 0.1076133 2.572181e-04 0.0017464252</span>
<span class="co">#&gt; 3 AT1G01030.1 AL1G11500.t1 0.05606632 0.2329670 1.808118e-04 0.0026270370</span>
<span class="co">#&gt; 4 AT1G01040.1 AL1G11480.t1 0.01388844 0.1121138 6.907144e-06 0.0001460956</span>
<span class="co">#&gt; 5 AT1G01050.1 AL1G11470.t1 0.00000000 0.1831839 0.000000e+00 0.0030223075</span>
<span class="co">#&gt; 6 AT1G01060.6 AL1G11460.t1 0.09054187 0.1854908 1.323098e-04 0.0007490874</span></pre></body></html></div>
<p>Like for the <code><a href="../reference/cds2kaks.html">cds2kaks()</a></code> function, all parameters which can be used by the <code><a href="../reference/cds2codonaln.html">cds2codonaln()</a></code> function can be overgiven to e.g. use an alternative <code>substitutionMatrix</code>.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="co">#calculate dN/dS usign the crbh matrix and multiple threads</span>
<span class="no">ath_aly_crbh.kaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/rbh2kaks.html">rbh2kaks</a></span>(<span class="no">ath_aly_crbh</span>$<span class="no">crbh.pairs</span>[<span class="fl">1</span>:<span class="fl">20</span>, ],
                              <span class="no">cds1</span>, <span class="no">cds2</span>, <span class="kw">model</span> <span class="kw">=</span> <span class="st">"Li"</span>, <span class="kw">threads</span> <span class="kw">=</span> <span class="fl">2</span>,
                              <span class="kw">substitutionMatrix</span> <span class="kw">=</span> <span class="st">"BLOSUM45"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ath_aly_crbh.kaks</span>)
<span class="co">#&gt;           aa1          aa2         ka        ks          vka          vks</span>
<span class="co">#&gt; 1 AT1G01010.1 AL1G11530.t1 0.10701711 0.2719592 2.625809e-04 0.0024722288</span>
<span class="co">#&gt; 2 AT1G01020.5 AL1G11510.t1 0.05027159 0.1076133 2.572181e-04 0.0017464252</span>
<span class="co">#&gt; 3 AT1G01030.1 AL1G11500.t1 0.05537564 0.2580275 1.783063e-04 0.0029488815</span>
<span class="co">#&gt; 4 AT1G01040.1 AL1G11480.t1 0.01388844 0.1121138 6.907144e-06 0.0001460956</span>
<span class="co">#&gt; 5 AT1G01050.1 AL1G11470.t1 0.00000000 0.1831839 0.000000e+00 0.0030223075</span>
<span class="co">#&gt; 6 AT1G01060.6 AL1G11460.t1 0.09140582 0.1860964 1.322842e-04 0.0007557110</span></pre></body></html></div>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>4. <span id="references">References</span>
</h2>
<p>Aubry S., Kelly S., Kümpers B. M., Smith-Unna R. D., and Hibberd J. M. (2014). <strong>Deep evolutionary comparison of gene expression identifies parallel recruitment of trans-factors in two independent origins of C4 photosynthesis.</strong> <em>PLoS genetics</em>, <strong>10(6)</strong>. <a href="https://doi.org/10.1371/journal.pgen.1004365">https://doi.org/10.1371/journal.pgen.1004365</a></p>
<p>Charif D., and Lobry J. R. (2007). <strong>SeqinR 1.0-2: a contributed package to the R project for statistical computing devoted to biological sequences retrieval and analysis.</strong> <em>In Structural approaches to sequence evolution</em> (pp. 207-232). Springer, Berlin, Heidelberg. <a href="https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10">https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10</a></p>
<p>Kiełbasa S. M., Wan R., Sato K., Horton P., and Frith M. C. (2011). <strong>Adaptive seeds tame genomic sequence comparison.</strong> <em>Genome research</em>, 21(3), 487-493. <a href="https://doi.org/10.1101/gr.113985.110">https://doi.org/10.1101/gr.113985.110</a></p>
<p>Li W. H. (1993). <strong>Unbiased estimation of the rates of synonymous and nonsynonymous substitution.</strong> <em>Journal of molecular evolution</em>, 36(1), 96-99. <a href="https://doi.org/10.1007/bf02407308">https://doi.org/10.1007/bf02407308</a></p>
<p>Pagès H., Aboyoun P., Gentleman R., and DebRoy S. (2017). <strong>Biostrings: Efficient manipulation of biological strings.</strong> <em>R package version</em>, 2(0).</p>
<p>Rost B. (1999). <strong>Twilight zone of protein sequence alignments.</strong> <em>Protein engineering</em>, 12(2), 85-94. <a href="https://doi.org/10.1093/protein/12.2.85">https://doi.org/10.1093/protein/12.2.85</a></p>
<p>Scott C. (2017). <strong>shmlast: an improved implementation of conditional reciprocal best hits with LAST and Python.</strong> <em>Journal of Open Source Software</em>, 2(9), 142. <a href="https://joss.theoj.org/papers/10.21105/joss.00142">https://joss.theoj.org/papers/10.21105/joss.00142</a></p>
<p>Yang Z., and Nielsen R. (2000). <strong>Estimating synonymous and nonsynonymous substitution rates under realistic evolutionary models.</strong> <em>Molecular biology and evolution</em>, 17(1), 32-43. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">https://doi.org/10.1093/oxfordjournals.molbev.a026236</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
