<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>KaKs Vignette • CRBHits</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="KaKs Vignette">
<meta property="og:description" content="CRBHits">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CRBHits</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>
    </li>
    <li>
      <a href="../articles/V02KaKsVignette.html">KaKs Vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="V02KaKsVignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="V02KaKsVignette_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="V02KaKsVignette_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>KaKs Vignette</h1>
                        <h4 class="author">Kristian K Ullrich</h4>
            
            <h4 class="date">2020-10-31</h4>
      
      <small class="dont-index">Source: <a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits/blob/mastervignettes/V02KaKsVignette.Rmd"><code>vignettes/V02KaKsVignette.Rmd</code></a></small>
      <div class="hidden name"><code>V02KaKsVignette.Rmd</code></div>

    </div>

    
    
<div id="kaks-vignette" class="section level1">
<h1 class="hasAnchor">
<a href="#kaks-vignette" class="anchor"></a>Ka/Ks Vignette</h1>
<ul>
<li>includes <em>A. thalina</em> and <em>A. lyrata</em> CRBHit pair calculation</li>
<li>includes <em>H. sapiens</em> and <em>P. troglodytes</em> CRBHit pair calculation</li>
<li>inlcudes Longest Isoform selection</li>
<li>includes Gene/Isoform chromosomal position extraction</li>
<li>includes Tandem Duplicate Assignment</li>
<li>includes Ka/Ks colored Dot-Plot</li>
</ul>
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li>
<a href="#crbhpairs">Conditional Reciprocal Best Hit pairs (CRBH pairs)</a>
<ol style="list-style-type: decimal">
<li><a href="#ensembl">Download Coding Sequences (CDS) from NCBI and ENSEMBL</a></li>
<li><a href="#isoform">Get Longest Isoform from NCBI or ENSEMBL Input</a></li>
<li><a href="#crbhfilter">Calculate/Filter CRBHit pairs</a></li>
<li>
<a href="#extractposition">Extract Gene/Isoform chromosomal position</a>
<ol style="list-style-type: decimal">
<li><a href="#ncbi2genepos">Get gene position from NCBI or ENSEMBL Input</a></li>
<li><a href="#gff2genepos">Use GTF/GFF3 file to obtain gene position</a></li>
</ol>
</li>
<li><a href="#tandemduplicates">Assign Tandem Duplicates</a></li>
<li><a href="#synteny">Synteny with DAGchainer</a></li>
</ol>
</li>
<li><a href="#KaKs">Ka/Ks Calculation</a></li>
<li><a href="#KaKsFiltering">Ka/Ks Filtering and Visualisation</a></li>
<li><a href="#homosapiens"><em>Homo sapiens</em> vs. <em>Pan troglodytes</em> example</a></li>
<li><a href="#references">References</a></li>
</ol>
<p>This vignette is supposed to explain in more detail Ka/Ks calculation and it’s downstream filtering and visualization. The Ka/Ks ratio was originally developed for the analysis of genetic sequences from divergent species (<a href="https://www.nature.com/articles/267275a0">Kimura (1977)</a>, <a href="(https://doi.org/10.1093/oxfordjournals.molbev.a026236)">Yang and Nielson (2000)</a>). In short, the Ka/Ks ratio quantifies the mode and strength of selection by comparing synonymous substitution rates (Ks) (assumed to be evolutionary neutral) with nonsynonymous rates (Ka), which are exposed to selection as they change the amino acid composition of a protein (<a href="https://academic.oup.com/mbe/article/31/1/212/1049986">Mugal et al (2014)</a>).</p>
<p><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a> is a reimplementation of the Conditional Reciprocal Best Hit (CRBH) algorithm <a href="https://github.com/cboursnell/crb-blast">crb-blast</a> in R. The CRBHit pairs can be directly used to calculate Ka/Ks ratios and to filter for tandem duplicates or syntenic groups.</p>
<p>See the R package page for a detailed description of the install process and its dependencies <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/">https://mpievolbio-it.pages.gwdg.de/crbhits/</a> or have a look at the <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>.</p>
<p><strong>Note:</strong></p>
<p>In this Vignette the <code>lastpath</code> is defined as <code>vignette.paths[1]</code> and the <code>kakscalcpath</code> as <code>vignette.paths[2]</code> to be able to build the Vignette.</p>
<p>However, once you have compiled last-1133 and KaKs_Calculator2.0 with the functions <code><a href="../reference/make.last.html">make.last()</a></code> and <code><a href="../reference/make.KaKs_Calculator2.html">make.KaKs_Calculator2()</a></code> you won’t need to specify the paths anymore. Please remove them if you would like to repeat the examples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co">#load vignette specific libraries</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.gwdg.de/mpievolbio-it/crbhits">CRBHits</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">Biostrings</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.cran.dev/curl">curl</a></span><span class="op">)</span><span class="op">)</span>
<span class="co">#compile LAST and KaKs_Calculator2.0 for the vignette</span>
<span class="va">vignette.paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.vignette.html">make.vignette</a></span><span class="op">(</span><span class="op">)</span></pre></div>
</div>
<div id="conditional-reciprocal-best-hit-pairs" class="section level2">
<h2 class="hasAnchor">
<a href="#conditional-reciprocal-best-hit-pairs" class="anchor"></a>1. <span id="crbhpairs">Conditional reciprocal best hit pairs</span>
</h2>
<p>The CRBH algorithm builds upon the classical reciprocal best hit (RBH) approach to find orthologous sequences between two sets of sequences by defining an expect-value cutoff per alignment length. Further, this cutoff is used to retain secondary hits as additional <em>bona-fide</em> homologues.</p>
<p>See the R package page and the <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a> for a more detailed description of the CRBH algorithm.</p>
<div id="download-coding-sequences-cds-from-ncbi-and-ensembl" class="section level3">
<h3 class="hasAnchor">
<a href="#download-coding-sequences-cds-from-ncbi-and-ensembl" class="anchor"></a>1.1. <span id="ensembl">Download Coding sequences (CDS) from NCBI and ENSEMBL</span>
</h3>
<p>To calculate CRBHit pairs between two species, one can directly use an URL to access the coding sequences and calculate the CRBH pairs matrix.</p>
<p>Two examples are given in this vignette.</p>
<p>The first example compares the coding sequences from <em>Arabidopsis thaliana</em> and <em>Arabidopsis lyrata</em>. The input sequences are from the FTP server from <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/all/">NCBI Genomes</a>.</p>
<p>The second example compares the CDS from <em>Homo sapiens</em> and <em>Pan troglodytes</em> (see <a href="#homosapiens"><em>Homo sapiens</em> vs. <em>Pan troglodytes</em> example</a>).</p>
<p>As an alternative the <a href="ftp://ftp.ensemblgenomes.org/pub/plants/release-48/fasta/">EnsemblPlants release-48</a> can used.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">##set URLs for Arabidopis thaliana and Arabidopsis lyrata from NCBI Genomes</span>

<span class="co">#set NCBI URL</span>
<span class="va">NCBI</span> <span class="op">&lt;-</span> <span class="st">"https://ftp.ncbi.nlm.nih.gov/genomes/all/"</span>
<span class="co">#set Arabidopsis thaliana CDS URL</span>
<span class="va">ARATHA.cds.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">NCBI</span>, <span class="st">"GCF/000/001/735/GCF_000001735.4_TAIR10.1/"</span>,
                      <span class="st">"GCF_000001735.4_TAIR10.1_cds_from_genomic.fna.gz"</span><span class="op">)</span>

<span class="co">#set Arabidopsis lyrata CDS URL</span>
<span class="va">ARALYR.cds.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">NCBI</span>, <span class="st">"GCF/000/004/255/GCF_000004255.2_v.1.0/"</span>,
                      <span class="st">"GCF_000004255.2_v.1.0_cds_from_genomic.fna.gz"</span><span class="op">)</span>
<span class="co">#get Arabidopsis thaliana CDS</span>
<span class="va">ARATHA.cds</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">ARATHA.cds.url</span><span class="op">)</span>
<span class="co">#get Arabidopsis lyrata CDS</span>
<span class="va">ARALYR.cds</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">ARALYR.cds.url</span><span class="op">)</span></pre></div>
</div>
<div id="get-longest-isoform-from-ncbi-or-ensembl-input" class="section level3">
<h3 class="hasAnchor">
<a href="#get-longest-isoform-from-ncbi-or-ensembl-input" class="anchor"></a>1.2. <span id="isoform">Get Longest Isoform from NCBI or ENSEMBL Input</span>
</h3>
<p>Prior CRBHit pairs calculation, the longest isoforms can be selected, if the data was downloaded from NCBI or Ensembl.</p>
<p><strong>Note:</strong> This is an important step. If e.g. isoforms are used, the Ka/Ks calculations might show biased values inspecting mean or median Ka/Ks values grouped per query sequence.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co">##get longest isoforms</span>

<span class="co">#get longest isoform</span>
<span class="va">ARATHA.cds.longest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="va">ARATHA.cds</span>, <span class="st">"NCBI"</span><span class="op">)</span>
<span class="va">ARALYR.cds.longest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="va">ARALYR.cds</span>, <span class="st">"NCBI"</span><span class="op">)</span>

<span class="co">#get help ?cds2longest</span></pre></div>
</div>
<div id="calculatefilter-crbhit-pairs" class="section level3">
<h3 class="hasAnchor">
<a href="#calculatefilter-crbhit-pairs" class="anchor"></a>1.3. <span id="crbhfilter">Calculate/Filter CRBHit pairs</span>
</h3>
<p>The blast-like software <a href="http://last.cbrc.jp/">LAST</a> is used to compare the translated CDS against each other and output a blast-like output table including the query and target length.</p>
<p>The hit pairs are filtered prior fitting for a query coverage of 50% and the <strong>twilight zone of protein sequence alignments</strong> (<a href="https://academic.oup.com/peds/article/12/2/85/1550637">Rost B. (1999)</a>).</p>
<p>The CRBhit pairs can be calculated directly from the URLs with the function <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> using multiple threads including longest isoform selection.</p>
<p>For a detailed description of the <code><a href="../reference/cdsfile2rbh.html">cdsfile2rbh()</a></code> function and the <code><a href="../reference/cds2rbh.html">cds2rbh()</a></code> function with its individual filtering steps prior CRBH algorithm fitting see the <a href="https://mpievolbio-it.pages.gwdg.de/crbhits/articles/V01CRBHitsBasicVignette.html">CRBHits Basic Vignette</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co">##calculate CRBHit pairs for A. thaliana and A. lyrata using 2 threads</span>

<span class="co">#input from CDS obtained from NCBI</span>
<span class="co">#longest isoform selection</span>
<span class="co">#query coverage &gt;= 50%</span>
<span class="co">#rost199 filter</span>
<span class="va">ARATHA_ARALYR_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">ARATHA.cds</span>,
                              <span class="va">ARALYR.cds</span>,
                              qcov <span class="op">=</span> <span class="fl">0.5</span>, rost1999 <span class="op">=</span> <span class="cn">TRUE</span>,
                              longest.isoform <span class="op">=</span> <span class="cn">TRUE</span>,
                              isoform.source <span class="op">=</span> <span class="st">"NCBI"</span>,
                              threads <span class="op">=</span> <span class="fl">2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                              lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></pre></div>
<p><img src="V02KaKsVignette_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb5"><pre class="downlit">

<span class="co">#get help ?cds2rbh</span></pre></div>
</div>
<div id="extract-geneisoform-chromosomal-position" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-geneisoform-chromosomal-position" class="anchor"></a>1.4. <span id="extractposition">Extract Gene/Isoform chromosomal position</span>
</h3>
<p>To be able to filter for <em>Tandem Duplicates</em> and to plot a dotplot, one needs to have annotated gene positions per chromosome/contig. With the <code><a href="../reference/cds2genepos.html">cds2genepos()</a></code> function it is possible to directly access this information if the data was obtained from NCBI or Ensembl.</p>
<p>Chromosomal gene positions might overlap and if one has not filtered to use the longest or primary isoform there will be positions overlap for CDS. To overcome this issue it is recommended to first reduce the CDS Input to the longest isoform (see <a href="#isoform">Get Longest Isoform from NCBI or ENSEMBL Input</a>).</p>
<p><strong>Note:</strong> It is also possible to obtain and define chromosomal gene positions from other sources, like parsing a GTF/GFF file and supply a manual curated genepos matrix. In the same way this process might be used to select the longest/primary isoform from CDS Input sources other than NCBI or Ensembl. This special case is handled here:</p>
<div id="get-gene-position-from-ncbi-or-ensembl-input" class="section level4">
<h4 class="hasAnchor">
<a href="#get-gene-position-from-ncbi-or-ensembl-input" class="anchor"></a>1.4.1. <span id="ncbi2genepos">Get gene position from NCBI or ENSEMBL Input</span>
</h4>
<p>If the CDS Input was obtained from NCBI or ENSEMBL, the gene position can be directly extracted from the <code>DNAStringSet</code> as follows.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co">##get gene position idx from NCBI CDS</span>

<span class="co">#extract gene position from CDS</span>
<span class="va">ARATHA.cds.genepos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2genepos.html">cds2genepos</a></span><span class="op">(</span><span class="va">ARATHA.cds</span>, source <span class="op">=</span> <span class="st">"NCBI"</span><span class="op">)</span>
<span class="co">#extract gene position from longest isoform CDS</span>
<span class="va">ARATHA.cds.longest.genepos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2genepos.html">cds2genepos</a></span><span class="op">(</span><span class="va">ARATHA.cds.longest</span>, source <span class="op">=</span> <span class="st">"NCBI"</span><span class="op">)</span>
<span class="co">#show first entries</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ARATHA.cds.genepos</span><span class="op">)</span>
<span class="co">#&gt;                             gene.seq.id    gene.chr gene.start gene.end</span>
<span class="co">#&gt; 1 lcl|NC_000932.1_cds_NP_051039.1_48182 NC_000932.1        383     1444</span>
<span class="co">#&gt; 2 lcl|NC_000932.1_cds_NP_051040.2_48183 NC_000932.1       2056     3570</span>
<span class="co">#&gt; 3 lcl|NC_000932.1_cds_NP_051041.1_48184 NC_000932.1       5084     6188</span>
<span class="co">#&gt; 4 lcl|NC_000932.1_cds_NP_051042.1_48185 NC_000932.1       7017     7202</span>
<span class="co">#&gt; 5 lcl|NC_000932.1_cds_NP_051043.1_48186 NC_000932.1       7583     7693</span>
<span class="co">#&gt; 6 lcl|NC_000932.1_cds_NP_051044.1_48187 NC_000932.1       9938    11461</span>
<span class="co">#&gt;   gene.mid gene.strand gene.idx</span>
<span class="co">#&gt; 1    913.5          -1        1</span>
<span class="co">#&gt; 2   2813.0          -1        2</span>
<span class="co">#&gt; 3   5636.0          -1        3</span>
<span class="co">#&gt; 4   7109.5           1        4</span>
<span class="co">#&gt; 5   7638.0           1        5</span>
<span class="co">#&gt; 6  10699.5          -1        6</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ARATHA.cds.longest.genepos</span><span class="op">)</span>
<span class="co">#&gt;                             gene.seq.id    gene.chr gene.start gene.end</span>
<span class="co">#&gt; 1 lcl|NC_000932.1_cds_NP_051039.1_48182 NC_000932.1        383     1444</span>
<span class="co">#&gt; 2 lcl|NC_000932.1_cds_NP_051040.2_48183 NC_000932.1       2056     3570</span>
<span class="co">#&gt; 3 lcl|NC_000932.1_cds_NP_051041.1_48184 NC_000932.1       5084     6188</span>
<span class="co">#&gt; 4 lcl|NC_000932.1_cds_NP_051042.1_48185 NC_000932.1       7017     7202</span>
<span class="co">#&gt; 5 lcl|NC_000932.1_cds_NP_051043.1_48186 NC_000932.1       7583     7693</span>
<span class="co">#&gt; 6 lcl|NC_000932.1_cds_NP_051044.1_48187 NC_000932.1       9938    11461</span>
<span class="co">#&gt;   gene.mid gene.strand gene.idx</span>
<span class="co">#&gt; 1    913.5          -1        1</span>
<span class="co">#&gt; 2   2813.0          -1        2</span>
<span class="co">#&gt; 3   5636.0          -1        3</span>
<span class="co">#&gt; 4   7109.5           1        4</span>
<span class="co">#&gt; 5   7638.0           1        5</span>
<span class="co">#&gt; 6  10699.5          -1        6</span>
<span class="co">#get number of gene isoforms with same index</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ARATHA.cds.genepos</span><span class="op">$</span><span class="va">gene.idx</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     1     2     3     4     5     6     7     8     9    10    11    12    15 </span>
<span class="co">#&gt; 29602  5062  1441   495   197    86    40    18    16     5     6     3     1</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">ARATHA.cds.longest.genepos</span><span class="op">$</span><span class="va">gene.idx</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     1 </span>
<span class="co">#&gt; 27562</span>

<span class="co">#get help ?cds2genepos</span></pre></div>
</div>
<div id="use-gtfgff3-file-to-obtain-gene-position" class="section level4">
<h4 class="hasAnchor">
<a href="#use-gtfgff3-file-to-obtain-gene-position" class="anchor"></a>1.4.2. <span id="gff2genepos">Use GTF/GFF3 file to obtain gene position</span>
</h4>
<p>As noted above, it is also possible to obtain gene position and corresponding the longest isoforms per gene annotated in a GTF/GFF3 file.</p>
<p>see <a href="http://www.ensembl.org/info/website/upload/gff3.html">GTF/GFF3 format</a></p>
<p><strong>Note:</strong> In some cases the IDs provided in the CDS and the GTF/GFF3 file differ, which make it more complicated to directly link gene position and CDS.</p>
<p>Here, as an example, the GFF3 file for the species <em>Arabidopsis thaliana</em> is obtained from the FTP server from <a href="https://plants.ensembl.org/index.html">Ensembl Plants</a>.</p>
<p><strong>Note:</strong> The GFF3 and GTF files can be very large and it might take some time until this data is loaded into memory.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co">##get gene position idx from GTF/GFF3</span>

<span class="co">#set NCBI URL</span>
<span class="va">ensemblPlants</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ensemblgenomes.org/pub/plants/release-48/"</span>
<span class="co">#set Arabidopsis thaliana GFF3 URL</span>
<span class="va">ARATHA.GFF.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">ensemblPlants</span>, <span class="st">"gff3/arabidopsis_thaliana/"</span>,
                      <span class="st">"Arabidopsis_thaliana.TAIR10.48.gff3.gz"</span><span class="op">)</span>
<span class="co">#downlaod and gunzip file</span>
<span class="va">ARATHA.GFF.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">ARATHA.GFF.file.gz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">ARATHA.GFF.file</span>, <span class="st">".gz"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">ARATHA.GFF.url</span>, <span class="va">ARATHA.GFF.file.gz</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.html">system</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"gunzip -f "</span>, <span class="va">ARATHA.GFF.file.gz</span><span class="op">)</span><span class="op">)</span>
<span class="co">#read ARATHA.GFF.file</span>
<span class="va">ARATHA.gff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="va">ARATHA.GFF.file</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, 
                         quote <span class="op">=</span> <span class="st">""</span>, header <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ARATHA.gff</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seqname"</span>, <span class="st">"source"</span>, <span class="st">"feature"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"score"</span>, <span class="st">"strand"</span>, <span class="st">"frame"</span>, <span class="st">"attribute"</span><span class="op">)</span>

<span class="co">#extract gene</span>
<span class="va">ARATHA.gff.gene</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">)</span>
<span class="co">#get gene ID</span>
<span class="va">ARATHA.gff.gene.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"ID\\="</span>, <span class="st">""</span>, 
                           <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/word.html">word</a></span><span class="op">(</span><span class="va">ARATHA.gff.gene</span><span class="op">$</span><span class="va">attribute</span>,
                           <span class="fl">1</span>, sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#extract mRNA</span>
<span class="va">ARATHA.gff.mRNA</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">==</span> <span class="st">"mRNA"</span><span class="op">)</span>
<span class="co">#get mRNA ID</span>
<span class="va">ARATHA.gff.mRNA.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"ID\\="</span>, <span class="st">""</span>, 
                           <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/word.html">word</a></span><span class="op">(</span><span class="va">ARATHA.gff.mRNA</span><span class="op">$</span><span class="va">attribute</span>,
                           <span class="fl">1</span>, sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#get mRNA based gene ID</span>
<span class="va">ARATHA.gff.mRNA.parent.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"Parent\\="</span>, <span class="st">""</span>, 
                                  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/word.html">word</a></span><span class="op">(</span><span class="va">ARATHA.gff.mRNA</span><span class="op">$</span><span class="va">attribute</span>,
                                  <span class="fl">2</span>, sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#add mRNA ID and mRNA based gene ID</span>
<span class="va">ARATHA.gff.mRNA</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.mRNA</span> <span class="op">%&gt;%</span> 
                     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gene.id <span class="op">=</span> <span class="va">ARATHA.gff.mRNA.parent.id</span>,
                                   mRNA.id <span class="op">=</span> <span class="va">ARATHA.gff.mRNA.id</span><span class="op">)</span>
<span class="co">#extract CDS</span>
<span class="va">ARATHA.gff.CDS</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">feature</span> <span class="op">==</span> <span class="st">"CDS"</span><span class="op">)</span>
<span class="co">#get CDS based mRNA ID</span>
<span class="va">ARATHA.gff.CDS.parent.id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"Parent\\="</span>, <span class="st">""</span>, 
                                 <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/word.html">word</a></span><span class="op">(</span><span class="va">ARATHA.gff.CDS</span><span class="op">$</span><span class="va">attribute</span>,
                                 <span class="fl">2</span>, sep<span class="op">=</span><span class="st">";"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#add CDS based mRNA ID</span>
<span class="va">ARATHA.gff.CDS</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.CDS</span> <span class="op">%&gt;%</span>
                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>mRNA.id <span class="op">=</span> <span class="va">ARATHA.gff.CDS.parent.id</span><span class="op">)</span>
<span class="co">#get width per mRNA isoform</span>
<span class="va">ARATHA.gff.mRNA.len</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.CDS</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">mRNA.id</span><span class="op">)</span> <span class="op">%&gt;%</span>
                         <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>mRNA.id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">mRNA.id</span><span class="op">)</span>,
                                          len <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">end</span><span class="op">-</span><span class="va">start</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; `summarise()` ungrouping output (override with `.groups` argument)</span>
<span class="co">#add mRNA isoform width</span>
<span class="va">ARATHA.gff.mRNA</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.mRNA</span> <span class="op">%&gt;%</span> 
                    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>mRNA.len <span class="op">=</span> 
                    <span class="va">ARATHA.gff.mRNA.len</span><span class="op">$</span><span class="va">len</span><span class="op">[</span>
                        <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">ARATHA.gff.mRNA</span><span class="op">$</span><span class="va">mRNA.id</span>, 
                              <span class="va">ARATHA.gff.mRNA.len</span><span class="op">$</span><span class="va">mRNA.id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#retain only longest mRNA isoform</span>
<span class="va">ARATHA.gff.mRNA.longest</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.mRNA</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">seqname</span>, <span class="va">start</span>,
  <span class="va">gene.id</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">mRNA.len</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">gene.id</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gene.mid <span class="op">=</span> <span class="op">(</span><span class="va">start</span><span class="op">+</span><span class="va">end</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#add gene idx</span>
<span class="va">ARATHA.gff.mRNA.longest</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.mRNA.longest</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gene.idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">1</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ARATHA.gff.mRNA.longest</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="co">#create gene position matrix to be used for downstream analysis</span>
<span class="va">ARATHA.gff.genepos</span> <span class="op">&lt;-</span> <span class="va">ARATHA.gff.mRNA.longest</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span>
  gene.seq.id <span class="op">=</span> <span class="va">mRNA.id</span>,
  gene.chr <span class="op">=</span> <span class="va">seqname</span>,
  gene.start <span class="op">=</span> <span class="va">start</span>,
  gene.end <span class="op">=</span> <span class="va">end</span>,
  gene.mid <span class="op">=</span> <span class="va">gene.mid</span>,
  gene.strand <span class="op">=</span> <span class="va">strand</span>,
  gene.idx <span class="op">=</span> <span class="va">gene.idx</span><span class="op">)</span>
<span class="co">#set attribute</span>
<span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="va">ARATHA.gff.genepos</span>, <span class="st">"CRBHits.class"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"genepos"</span>
<span class="co">#show first entries of generated gene position</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ARATHA.gff.genepos</span><span class="op">)</span>
<span class="co">#&gt;              gene.seq.id gene.chr gene.start gene.end gene.mid gene.strand</span>
<span class="co">#&gt; 1 transcript:AT1G01010.1        1       3631     5899   4765.0           +</span>
<span class="co">#&gt; 2 transcript:AT1G01020.1        1       6788     9130   7959.0           -</span>
<span class="co">#&gt; 3 transcript:AT1G01030.1        1      11649    13714  12681.5           -</span>
<span class="co">#&gt; 4 transcript:AT1G01040.1        1      23121    31227  27174.0           +</span>
<span class="co">#&gt; 5 transcript:AT1G01050.1        1      31170    33171  32170.5           -</span>
<span class="co">#&gt; 6 transcript:AT1G01060.6        1      33365    37871  35618.0           -</span>
<span class="co">#&gt;   gene.idx</span>
<span class="co">#&gt; 1        1</span>
<span class="co">#&gt; 2        2</span>
<span class="co">#&gt; 3        3</span>
<span class="co">#&gt; 4        4</span>
<span class="co">#&gt; 5        5</span>
<span class="co">#&gt; 6        6</span></pre></div>
</div>
</div>
<div id="assign-tandem-duplicates" class="section level3">
<h3 class="hasAnchor">
<a href="#assign-tandem-duplicates" class="anchor"></a>1.5 <span id="tandemduplicates">Assign Tandem Duplicates</span>
</h3>
<p>Tandem duplicated genes (gene family members that are tightly clustered on a chromosome) are common in plant genomes and can show different Ka/Ks distribution than non tandem duplicates or other segmental duplicated genes (<a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.0020115">Rizzon et al., 2006</a>).</p>
<p>In order to account for this, it is possible to classify these tandem duplicates with the <code><a href="../reference/tandemdups.html">tandemdups()</a></code> function.</p>
<p><strong>Note:</strong> To be able to classify tandem duplicated genes within one genome, one needs to calculate selfblast based CRBHit pairs and provide gene position information for the same input data.</p>
<p>The following example shows how one can:</p>
<ol style="list-style-type: decimal">
<li>get selfblast CRBHit pairs</li>
<li>get gene positions</li>
<li>assign tandem duplicates</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="co">##example to assign tandem duplicates given selfblast CRBHit pairs and gene position</span>

<span class="co">#get selfblast CRBHit pairs for A. thaliana</span>
<span class="va">ARATHA_selfblast_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">ARATHA.cds</span>,
                              <span class="va">ARATHA.cds</span>,
                              qcov <span class="op">=</span> <span class="fl">0.5</span>, rost1999 <span class="op">=</span> <span class="cn">TRUE</span>,
                              longest.isoform <span class="op">=</span> <span class="cn">TRUE</span>,
                              isoform.source <span class="op">=</span> <span class="st">"NCBI"</span>,
                              threads <span class="op">=</span> <span class="fl">2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                              lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></pre></div>
<p><img src="V02KaKsVignette_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit">

<span class="co">#get selfblast CRBHit pairs for A. lyrata</span>
<span class="va">ARALYR_selfblast_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">ARALYR.cds</span>,
                              <span class="va">ARALYR.cds</span>,
                              qcov <span class="op">=</span> <span class="fl">0.5</span>, rost1999 <span class="op">=</span> <span class="cn">TRUE</span>,
                              longest.isoform <span class="op">=</span> <span class="cn">TRUE</span>,
                              isoform.source <span class="op">=</span> <span class="st">"NCBI"</span>,
                              threads <span class="op">=</span> <span class="fl">2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                              lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></pre></div>
<p><img src="V02KaKsVignette_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit">

<span class="co">#get gene position for A. thaliana longest isoforms</span>
<span class="va">ARATHA.cds.longest.genepos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2genepos.html">cds2genepos</a></span><span class="op">(</span><span class="va">ARATHA.cds.longest</span>, source <span class="op">=</span> <span class="st">"NCBI"</span><span class="op">)</span>
<span class="co">#get gene position for A. lyrata longest isoforms</span>
<span class="va">ARALYR.cds.longest.genepos</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2genepos.html">cds2genepos</a></span><span class="op">(</span><span class="va">ARALYR.cds.longest</span>, source <span class="op">=</span> <span class="st">"NCBI"</span><span class="op">)</span>

<span class="co">#assign tandem duplicates for A. thaliana</span>
<span class="va">ARATHA.cds.longest.tandemdups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tandemdups.html">tandemdups</a></span><span class="op">(</span><span class="va">ARATHA_selfblast_crbh</span>,
                                            <span class="va">ARATHA.cds.longest.genepos</span>,
                                            dupdist <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="co">#assign tandem duplicates for A. lyrata</span>
<span class="va">ARALYR.cds.longest.tandemdups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tandemdups.html">tandemdups</a></span><span class="op">(</span><span class="va">ARALYR_selfblast_crbh</span>,
                                            <span class="va">ARALYR.cds.longest.genepos</span>,
                                            dupdist <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="co">#get help ?tandemdups</span></pre></div>
<p>The resulting tandem duplicated genes detected can be plotted given their chromosomal gene position as follows:</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co">##example how to plot tandem duplicated gene groups</span>

<span class="co">#get tandem group size</span>
<span class="va">tandem_group_size</span> <span class="op">&lt;-</span> <span class="va">ARATHA.cds.longest.tandemdups</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">tandem_group</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_data.html">group_size</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">tandem_group_size</span><span class="op">)</span>
<span class="co">#&gt; tandem_group_size</span>
<span class="co">#&gt;    2    3    4    5    6    7    8    9   10   11   13   14 </span>
<span class="co">#&gt; 1051  169   54   28   12    6    3    1    1    1    1    1</span>

<span class="co">#use dplyr::mutate to assign group size</span>
<span class="va">ARATHA.cds.longest.tandemdups</span> <span class="op">&lt;-</span> <span class="va">ARATHA.cds.longest.tandemdups</span> <span class="op">%&gt;%</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>tandem_group_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">tandem_group_size</span>,
  <span class="va">tandem_group_size</span><span class="op">)</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co">#use dplyr::group_by to plot group by chromosome and colored by group size</span>
<span class="va">ARATHA.cds.longest.tandemdups</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">gene.chr</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">gene.mid</span>, y <span class="op">=</span> <span class="va">gene.mid</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">20</span>, <span class="fu">aes</span><span class="op">(</span>col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">tandem_group_size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">gene.chr</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="../reference/CRBHitsColors.html">CRBHitsColors</a></span><span class="op">(</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="V02KaKsVignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div id="synteny-with-dagchainer" class="section level3">
<h3 class="hasAnchor">
<a href="#synteny-with-dagchainer" class="anchor"></a>1.6 <span id="synteny">Synteny with DAGchainer</span>
</h3>
<p>Chains of collinear gene pairs can represent segmentally duplicated regions and genes within a single genome or syntenic regions between related genomes (<a href="https://doi.org/10.1093/bioinformatics/bth397">Haas et al., 2004</a>).</p>
<p>There exists external tools which can be used to infer syntenic regions, like e.g. <a href="https://genomevolution.org/CoGe/SynMap.pl">SynMap</a> as published by <a href="https://academic.oup.com/bioinformatics/article/33/14/2197/3072872">Haug-Baltzell A. et al., 2017</a>.</p>
<p>Within CRBHits the tool <a href="http://dagchainer.sourceforge.net/">DAGchainer</a> published by <a href="https://doi.org/10.1093/bioinformatics/bth397">Haas et al., 2004</a> can be directly applied on the CRBHit pairs.</p>
<p>The following example shows how one can run <a href="http://dagchainer.sourceforge.net/">DAGchainer</a> on CRBHit pairs and gene positions to get syntenic information about the CRBHit pairs.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="co">##example how to run DAGchainer on CRBHit pairs and gene positions</span>

<span class="co">#DAGchainer using gene base pair (gene bp start end)</span>
<span class="co">#Note: change parameter to fit bp option</span>
<span class="va">ARATHA_ARALYR_crbh.dagchainer.bp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2dagchainer.html">rbh2dagchainer</a></span><span class="op">(</span>rbhpairs <span class="op">=</span> <span class="va">ARATHA_ARALYR_crbh</span>, 
                                                gene.position.cds1 <span class="op">=</span> <span class="va">ARATHA.cds.longest.genepos</span>,
                                                gene.position.cds2 <span class="op">=</span> <span class="va">ARALYR.cds.longest.genepos</span>,
                                                type <span class="op">=</span> <span class="st">"bp"</span>,
                                                gap_length <span class="op">=</span> <span class="fl">10000</span>,
                                                max_dist_allowed <span class="op">=</span> <span class="fl">200000</span>,
                                                dagchainerpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>

<span class="co">#DAGchainer using gene index (gene order)</span>
<span class="va">ARATHA_ARALYR_crbh.dagchainer.idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rbh2dagchainer.html">rbh2dagchainer</a></span><span class="op">(</span>rbhpairs <span class="op">=</span> <span class="va">ARATHA_ARALYR_crbh</span>, 
                                                gene.position.cds1 <span class="op">=</span> <span class="va">ARATHA.cds.longest.genepos</span>,
                                                gene.position.cds2 <span class="op">=</span> <span class="va">ARALYR.cds.longest.genepos</span>,
                                                type <span class="op">=</span> <span class="st">"idx"</span>,
                                                gap_length <span class="op">=</span> <span class="fl">1</span>,
                                                max_dist_allowed <span class="op">=</span> <span class="fl">20</span>,
                                                dagchainerpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="co">##example how to plot pairwise chromosomal syntenic groups (DAGchainer results)</span>

<span class="co">#plot DAGchainer results for each chromosome combination</span>
<span class="fu"><a href="../reference/plot.dagchainer.html">plot.dagchainer</a></span><span class="op">(</span><span class="va">ARATHA_ARALYR_crbh.dagchainer.bp</span><span class="op">)</span></pre></div>
<p><img src="V02KaKsVignette_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div id="kaks-calculation" class="section level2">
<h2 class="hasAnchor">
<a href="#kaks-calculation" class="anchor"></a>2. <span id="KaKs">Ka/Ks Calculation</span>
</h2>
<p>model = “Li”</p>
<pre><code>ARATHA.cds &lt;- Biostrings::readDNAStringSet(paste0(ensemblPlants, ARATHA.CDS.file))
ARALYR.cds &lt;- Biostrings::readDNAStringSet(paste0(ensemblPlants, ARALYR.CDS.file))
ARATHA_ARALYR_crbh.kaks.Li &lt;- rbh2kaks(ARATHA_ARALYR_crbh$crbh.pairs,
                                       ARATHA.cds, ARALYR.cds, model = "Li",
                                       threads = 4)</code></pre>
<p>model = “YN”</p>
<pre><code>HOMSAP_PANTRO_crbh.kaks.YN &lt;- rbh2kaks(HOMSAP_PANTRO_crbh$crbh.pairs,
                                       HOMSAP.cds, PANTRO.cds, model = "YN",
                                       threads = 8)</code></pre>
</div>
<div id="kaks-filtering-and-visualisation" class="section level2">
<h2 class="hasAnchor">
<a href="#kaks-filtering-and-visualisation" class="anchor"></a>3. <span id="KaKsFiltering">Ka/Ks Filtering and Visualisation</span>
</h2>
<p>In some cases the Ka/Ks calculation fails, since e.g. there are no synonymous (Ks) or nonsynonymous substitutions (Ka) between cds1 and cds2 or one wants to get rid of high Ks values due to substitution saturation. One can easily remove these cases before doing further analysis.</p>
<pre><code>#remove all values that are negative for either ka or ks
ath_aly_crbh.kaks.Li.red &lt;- ath_aly_crbh.kaks.Li %&gt;%
  subset(ka != 9.999999) %&gt;% subset(ks != 9.999999)</code></pre>
</div>
<div id="homo-sapiens-vs--pan-troglodytes-example" class="section level2">
<h2 class="hasAnchor">
<a href="#homo-sapiens-vs--pan-troglodytes-example" class="anchor"></a>4. <span id="homosapiens"><em>Homo sapiens</em> vs. <em>Pan troglodytes</em> example</span>
</h2>
<p>The second example compares the CDS from <em>Homo sapiens</em> and <em>Pan troglodytes</em>. The input sequences are from the FTP server from <a href="https://ensembl.org/index.html">Ensembl</a>.</p>
<p>The <a href="ftp://ftp.ensembl.org/pub/release-101/fasta/">Ensembl release-101</a> is used.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="co">###example comparing Homo sapiens and Pan troglodytes</span>

<span class="co">##set URLs for Homo sapiens and Pan troglodytes from Ensembl</span>

<span class="co">#set Ensembl URL</span>
<span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-101/fasta/"</span>
<span class="co">#set Homo sapiens CDS URL</span>
<span class="va">HOMSAP.cds.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">ensembl</span>, <span class="st">"homo_sapiens/cds/Homo_sapiens.GRCh38.cds.all.fa.gz"</span><span class="op">)</span>
<span class="co">#set Pan troglodytes CDS URL</span>
<span class="va">PANTRO.cds.url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">ensembl</span>, <span class="st">"pan_troglodytes/cds/Pan_troglodytes.Pan_tro_3.0.cds.all.fa.gz"</span><span class="op">)</span>

<span class="co">#get Homo sapiens CDS</span>
<span class="va">HOMSAP.cds.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">HOMSAP.cds.url</span>, <span class="va">HOMSAP.cds.file</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">HOMSAP.cds</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">HOMSAP.cds.file</span><span class="op">)</span>

<span class="co">#get Pan troglodytes CDS</span>
<span class="va">PANTRO.cds.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html">tempfile</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">PANTRO.cds.url</span>, <span class="va">PANTRO.cds.file</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">PANTRO.cds</span> <span class="op">&lt;-</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html">readDNAStringSet</a></span><span class="op">(</span><span class="va">PANTRO.cds.file</span><span class="op">)</span>

<span class="co">#get longest isoform</span>
<span class="va">HOMSAP.cds.longest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="va">HOMSAP.cds</span>, <span class="st">"ENSEMBL"</span><span class="op">)</span>
<span class="va">PANTRO.cds.longest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoform2longest.html">isoform2longest</a></span><span class="op">(</span><span class="va">PANTRO.cds</span>, <span class="st">"ENSEMBL"</span><span class="op">)</span>

<span class="co">##calculate CRBHit pairs for H. sapiens and P. troglodytes using 2 threads</span>

<span class="co">#input from CDS obtained from Ensembl</span>
<span class="co">#longest isoform selection</span>
<span class="co">#query coverage &gt;= 50%</span>
<span class="co">#rost199 filter</span>
<span class="va">HOMSAP_PANTRO_crbh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cds2rbh.html">cds2rbh</a></span><span class="op">(</span><span class="va">HOMSAP.cds</span>,
                              <span class="va">PANTRO.cds</span>,
                              qcov <span class="op">=</span> <span class="fl">0.5</span>, rost1999 <span class="op">=</span> <span class="cn">TRUE</span>,
                              longest.isoform <span class="op">=</span> <span class="cn">TRUE</span>,
                              isoform.source <span class="op">=</span> <span class="st">"ENSEMBL"</span>,
                              threads <span class="op">=</span> <span class="fl">2</span>, plotCurve <span class="op">=</span> <span class="cn">TRUE</span>,
                              lastpath <span class="op">=</span> <span class="va">vignette.paths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></pre></div>
<p><img src="V02KaKsVignette_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<pre><code>HOMSAP.cds &lt;- Biostrings::readDNAStringSet(paste0(ensembl, HOMSAP.CDS.file))
PANTRO.cds &lt;- Biostrings::readDNAStringSet(paste0(ensembl, PANTRO.CDS.file))
HOMSAP_PANTRO_crbh.kaks.Li &lt;- rbh2kaks(HOMSAP_PANTRO_crbh$crbh.pairs,
                                       HOMSAP.cds, PANTRO.cds, model = "Li",
                                       threads = 4)</code></pre>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>5. <span id="references">References</span>
</h2>
<p>Aubry S., Kelly S., Kümpers B. M., Smith-Unna R. D., and Hibberd J. M. (2014). <strong>Deep evolutionary comparison of gene expression identifies parallel recruitment of trans-factors in two independent origins of C4 photosynthesis.</strong> <em>PLoS genetics</em>, <strong>10(6)</strong>. <a href="https://doi.org/10.1371/journal.pgen.1004365">https://doi.org/10.1371/journal.pgen.1004365</a></p>
<p>Charif D., and Lobry J. R. (2007). <strong>SeqinR 1.0-2: a contributed package to the R project for statistical computing devoted to biological sequences retrieval and analysis.</strong> <em>In Structural approaches to sequence evolution</em> (pp. 207-232). Springer, Berlin, Heidelberg. <a href="https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10">https://link.springer.com/chapter/10.1007/978-3-540-35306-5_10</a></p>
<p>Haas B. J., Delcher A. L., Wortman J. R., and Salzberg S. L. (2004). <strong>DAGchainer: a tool for mining segmental genome duplications and synteny.</strong> <em>Bioinformatics</em>, <strong>20(18)</strong>. <a href="https://doi.org/10.1093/bioinformatics/bth397">https://doi.org/10.1093/bioinformatics/bth397</a></p>
<p>Haug-Baltzell A., Stephens S. A., Davey S., Scheidegger C. E., Lyons E. (2017). <strong>SynMap2 and SynMap3D: web-based wholge-genome synteny browsers.</strong> <em>Bioinformatics</em>, <strong>33(14)</strong>. <a href="https://academic.oup.com/bioinformatics/article/33/14/2197/3072872">https://academic.oup.com/bioinformatics/article/33/14/2197/3072872</a></p>
<p>Kiełbasa S. M., Wan R., Sato K., Horton P., and Frith M. C. (2011). <strong>Adaptive seeds tame genomic sequence comparison.</strong> <em>Genome research</em>, 21(3), 487-493. <a href="https://doi.org/10.1101/gr.113985.110">https://doi.org/10.1101/gr.113985.110</a></p>
<p>Kimura M. (1977). <strong>Preponderance of synonymous changes as evidence for the neutral theory of molecular evolution.</strong> <em>Nature</em>, 267, 275-276.</p>
<p>Li W. H. (1993). <strong>Unbiased estimation of the rates of synonymous and nonsynonymous substitution.</strong> <em>Journal of molecular evolution</em>, 36(1), 96-99. <a href="https://doi.org/10.1007/bf02407308">https://doi.org/10.1007/bf02407308</a></p>
<p>Microsoft, and Weston S. (2020). <strong>foreach: Provides Foreach Looping Construct.</strong> <em>R package version</em>, 1.5.1. <a href="https://cran.r-project.org/web/packages/foreach/index.html">foreach</a></p>
<p>Mugal C. F., Wolf J. B. W., Kaj I. (2014). <strong>Why Time Matters: Codon Evolution and the Temproal Dynamics of dN/dS.</strong> <em>Molecular Biology and Evolution</em>, 31(1), 212-231.</p>
<p>Ooms J. (2019). <strong>curl: A Modern and Flexible Web Client for R.</strong> <em>R package version</em>, 4.3. <a href="https://cran.r-project.org/web/packages/curl/index.html">curl</a></p>
<p>Pagès H., Aboyoun P., Gentleman R., and DebRoy S. (2017). <strong>Biostrings: Efficient manipulation of biological strings.</strong> <em>R package version</em>, 2.56.0. <a href="http://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a></p>
<p>Revolution Analytics, and Weston S. (2020). <strong>doMC: Foreach Parallel Adaptor for ‘parallel’.</strong> <em>R package version</em>, 1.3.7. <a href="https://cran.r-project.org/web/packages/doMC/index.html">doMC</a></p>
<p>Rost B. (1999). <strong>Twilight zone of protein sequence alignments.</strong> <em>Protein engineering</em>, 12(2), 85-94. <a href="https://doi.org/10.1093/protein/12.2.85">https://doi.org/10.1093/protein/12.2.85</a></p>
<p>Scott C. (2017). <strong>shmlast: an improved implementation of conditional reciprocal best hits with LAST and Python.</strong> <em>Journal of Open Source Software</em>, 2(9), 142. <a href="https://joss.theoj.org/papers/10.21105/joss.00142">https://joss.theoj.org/papers/10.21105/joss.00142</a></p>
<p>Scrucca L., Fop M., Murphy T. B., and Raftery A. E. (2016) <strong>mclust 5: clustering, classification and density estimation using Gaussian finite mixture models.</strong> <em>The R Journal</em>, 8(1), 289-317. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5096736/</a></p>
<p>Duong T., and Wand M. (2015). <strong>feature: Local Inferential Feature Significance for Multivariate Kernel Density Estimation.</strong> <em>R package version 1.2.13</em>. <a href="https://cran.r-project.org/web/packages/feature/">https://cran.r-project.org/web/packages/feature/</a></p>
<p>Wickham H. (2011). <strong>testthat: Get Started with Testing.</strong> <em>The R Journal</em>, 3(1), 5. <a href="https://cran.r-project.org/web/packages/testthat/index.html">testthat</a></p>
<p>Wickham H. (2019). <strong>stringr: Simple, Consistent Wrappers for Common String Operations.</strong> <em>R package version</em>, 1.4.0. <a href="https://cran.r-project.org/web/packages/stringr/index.html">stringr</a></p>
<p>Wickham H. (2020). <strong>tidyr: Tidy Messy Data.</strong> <em>R package version</em>, 1.1.2. <a href="https://cran.r-project.org/web/packages/tidyr/index.html">tidyr</a></p>
<p>Wickham H., Hester J., and Chang W. (2020). <strong>devtools: Tools to make Developing R Packages Easier.</strong> <em>R package version</em>, (2.3.2). <a href="https://cran.r-project.org/web/packages/devtools/index.html">devtools</a></p>
<p>Wickham H., François R., Henry L., and Müller K. (2020). <strong>dplyr: A Grammar of Data Manipulation.</strong> <em>R package version</em>, 1.0.2. <a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a></p>
<p>Yang Z., and Nielsen R. (2000). <strong>Estimating synonymous and nonsynonymous substitution rates under realistic evolutionary models.</strong> <em>Molecular biology and evolution</em>, 17(1), 32-43. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a026236">https://doi.org/10.1093/oxfordjournals.molbev.a026236</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kristian K Ullrich.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
